{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import \n",
    "import os\n",
    "import re\n",
    "import time\n",
    "from pprint import pprint \n",
    "from datetime import datetime, timedelta\n",
    "from typing import Annotated, List, Dict, Tuple, Optional, Union, Any\n",
    "import xml.etree.ElementTree as ET\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain_core.tools import tool\n",
    "from langchain_core.runnables import RunnableConfig\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"nick.youngblut@arcinstitute.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'database: sra, IDs: 36004814, 36004694, 36004643, 36004624, 36004101, 36003953, 36003799, 36003617, 36003381, 35966237, 35966236, 35966235, 35966234, 35966233, 35966232, 35960653, 35960652, 35860023, 35812982, 35812981'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@tool \n",
    "def esearch(\n",
    "    esearch_query: Annotated[str, \"Entrez query string.\"],\n",
    "    database: Annotated[str, \"Database name ('sra' or 'gds')\"]=\"sra\",\n",
    "    #organism: Annotated[str, \"Organism name ('human' or 'mouse')\"]=\"human\",\n",
    "    )-> Annotated[List[str], \"All tasks in the Asana project.\"]:\n",
    "    \"\"\"\n",
    "    Run an Entrez search query and return the IDs of the results.\n",
    "    \"\"\"\n",
    "    # date range\n",
    "    start_date = datetime.now() - timedelta(days=7)\n",
    "    end_date = datetime.now()\n",
    "    date_range = f\"{start_date.strftime('%Y/%m/%d')}:{end_date.strftime('%Y/%m/%d')}[PDAT]\"\n",
    "    esearch_query += f\" AND {date_range}\"\n",
    "\n",
    "    max_ids = 20  # DEBUG\n",
    "\n",
    "    # query\n",
    "    ids = []\n",
    "    retstart = 0\n",
    "    retmax = 50\n",
    "    while True:\n",
    "        try:\n",
    "            search_handle = Entrez.esearch(\n",
    "                db=database, \n",
    "                term=esearch_query, \n",
    "                retstart=retstart, \n",
    "                retmax=retmax\n",
    "            )\n",
    "            search_results = Entrez.read(search_handle)\n",
    "            search_handle.close()\n",
    "            ids.extend(search_results[\"IdList\"])\n",
    "            retstart += retmax\n",
    "            time.sleep(0.5)\n",
    "            if max_ids and len(ids) >= max_ids:\n",
    "                break\n",
    "            if retstart >= int(search_results['Count']):\n",
    "                break\n",
    "        except Exception as e:\n",
    "            print(f\"Error searching {database} with query: {esearch_query}: {str(e)}\")\n",
    "            break \n",
    "    # create return string\n",
    "    ids = \", \".join(ids[:max_ids])\n",
    "    return f\"database: {database}, IDs: {ids}\"\n",
    "\n",
    "query = '(\"single cell RNA sequencing\" OR \"single cell RNA-seq\")'\n",
    "IDs = esearch.invoke({\"esearch_query\" : query, \"database\" : \"sra\"})\n",
    "IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('<?xml version=\"1.0\" encoding=\"UTF-8\"  ?>\\n'\n",
      " '<EXPERIMENT_PACKAGE_SET>\\n'\n",
      " '<EXPERIMENT_PACKAGE><EXPERIMENT accession=\"SRX26623215\" '\n",
      " 'alias=\"220857C_LM_3\"><IDENTIFIERS><PRIMARY_ID>SRX26623215</PRIMARY_ID></IDENTIFIERS><TITLE>RNA-Seq '\n",
      " 'of Homo sapiens: adult female pancreas</TITLE><STUDY_REF '\n",
      " 'accession=\"SRP543450\"><IDENTIFIERS><PRIMARY_ID>SRP543450</PRIMARY_ID></IDENTIFIERS></STUDY_REF><DESIGN><DESIGN_DESCRIPTION>Single-cell '\n",
      " 'RNA-seq libraries were prepared using Single Cell 3 Library Gel Bead Kit V2 '\n",
      " 'following the manufactures introduction. Finally sequencing was performed on '\n",
      " 'an Illumina Novaseq6000 with a sequencing depth of at least 100,000 reads '\n",
      " 'per cell and pair end 150bp (PE150).</DESIGN_DESCRIPTION><SAMPLE_DESCRIPTOR '\n",
      " 'accession=\"SRS23119663\"><IDENTIFIERS><PRIMARY_ID>SRS23119663</PRIMARY_ID></IDENTIFIERS></SAMPLE_DESCRIPTOR><LIBRARY_DESCRIPTOR><LIBRARY_NAME>220857C_LM_3</LIBRARY_NAME><LIBRARY_STRATEGY>RNA-Seq</LIBRARY_STRATEGY><LIBRARY_SOURCE>TRANSCRIPTOMIC</LIBRARY_SOURCE><LIBRARY_SELECTION>RT-PCR</LIBRARY_SELECTION><LIBRARY_LAYOUT><PAIRED/></LIBRARY_LAYOUT></LIBRARY_DESCRIPTOR></DESIGN><PLATFORM><ILLUMINA><INSTRUMENT_MODEL>Illumina '\n",
      " 'NovaSeq '\n",
      " '6000</INSTRUMENT_MODEL></ILLUMINA></PLATFORM></EXPERIMENT><SUBMISSION '\n",
      " 'lab_name=\"general surgical\" center_name=\"Peking Union Medical College '\n",
      " 'Hospital\" accession=\"SRA2008902\" '\n",
      " 'alias=\"SUB14817120\"><IDENTIFIERS><PRIMARY_ID>SRA2008902</PRIMARY_ID><SUBMITTER_ID '\n",
      " 'namespace=\"Peking Union Medical College '\n",
      " 'Hospital\">SUB14817120</SUBMITTER_ID></IDENTIFIERS></SUBMISSION><Organization '\n",
      " 'type=\"institute\"><Name>Peking Union Medical College Hospital</Name><Address '\n",
      " 'postal_code=\"100000\"><Department>general '\n",
      " 'surgical</Department><Institution>Peking Union Medical College '\n",
      " 'Hospital</Institution><Street>Peking Union Medical College Hospital (East), '\n",
      " 'No.1 Shuaifuyuan,Wangfujing, Dongcheng '\n",
      " 'District</Street><City>Beijing</City><Sub>Beijing '\n",
      " 'Municipality</Sub><Country>China</Country></Address><Contact '\n",
      " 'email=\"m13717008081@163.com\" sec_email=\"hhr2015530274@outlook.com\"><Address '\n",
      " 'postal_code=\"100000\"><Department>general '\n",
      " 'surgical</Department><Institution>Peking Union Medical College '\n",
      " 'Hospital</Institution><Street>Peking Union Medical College Hospital (East), '\n",
      " 'No.1 Shuaifuyuan,Wangfujing, Dongcheng '\n",
      " 'District</Street><City>Beijing</City><Sub>Beijing '\n",
      " 'Municipality</Sub><Country>China</Country></Address><Name><First>Haoran</First><Last>Huang</Last></Name></Contact></Organization><STUDY '\n",
      " 'center_name=\"BioProject\" alias=\"PRJNA1182665\" '\n",
      " 'accession=\"SRP543450\"><IDENTIFIERS><PRIMARY_ID>SRP543450</PRIMARY_ID><EXTERNAL_ID '\n",
      " 'namespace=\"BioProject\" '\n",
      " 'label=\"primary\">PRJNA1182665</EXTERNAL_ID></IDENTIFIERS><DESCRIPTOR><STUDY_TITLE>scRNA '\n",
      " 'seq on paired samples from three advanced pancreatic cancer patients with '\n",
      " 'liver metastasis</STUDY_TITLE><STUDY_TYPE '\n",
      " 'existing_study_type=\"Other\"/><STUDY_ABSTRACT>scRNA seq on paired samples '\n",
      " 'from three advanced PC patients (P1-P3) with LM . Using resected tissue '\n",
      " 'specimens from LM allowed us to characterize the cellular components and '\n",
      " 'spatial distribution characteristics '\n",
      " 'comprehensively.</STUDY_ABSTRACT></DESCRIPTOR></STUDY><SAMPLE '\n",
      " 'alias=\"220857C_LM_3\" '\n",
      " 'accession=\"SRS23119663\"><IDENTIFIERS><PRIMARY_ID>SRS23119663</PRIMARY_ID><EXTERNAL_ID '\n",
      " 'namespace=\"BioSample\">SAMN44601143</EXTERNAL_ID><SUBMITTER_ID '\n",
      " 'namespace=\"pda|huanghr18\" label=\"Sample '\n",
      " 'name\">220857C_LM_3</SUBMITTER_ID></IDENTIFIERS><SAMPLE_NAME><TAXON_ID>9606</TAXON_ID><SCIENTIFIC_NAME>Homo '\n",
      " 'sapiens</SCIENTIFIC_NAME></SAMPLE_NAME><SAMPLE_LINKS><SAMPLE_LINK><XREF_LINK><DB>bioproject</DB><ID>1182665</ID><LABEL>PRJNA1182665</LABEL></XREF_LINK></SAMPLE_LINK></SAMPLE_LINKS><SAMPLE_ATTRIBUTES><SAMPLE_ATTRIBUTE><TAG>isolate</TAG><VALUE>LM_3</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>age</TAG><VALUE>54</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>biomaterial_provider</TAG><VALUE>Peking '\n",
      " 'Union Medical College Hospital (East), No.1 Shuaifuyuan,Wangfujing, '\n",
      " 'Dongcheng District, Beijing, '\n",
      " 'China</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>collection_date</TAG><VALUE>2022-02-05</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>geo_loc_name</TAG><VALUE>China: '\n",
      " 'Beijing</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>sex</TAG><VALUE>female</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>tissue</TAG><VALUE>liver</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>BioSampleModel</TAG><VALUE>Human</VALUE></SAMPLE_ATTRIBUTE></SAMPLE_ATTRIBUTES></SAMPLE><Pool><Member '\n",
      " 'member_name=\"\" accession=\"SRS23119663\" sample_name=\"220857C_LM_3\" '\n",
      " 'spots=\"227438532\" bases=\"68231559600\" tax_id=\"9606\" organism=\"Homo '\n",
      " 'sapiens\"><IDENTIFIERS><PRIMARY_ID>SRS23119663</PRIMARY_ID><EXTERNAL_ID '\n",
      " 'namespace=\"BioSample\">SAMN44601143</EXTERNAL_ID></IDENTIFIERS></Member></Pool><RUN_SET '\n",
      " 'runs=\"1\" bases=\"68231559600\" spots=\"227438532\" bytes=\"26895277702\"><RUN '\n",
      " 'accession=\"SRR31243301\" alias=\"220857C_LM_3_S1_L001_R1_001.fastq.gz\" '\n",
      " 'total_spots=\"227438532\" total_bases=\"68231559600\" size=\"26895277702\" '\n",
      " 'load_done=\"true\" published=\"2024-11-06 09:50:21\" is_public=\"true\" '\n",
      " 'cluster_name=\"public\" has_taxanalysis=\"1\" '\n",
      " 'static_data_available=\"1\"><IDENTIFIERS><PRIMARY_ID>SRR31243301</PRIMARY_ID><SUBMITTER_ID '\n",
      " 'namespace=\"SUB14817120\" '\n",
      " 'label=\"5\">220857C_LM_3_S1_L001_R1_001.fastq.gz</SUBMITTER_ID></IDENTIFIERS><EXPERIMENT_REF '\n",
      " 'accession=\"SRX26623215\"><IDENTIFIERS><SUBMITTER_ID '\n",
      " 'namespace=\"SUB14817120\">220857C_LM_3</SUBMITTER_ID></IDENTIFIERS></EXPERIMENT_REF><Pool><Member '\n",
      " 'member_name=\"\" accession=\"SRS23119663\" sample_name=\"220857C_LM_3\" '\n",
      " 'spots=\"227438532\" bases=\"68231559600\" tax_id=\"9606\" organism=\"Homo '\n",
      " 'sapiens\"><IDENTIFIERS><PRIMARY_ID>SRS23119663</PRIMARY_ID><EXTERNAL_ID '\n",
      " 'namespace=\"BioSample\">SAMN44601143</EXTERNAL_ID></IDENTIFIERS></Member></Pool><SRAFiles><SRAFile '\n",
      " 'cluster=\"public\" filename=\"220857C_LM_3_S1_L001_R1_001.fastq.gz\" '\n",
      " 'size=\"11485619886\" date=\"2024-11-06 09:20:10\" '\n",
      " 'md5=\"ef1da9c26515146df1b6433a187ea71c\" version=\"1\" semantic_name=\"fastq\" '\n",
      " 'supertype=\"Original\" sratoolkit=\"0\"><Alternatives '\n",
      " 'url=\"s3://sra-pub-src-16/SRR31243301/220857C_LM_3_S1_L001_R1_001.fastq.gz.1\" '\n",
      " 'free_egress=\"-\" access_type=\"Use Cloud Data Delivery\" '\n",
      " 'org=\"AWS\"/></SRAFile><SRAFile cluster=\"public\" '\n",
      " 'filename=\"220857C_LM_3_S1_L001_R2_001.fastq.gz\" size=\"9178998612\" '\n",
      " 'date=\"2024-11-06 09:20:31\" md5=\"9f1777d2856e411d9ff711ef5832ffc7\" '\n",
      " 'version=\"1\" semantic_name=\"fastq\" supertype=\"Original\" '\n",
      " 'sratoolkit=\"0\"><Alternatives '\n",
      " 'url=\"s3://sra-pub-src-16/SRR31243301/220857C_LM_3_S1_L001_R2_001.fastq.gz.1\" '\n",
      " 'free_egress=\"-\" access_type=\"Use Cloud Data Delivery\" '\n",
      " 'org=\"AWS\"/></SRAFile><SRAFile cluster=\"public\" '\n",
      " 'filename=\"220857C_LM_3_S1_L002_R1_001.fastq.gz\" size=\"10102292637\" '\n",
      " 'date=\"2024-11-06 09:19:47\" md5=\"98be9613a5351bbea670fd011fbbd619\" '\n",
      " 'version=\"1\" semantic_name=\"fastq\" supertype=\"Original\" '\n",
      " 'sratoolkit=\"0\"><Alternatives '\n",
      " 'url=\"s3://sra-pub-src-16/SRR31243301/220857C_LM_3_S1_L002_R1_001.fastq.gz.1\" '\n",
      " 'free_egress=\"-\" access_type=\"Use Cloud Data Delivery\" '\n",
      " 'org=\"AWS\"/></SRAFile><SRAFile cluster=\"public\" '\n",
      " 'filename=\"220857C_LM_3_S1_L002_R2_001.fastq.gz\" size=\"8074476463\" '\n",
      " 'date=\"2024-11-06 09:20:44\" md5=\"8510f6d25259e32f6bc6cb647e7f0bdd\" '\n",
      " 'version=\"1\" semantic_name=\"fastq\" supertype=\"Original\" '\n",
      " 'sratoolkit=\"0\"><Alternatives '\n",
      " 'url=\"s3://sra-pub-src-16/SRR31243301/220857C_LM_3_S1_L002_R2_001.fastq.gz.1\" '\n",
      " 'free_egress=\"-\" access_type=\"Use Cloud Data Delivery\" '\n",
      " 'org=\"AWS\"/></SRAFile><SRAFile cluster=\"public\" filename=\"SRR31243301\" '\n",
      " 'url=\"https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR31243301/SRR31243301\" '\n",
      " 'size=\"26895280219\" date=\"2024-11-06 09:24:41\" '\n",
      " 'md5=\"dd8019b9d9b7aacfef71e36c8e37af67\" version=\"1\" semantic_name=\"SRA '\n",
      " 'Normalized\" supertype=\"Primary ETL\" sratoolkit=\"1\"><Alternatives '\n",
      " 'url=\"https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR31243301/SRR31243301\" '\n",
      " 'free_egress=\"worldwide\" access_type=\"anonymous\" '\n",
      " 'org=\"AWS\"/></SRAFile><SRAFile cluster=\"public\" filename=\"SRR31243301.lite\" '\n",
      " 'url=\"https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos2/sra-pub-zq-20/SRR031/31243/SRR31243301/SRR31243301.lite.1\" '\n",
      " 'size=\"17461803811\" date=\"2024-11-06 11:05:00\" '\n",
      " 'md5=\"7220b0597bf418fb01a1ceebaca73eab\" version=\"1\" semantic_name=\"SRA Lite\" '\n",
      " 'supertype=\"Primary ETL\" sratoolkit=\"1\"><Alternatives '\n",
      " 'url=\"https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos2/sra-pub-zq-20/SRR031/31243/SRR31243301/SRR31243301.lite.1\" '\n",
      " 'free_egress=\"worldwide\" access_type=\"anonymous\" org=\"NCBI\"/><Alternatives '\n",
      " 'url=\"s3://sra-pub-zq-3/SRR31243301/SRR31243301.lite.1\" '\n",
      " 'free_egress=\"s3.us-east-1\" access_type=\"aws identity\" '\n",
      " 'org=\"AWS\"/><Alternatives '\n",
      " 'url=\"gs://sra-pub-zq-107/SRR31243301/SRR31243301.lite.1\" '\n",
      " 'free_egress=\"gs.us-east1\" access_type=\"gcp identity\" '\n",
      " 'org=\"GCP\"/></SRAFile></SRAFiles><CloudFiles><CloudFile filetype=\"fastq\" '\n",
      " 'provider=\"s3\" location=\"s3.us-east-1\"/><CloudFile filetype=\"run\" '\n",
      " 'provider=\"s3\" location=\"s3.us-east-1\"/><CloudFile filetype=\"run.zq\" '\n",
      " 'provider=\"gs\" location=\"gs.us-east1\"/><CloudFile filetype=\"run.zq\" '\n",
      " 'provider=\"s3\" location=\"s3.us-east-1\"/></CloudFiles><Statistics nreads=\"2\" '\n",
      " 'nspots=\"227438532\"><Read index=\"0\" count=\"227438532\" average=\"150\" '\n",
      " 'stdev=\"0\"/><Read index=\"1\" count=\"227438532\" average=\"150\" '\n",
      " 'stdev=\"0\"/></Statistics><Databases><Database><Table '\n",
      " 'name=\"SEQUENCE\"><Statistics source=\"meta\"><Rows count=\"227438532\"/><Elements '\n",
      " 'count=\"68231559600\"/></Statistics></Table></Database></Databases><Bases '\n",
      " 'cs_native=\"false\" count=\"68231559600\"><Base value=\"A\" '\n",
      " 'count=\"19648710546\"/><Base value=\"C\" count=\"14680002479\"/><Base value=\"G\" '\n",
      " 'count=\"13741048613\"/><Base value=\"T\" count=\"20160768783\"/><Base value=\"N\" '\n",
      " 'count=\"1029179\"/></Bases></RUN></RUN_SET></EXPERIMENT_PACKAGE></EXPERIMENT_PACKAGE_SET>\\n')\n"
     ]
    }
   ],
   "source": [
    "@tool \n",
    "def efetch(\n",
    "    database: Annotated[str, \"Database name ('sra' or 'gds')\"],\n",
    "    dataset_id: Annotated[str, \"Entrez ID\"],\n",
    "    )-> Annotated[List[str], \"eFetch results in xml format\"]:\n",
    "    \"\"\"\n",
    "    Run an Entrez efetch query and return the results.\n",
    "    \"\"\"\n",
    "\n",
    "    time.sleep(0.5)\n",
    "\n",
    "    # Fetch dataset record\n",
    "    handle = Entrez.efetch(db=database, id=dataset_id, retmode=\"xml\")\n",
    "    record = handle.read()\n",
    "    handle.close()\n",
    "\n",
    "    try:\n",
    "        record = record.decode(\"utf-8\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    return str(record)\n",
    "\n",
    "record = efetch.invoke({\"database\" : \"sra\", \"dataset_id\" : \"35966237\"})\n",
    "pprint(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SRR29263068,SRR29263069'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@tool \n",
    "def geo2sra(\n",
    "    dataset_id: Annotated[str, \"GEO accession\"],\n",
    "    )-> Annotated[List[str], \"SRA accession\"]:\n",
    "    \"\"\"\n",
    "    Convert GEO accession to SRA accession.\n",
    "    \"\"\"\n",
    "    # Fetch detailed GEO record to get links to SRA\n",
    "    handle = Entrez.elink(dbfrom=\"gds\", db=\"sra\", id=dataset_id)\n",
    "    links = Entrez.read(handle)\n",
    "    handle.close()\n",
    "        \n",
    "    sra_ids = []\n",
    "    if links[0]['LinkSetDb']:\n",
    "        sra_ids = [link['Id'] for link in links[0]['LinkSetDb'][0]['Link']]\n",
    "\n",
    "    # Get SRA accessions from IDs\n",
    "    sra_accessions = []\n",
    "    for sra_id in sra_ids:\n",
    "        handle = Entrez.esummary(db=\"sra\", id=sra_id)\n",
    "        summary = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        # Extract SRA accessions from summary\n",
    "        for entry in summary:\n",
    "            try:\n",
    "                run_xml = entry[\"Runs\"]\n",
    "            except KeyError:\n",
    "                continue\n",
    "            \n",
    "            hit = re.search(r'Run acc=\"([A-Z]+\\d+)\"', run_xml)\n",
    "            if hit:\n",
    "                sra_accessions.append(hit.group(1))\n",
    "        # Add delay to comply with NCBI's guidelines\n",
    "        time.sleep(0.34) \n",
    "    if len(sra_accessions) == 0:\n",
    "        return \"No SRA accessions found.\"\n",
    "\n",
    "    # return accessions as string\n",
    "    return ','.join(sra_accessions)\n",
    "\n",
    "#geo2sra.invoke({\"dataset_id\" : \"200254051\"})\n",
    "geo2sra.invoke({\"dataset_id\" : \"200268899\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<?xml version=\"1.0\" ?>\\n<!DOCTYPE PubmedArticleSet PUBLIC \"-//NLM//DTD PubMedArticle, 1st January 2024//EN\" \"https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_240101.dtd\">\\n<PubmedArticleSet>\\n<PubmedArticle><MedlineCitation Status=\"Publisher\" Owner=\"NLM\"><PMID Version=\"1\">39492543</PMID><DateRevised><Year>2024</Year><Month>11</Month><Day>04</Day></DateRevised><Article PubModel=\"Print-Electronic\"><Journal><ISSN IssnType=\"Electronic\">1096-0007</ISSN><JournalIssue CitedMedium=\"Internet\"><PubDate><Year>2023</Year><Month>Oct</Month><Day>28</Day></PubDate></JournalIssue><Title>Experimental eye research</Title><ISOAbbreviation>Exp Eye Res</ISOAbbreviation></Journal><ArticleTitle>Comprehensive landscape of RNA N6-methyladenosine modification in lens epithelial cells from normal and diabetic cataract.</ArticleTitle><Pagination><StartPage>109702</StartPage><MedlinePgn>109702</MedlinePgn></Pagination><ELocationID EIdType=\"doi\" ValidYN=\"Y\">10.1016/j.exer.2023.109702</ELocationID><ELocationID EIdType=\"pii\" ValidYN=\"Y\">S0014-4835(23)00323-8</ELocationID><Abstract><AbstractText>To gain more insight into the mechanism of cataract formation from the perspective of epigenetics in the diabetic population, lens epithelium from diabetic cataract patients and health individuals were collected separately and analyzed for N6-methyladenosine (m<sup>6</sup>A)-modified RNA using methylated RNA immunoprecipitation sequencing (MeRIP-Seq). Subsequently, differential expression analysis was performed on m<sup>6</sup>A-regulated messenger RNA (mRNA), circular RNA (circRNA), and long non-coding RNA (lncRNA), followed by functional annotation using the Gene Ontology (GO) database. Furthermore, analysis of single-cell data of lens complemented the intrinsic association and cellular heterogeneity of cataract and m<sup>6</sup>A regulators. In this study, both the global expression levels and peak intensity of m<sup>6</sup>A-tagged RNAs were increased in patients with diabetic cataract. And we noted multiple core enzymes were upregulated in the diabetic cataract (DC) samples. Besides, single-cell RNA sequencing analysis of the lens revealed the heterogeneous expression of RNA m<sup>6</sup>A regulators across different cell types, and we noted that the early fiber cell cluster was also closely associated with the onset of cataract and m<sup>6</sup>A modification. The results comprehensively revealed the dynamic modification landscape of m<sup>6</sup>A on mRNA, circRNA, and lncRNA, which might provide valuable resources for future studies of the pathogenesis of DCs.</AbstractText><CopyrightInformation>Copyright &#xa9; 2023. Published by Elsevier Ltd.</CopyrightInformation></Abstract><AuthorList CompleteYN=\"Y\"><Author ValidYN=\"Y\"><LastName>Zhang</LastName><ForeName>Xiaowen</ForeName><Initials>X</Initials><AffiliationInfo><Affiliation>Eye Institute of Shandong First Medical University, Qingdao Eye Hospital of Shandong First Medical University, Qingdao, China; State Key Laboratory Cultivation Base, Shandong Provincial Key Laboratory of Ophthalmology, Qingdao, China; School of Ophthalmology, Shandong First Medical University, Jinan, China.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Dou</LastName><ForeName>Shengqian</ForeName><Initials>S</Initials><AffiliationInfo><Affiliation>Eye Institute of Shandong First Medical University, Qingdao Eye Hospital of Shandong First Medical University, Qingdao, China; State Key Laboratory Cultivation Base, Shandong Provincial Key Laboratory of Ophthalmology, Qingdao, China. Electronic address: doushq@126.com.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Huang</LastName><ForeName>Yusen</ForeName><Initials>Y</Initials><AffiliationInfo><Affiliation>Eye Institute of Shandong First Medical University, Qingdao Eye Hospital of Shandong First Medical University, Qingdao, China; State Key Laboratory Cultivation Base, Shandong Provincial Key Laboratory of Ophthalmology, Qingdao, China. Electronic address: huang_yusen@126.com.</Affiliation></AffiliationInfo></Author></AuthorList><Language>eng</Language><PublicationTypeList><PublicationType UI=\"D016428\">Journal Article</PublicationType></PublicationTypeList><ArticleDate DateType=\"Electronic\"><Year>2023</Year><Month>10</Month><Day>28</Day></ArticleDate></Article><MedlineJournalInfo><Country>England</Country><MedlineTA>Exp Eye Res</MedlineTA><NlmUniqueID>0370707</NlmUniqueID><ISSNLinking>0014-4835</ISSNLinking></MedlineJournalInfo><CitationSubset>IM</CitationSubset><KeywordList Owner=\"NOTNLM\"><Keyword MajorTopicYN=\"N\">Circular RNA</Keyword><Keyword MajorTopicYN=\"N\">Diabetic cataract</Keyword><Keyword MajorTopicYN=\"N\">Long non-coding RNA</Keyword><Keyword MajorTopicYN=\"N\">N6-methyladenosine modification</Keyword><Keyword MajorTopicYN=\"N\">mRNA</Keyword></KeywordList><CoiStatement>Declaration of competing interest The authors have declared that no competing interest exists.</CoiStatement></MedlineCitation><PubmedData><History><PubMedPubDate PubStatus=\"received\"><Year>2023</Year><Month>9</Month><Day>1</Day></PubMedPubDate><PubMedPubDate PubStatus=\"revised\"><Year>2023</Year><Month>10</Month><Day>22</Day></PubMedPubDate><PubMedPubDate PubStatus=\"accepted\"><Year>2023</Year><Month>10</Month><Day>27</Day></PubMedPubDate><PubMedPubDate PubStatus=\"medline\"><Year>2024</Year><Month>11</Month><Day>4</Day><Hour>6</Hour><Minute>24</Minute></PubMedPubDate><PubMedPubDate PubStatus=\"pubmed\"><Year>2024</Year><Month>11</Month><Day>4</Day><Hour>6</Hour><Minute>24</Minute></PubMedPubDate><PubMedPubDate PubStatus=\"entrez\"><Year>2024</Year><Month>11</Month><Day>4</Day><Hour>2</Hour><Minute>54</Minute></PubMedPubDate></History><PublicationStatus>aheadofprint</PublicationStatus><ArticleIdList><ArticleId IdType=\"pubmed\">39492543</ArticleId><ArticleId IdType=\"doi\">10.1016/j.exer.2023.109702</ArticleId><ArticleId IdType=\"pii\">S0014-4835(23)00323-8</ArticleId></ArticleIdList></PubmedData></PubmedArticle></PubmedArticleSet>'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@tool \n",
    "def get_pubmed_article(\n",
    "    pubmed_id: Annotated[str, \"PubMed ID\"],\n",
    "    )-> Annotated[str, \"PubMed article metadata in xml format\"]:\n",
    "    \"\"\"\n",
    "    Get a PubMed article.\n",
    "    \"\"\"\n",
    "    handle = Entrez.efetch(db=\"pubmed\", id=pubmed_id, retmode=\"xml\")\n",
    "    record = handle.read()\n",
    "    handle.close()\n",
    "    try:\n",
    "        record = record.decode(\"utf-8\")\n",
    "    except:\n",
    "        pass\n",
    "    return str(record)\n",
    "\n",
    "get_pubmed_article.invoke({\"pubmed_id\" : \"39492543\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subprocess import Popen, PIPE\n",
    "\n",
    "def run_cmd(cmd: str) -> tuple:\n",
    "    \"\"\"\n",
    "    Run sub-command and return returncode, output, and error.\n",
    "    Args:\n",
    "        cmd: Command to run\n",
    "    Returns:\n",
    "        tuple: (returncode, output, error)\n",
    "    \"\"\"\n",
    "    p = Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True)\n",
    "    output, err = p.communicate()\n",
    "    return p.returncode, output, err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "@tool\n",
    "def run_sra_stat(\n",
    "    accession: Annotated[str, \"SRA accession\"],\n",
    "    tries: Annotated[int, \"Number of attempts\"]=3\n",
    "    ) -> str: \n",
    "    \"\"\"\n",
    "    Run sra-stat on an SRA accession.\n",
    "    \"\"\"\n",
    "    cmd = f'sra-stat --xml --quick {accession}'\n",
    "    err = \"\"\n",
    "    for i in range(tries):\n",
    "        rc,output,err = run_cmd(cmd)\n",
    "        if rc == 0:\n",
    "            return output.decode(\"utf-8\")\n",
    "        # sleep prior to next attempt\n",
    "        sleep_time = 10 * (i + 1)\n",
    "        time.sleep(sleep_time)\n",
    "    return f\"Failed to run sra-stat: {err}\"\n",
    "\n",
    "#run_sra_stat.invoke({\"accession\" : \"SRR13112659\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
