{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Check on the test and/or production databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pypika import Query, Table, Field, Column, Criterion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.upsert import db_upsert\n",
    "from SRAgent.db.utils import db_list_tables, db_glimpse_tables, db_get_table, execute_query\n",
    "from SRAgent.db.get import db_find_srx\n",
    "from SRAgent.db.create import create_table, create_table_router"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eval\n",
      "screcounter_log\n",
      "srx_srr\n",
      "srx_metadata\n",
      "screcounter_star_params\n",
      "screcounter_star_results\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in prod\n",
    "os.environ['DYNACONF'] = 'prod'\n",
    "conn_prod = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_prod)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['eval',\n",
       " 'screcounter_log',\n",
       " 'srx_srr',\n",
       " 'srx_metadata',\n",
       " 'screcounter_star_params',\n",
       " 'screcounter_star_results']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_list_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX25926423\tSRR30502613\t2025-01-03 18:43:38.627686\t2025-01-03 18:43:38.627686\n",
      "SRX25926423\tSRR30502600\t2025-01-03 18:43:38.627686\t2025-01-03 18:43:38.627686\n",
      "SRX24914484\tSRR29400843\t2025-01-03 18:43:49.275138\t2025-01-03 18:43:49.275138\n",
      "SRX22006379\tSRR26297925\t2025-01-03 18:43:57.864762\t2025-01-03 18:43:57.864762\n",
      "SRX21883929\tSRR26171850\t2025-01-03 18:44:08.340937\t2025-01-03 18:44:08.340937\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t29110064\tERX11148781\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t73f82ac8-15cc-4fcd-87f8-5683723fce7f\tDevelopmental cell programs are co-opted in inflammatory skin disease\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t19007785\tSRX13670569\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t91c8e321-566f-4f9d-b89e-3a164be654d5\tNeuron type-specific effects of human aging and sex on DNA methylation and transcription\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t12488012\tSRX9556597\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\ted9185e3-5b82-40c7-9824-b2141590c7f0\tTime-resolved Systems Immunology Reveals a Late Juncture Linked to Fatal COVID-19\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t21270435\tERX8791959\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t62ef75e4-cbea-454e-a0ce-998ec40223d3\tCross-tissue immune cell analysis reveals tissue-specific features in humans\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t34493113\tERX12558784\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\tec329aed-22bc-4d6e-8935-8282dcb1acac\tEarly human lung immune cell development and its role in epithelial cell fate\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "\n",
      "#-- Table: screcounter_star_params --#\n",
      "sample\tbarcodes\tstar_index\tcell_barcode_length\tumi_length\tstrand\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_star_results --#\n",
      "sample\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>26330964</td>\n",
       "      <td>SRX19136133</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Apical Ectodermal Ridge (AER)</td>\n",
       "      <td>Split-Hand/Foot Malformation type 3 (SHFM3)</td>\n",
       "      <td>Investigated functional consequences of SHFM3-...</td>\n",
       "      <td>other</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:00.368512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>27694563</td>\n",
       "      <td>SRX20273630</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>not specified</td>\n",
       "      <td>infected</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:44:17.048627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>29777569</td>\n",
       "      <td>SRX21883929</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>lymph node</td>\n",
       "      <td>ovarian tumors, immune checkpoint blockade sen...</td>\n",
       "      <td>2-Deoxy-D-glucose (2DG)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:44:08.060710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>29919592</td>\n",
       "      <td>SRX22006379</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blastocyst</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>inner cell mass (ICM)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:43:57.602382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>30865352</td>\n",
       "      <td>SRX22823407</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>tumor</td>\n",
       "      <td>phaeohyphomycosis</td>\n",
       "      <td>anti-PD1 treatment</td>\n",
       "      <td>endogenous CD8+ T cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:48.003195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249222</td>\n",
       "      <td>SRX24914484</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>hepatitis C virus (HCV)</td>\n",
       "      <td>E2-specific B cells sorted from PBMCs</td>\n",
       "      <td>memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:43:48.934750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sra</td>\n",
       "      <td>34602788</td>\n",
       "      <td>SRX25602002</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>Type 2 Diabetes Mellitus</td>\n",
       "      <td>Randomized controlled study on salt intake eff...</td>\n",
       "      <td>other</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:27.430411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sra</td>\n",
       "      <td>34648698</td>\n",
       "      <td>SRX25634198</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>prostate</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>anti-PD1 + Vehicle</td>\n",
       "      <td>Myc-CaP</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:44:30.489428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sra</td>\n",
       "      <td>34996766</td>\n",
       "      <td>SRX25926423</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>graft vs. host disease (GVHD)</td>\n",
       "      <td>STAT1 deletion in murine intestinal epithelium</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:43:38.319328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sra</td>\n",
       "      <td>36086856</td>\n",
       "      <td>SRX26710454</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Duodenal Lymph Nodes)</td>\n",
       "      <td>none</td>\n",
       "      <td>HP (Helminth Infection)</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:29.642017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  database  entrez_id srx_accession is_illumina is_single_cell is_paired_end  \\\n",
       "0      sra   26330964   SRX19136133         yes            yes           yes   \n",
       "1      sra   27694563   SRX20273630         yes            yes           yes   \n",
       "2      sra   29777569   SRX21883929         yes            yes           yes   \n",
       "3      sra   29919592   SRX22006379         yes            yes           yes   \n",
       "4      sra   30865352   SRX22823407         yes            yes           yes   \n",
       "5      sra   33249222   SRX24914484         yes            yes           yes   \n",
       "6      sra   34602788   SRX25602002         yes            yes           yes   \n",
       "7      sra   34648698   SRX25634198         yes            yes           yes   \n",
       "8      sra   34996766   SRX25926423         yes            yes           yes   \n",
       "9      sra   36086856   SRX26710454         yes            yes           yes   \n",
       "\n",
       "       lib_prep        tech_10x    cell_prep organism  \\\n",
       "0  10x_Genomics     3_prime_gex  single_cell    mouse   \n",
       "1         other  not_applicable  single_cell    human   \n",
       "2  10x_Genomics     3_prime_gex  single_cell    mouse   \n",
       "3    Smart-seq2  not_applicable  single_cell    human   \n",
       "4         other  not_applicable  single_cell    mouse   \n",
       "5    Smart-seq2  not_applicable  single_cell    human   \n",
       "6         other  not_applicable  single_cell    human   \n",
       "7         other  not_applicable  single_cell    mouse   \n",
       "8  10x_Genomics     3_prime_gex  single_cell    mouse   \n",
       "9    Smart-seq2           other  single_cell    mouse   \n",
       "\n",
       "                          tissue  \\\n",
       "0  Apical Ectodermal Ridge (AER)   \n",
       "1                         HCT116   \n",
       "2                     lymph node   \n",
       "3                     blastocyst   \n",
       "4                          tumor   \n",
       "5                          blood   \n",
       "6                          PBMCs   \n",
       "7                       prostate   \n",
       "8                small intestine   \n",
       "9    D-LN (Duodenal Lymph Nodes)   \n",
       "\n",
       "                                             disease  \\\n",
       "0        Split-Hand/Foot Malformation type 3 (SHFM3)   \n",
       "1                                      not specified   \n",
       "2  ovarian tumors, immune checkpoint blockade sen...   \n",
       "3                                             unsure   \n",
       "4                                  phaeohyphomycosis   \n",
       "5                            hepatitis C virus (HCV)   \n",
       "6                           Type 2 Diabetes Mellitus   \n",
       "7                                    prostate cancer   \n",
       "8                      graft vs. host disease (GVHD)   \n",
       "9                                               none   \n",
       "\n",
       "                                        purturbation                cell_line  \\\n",
       "0  Investigated functional consequences of SHFM3-...                    other   \n",
       "1                                           infected                   HCT116   \n",
       "2                            2-Deoxy-D-glucose (2DG)                   unsure   \n",
       "3                                             unsure    inner cell mass (ICM)   \n",
       "4                                 anti-PD1 treatment  endogenous CD8+ T cells   \n",
       "5              E2-specific B cells sorted from PBMCs           memory B cells   \n",
       "6  Randomized controlled study on salt intake eff...                    other   \n",
       "7                                 anti-PD1 + Vehicle                  Myc-CaP   \n",
       "8     STAT1 deletion in murine intestinal epithelium                   unsure   \n",
       "9                            HP (Helminth Infection)                     none   \n",
       "\n",
       "  czi_collection_id czi_collection_name                         notes  \\\n",
       "0              None                None  Metadata obtained by SRAgent   \n",
       "1              None                None  Metadata obtained by SRAgent   \n",
       "2              None                None  Metadata obtained by SRAgent   \n",
       "3              None                None  Metadata obtained by SRAgent   \n",
       "4              None                None  Metadata obtained by SRAgent   \n",
       "5              None                None  Metadata obtained by SRAgent   \n",
       "6              None                None  Metadata obtained by SRAgent   \n",
       "7              None                None  Metadata obtained by SRAgent   \n",
       "8              None                None  Metadata obtained by SRAgent   \n",
       "9              None                None  Metadata obtained by SRAgent   \n",
       "\n",
       "                  created_at                 updated_at  \n",
       "0 2025-01-03 18:42:38.083215 2025-01-03 18:45:00.368512  \n",
       "1 2025-01-03 18:42:38.083215 2025-01-03 18:44:17.048627  \n",
       "2 2025-01-03 18:42:38.083215 2025-01-03 18:44:08.060710  \n",
       "3 2025-01-03 18:42:38.083215 2025-01-03 18:43:57.602382  \n",
       "4 2025-01-03 18:42:38.083215 2025-01-03 18:45:48.003195  \n",
       "5 2025-01-03 18:42:38.083215 2025-01-03 18:43:48.934750  \n",
       "6 2025-01-03 18:42:38.083215 2025-01-03 18:45:27.430411  \n",
       "7 2025-01-03 18:42:38.083215 2025-01-03 18:44:30.489428  \n",
       "8 2025-01-03 18:42:38.083215 2025-01-03 18:43:38.319328  \n",
       "9 2025-01-03 18:42:38.083215 2025-01-03 18:45:29.642017  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \\\n",
    "    .distinct() \\\n",
    "    .where(tbl.notes != \"Processed by Chris Carpenter\")\n",
    "srx_metadata = pd.read_sql(str(stmt), conn_prod)\n",
    "srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
