{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Check on the test and/or production databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv(override=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pypika import Query, Table, Field, Column, Criterion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.upsert import db_upsert\n",
    "from SRAgent.db.utils import db_list_tables, db_glimpse_tables, db_get_table, execute_query\n",
    "from SRAgent.db.get import db_find_srx\n",
    "from SRAgent.db.create import create_table, create_table_router"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "screcounter_star_results\n",
      "eval\n",
      "screcounter_trace\n",
      "srx_srr\n",
      "srx_metadata\n",
      "screcounter_log\n",
      "screcounter_star_params\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in prod\n",
    "os.environ['DYNACONF'] = 'prod'\n",
    "conn_prod = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_prod)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['screcounter_star_results',\n",
       " 'eval',\n",
       " 'screcounter_trace',\n",
       " 'srx_srr',\n",
       " 'srx_metadata',\n",
       " 'screcounter_log',\n",
       " 'screcounter_star_params']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_list_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: screcounter_star_results --#\n",
      "sample\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\tmedian_reads_per_cell\tq30_bases_in_cb_umi\tq30_bases_in_rna_read\treads_mapped_to_gene__unique_gene\treads_mapped_to_gene__unique_multiple_gene\treads_mapped_to_genefull__unique_genefull\treads_mapped_to_genefull__unique_multiple_genefull\treads_mapped_to_genefull_ex50pas__unique_genefull_ex50pas\treads_mapped_to_genefull_ex50pas__unique_multiple_genefull_ex50\treads_mapped_to_genefull_exonoverintron__unique_genefull_exonov\treads_mapped_to_genefull_exonoverintron__unique_multiple_genefu\treads_mapped_to_genome__unique\treads_mapped_to_genome__unique_multiple\treads_mapped_to_velocyto__unique_velocyto\treads_mapped_to_velocyto__unique_multiple_velocyto\treads_with_valid_barcodes\tsequencing_saturation\ttotal_feature_detected\tumis_in_cells\tunique_reads_in_cells_mapped_to_gene\tunique_reads_in_cells_mapped_to_genefull\tunique_reads_in_cells_mapped_to_genefull_ex50pas\tunique_reads_in_cells_mapped_to_genefull_exonoverintron\tcreated_at\tupdated_at\n",
      "ERX10024831\tgene_full\t2292\t0.911745\t\t12557.0\t2730.0\t\t10513.0\t2775.0\t146018831\t25399.0\t0.968155\t0.918635\t\t\t0.534665\t0.788618\t\t\t\t\t0.734504\t0.87767\t\t\t0.983305\t0.593299\t21731.0\t28780763\t\t71181072.0\t\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tgene_ex50\t2290\t0.912572\t\t17603.0\t2756.0\t\t11784.0\t2799.0\t146018831\t28689.0\t0.968155\t0.918635\t\t\t\t\t0.762455\t\t\t\t0.734504\t0.87767\t\t\t0.983334\t0.601129\t21911.0\t40311609\t\t\t101599277.0\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tgene_ex_int\t2290\t0.912561\t\t17625.0\t2767.0\t\t11803.0\t2809.0\t146018831\t28735.0\t0.968155\t0.918635\t\t\t\t\t\t\t\t\t0.734504\t0.87767\t\t\t0.983322\t0.601073\t22001.0\t40362108\t\t\t\t101712901.0\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tvelocyto\t0\t\t\t\t\t\t\t\t146018831\t\t0.968155\t0.918635\t\t\t\t\t\t\t\t\t0.734504\t0.87767\t0.0\t0.0\t0.984647\t-inf\t\t0\t\t\t\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10086874\tgene_full\t2262\t0.909667\t\t4977.0\t2018.0\t\t4615.0\t2116.0\t160000000\t43880.0\t0.952429\t0.920767\t\t\t0.745753\t0.801599\t\t\t\t\t0.800461\t0.924752\t\t\t0.967224\t0.887163\t21453.0\t11258108\t\t108541887.0\t\t\t2025-01-13 15:17:54.767872\t2025-01-14 20:06:25.557352\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "\n",
      "#-- Table: screcounter_trace --#\n",
      "task_id\thash\tnative_id\tname\tstatus\texit\tsubmit\tcontainer\tcpus\ttime\tdisk\tmemory\tattempt\tduration\trealtime\tcpu_percent\tpeak_rss\tpeak_vmem\trchar\twchar\tworkdir\tscratch\tcreated_at\tupdated_at\n",
      "1\t56/d8d487\tnf-56d8d487-1736775758365/0/nf-56d8d487-173677-6a5e0c92-ed1e-4c820\tDB_ACC_WF:GET_DB_ACCESSIONS\tCOMPLETED\t0\t2025-01-13 05:42:40.879\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 53s\t1.1s\t66.4%\t3 MB\t4.2 MB\t20.1 MB\t91.6 KB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/56/d8d4879ac276a9d30ec850f1ecf20a\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "2\te8/3237c0\tnf-e83237c0-1736775874725/0/nf-e83237c0-173677-e84d9f6b-20f7-48380\tSRA_STAT (1)\tCOMPLETED\t0\t2025-01-13 05:44:35.213\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t40.2%\t79.2 MB\t543.5 MB\t14.8 MB\t1.8 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/e8/3237c03758be58eca921e818fa3e0e\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "3\t64/3df771\tnf-643df771-1736775874726/0/nf-643df771-173677-54f5b596-489b-43870\tSRA_STAT (2)\tCOMPLETED\t0\t2025-01-13 05:44:35.674\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t37.7%\t79.5 MB\t543.5 MB\t14.8 MB\t755 KB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/64/3df7716232b892d0d8a7fa72ece0b8\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "4\t4e/793c08\tnf-4e793c08-1736775874789/0/nf-4e793c08-173677-239dfe74-6cc5-41130\tSRA_STAT (3)\tCOMPLETED\t0\t2025-01-13 05:44:36.077\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t39.1%\t79.4 MB\t543.5 MB\t14.8 MB\t1.7 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/4e/793c0898e1f5b974d97dca646c8fa2\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "5\tce/e8e21e\tnf-cee8e21e-1736775874795/0/nf-cee8e21e-173677-f6013cb6-ca08-40df0\tSRA_STAT (4)\tCOMPLETED\t0\t2025-01-13 05:44:36.533\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 38s\t1.3s\t39.7%\t79.1 MB\t543.5 MB\t14.8 MB\t1.7 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/ce/e8e21e2d91358d48f9b67118002b20\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX24914484\tSRR29400843\t2025-01-03 18:43:49.275138\t2025-01-03 18:43:49.275138\n",
      "SRX22006379\tSRR26297925\t2025-01-03 18:43:57.864762\t2025-01-03 18:43:57.864762\n",
      "SRX20273630\tSRR24488012\t2025-01-03 18:44:17.303662\t2025-01-03 18:44:17.303662\n",
      "SRX25634198\tSRR30167498\t2025-01-03 18:44:30.759838\t2025-01-03 18:44:30.759838\n",
      "SRX25602002\tSRR30132651\t2025-01-03 18:45:27.695421\t2025-01-03 18:45:27.695421\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t29110064\tERX11148781\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t73f82ac8-15cc-4fcd-87f8-5683723fce7f\tDevelopmental cell programs are co-opted in inflammatory skin disease\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t19007785\tSRX13670569\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t91c8e321-566f-4f9d-b89e-3a164be654d5\tNeuron type-specific effects of human aging and sex on DNA methylation and transcription\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t12488012\tSRX9556597\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\ted9185e3-5b82-40c7-9824-b2141590c7f0\tTime-resolved Systems Immunology Reveals a Late Juncture Linked to Fatal COVID-19\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t21270435\tERX8791959\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t62ef75e4-cbea-454e-a0ce-998ec40223d3\tCross-tissue immune cell analysis reveals tissue-specific features in humans\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t34493113\tERX12558784\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\tec329aed-22bc-4d6e-8935-8282dcb1acac\tEarly human lung immune cell development and its role in epithelial cell fate\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "\n",
      "#-- Table: screcounter_star_params --#\n",
      "sample\tbarcodes\tstar_index\tcell_barcode_length\tumi_length\tstrand\tcreated_at\tupdated_at\n",
      "ERX10086874\t737K-august-2016.txt\tstar_refData_2020_mm10\t16\t10\tForward\t2025-01-13 13:58:53.927797\t2025-01-13 13:58:53.927797\n",
      "ERX10396585\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:56.112918\t2025-01-13 13:58:56.112918\n",
      "ERX10396584\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:58.510780\t2025-01-13 13:58:58.510780\n",
      "ERX10690512\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:58.640112\t2025-01-13 13:58:58.640112\n",
      "ERX10855739\t737K-august-2016.txt\tstar_refData_2020_hg38\t16\t10\tForward\t2025-01-13 13:58:59.433825\t2025-01-13 13:58:59.433825\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Database updates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CZI datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Organism "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>26659536</td>\n",
       "      <td>SRX19392127</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>S1 cortex</td>\n",
       "      <td>neuropsychiatric outcomes</td>\n",
       "      <td>maternal immune activation</td>\n",
       "      <td>microglia, P14, male, C57Bl/6N, WT</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:19:05.896057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>27454961</td>\n",
       "      <td>ERX10668462</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ILC progenitors from femur</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:19:02.038324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>27454971</td>\n",
       "      <td>ERX10668472</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ILC progenitors from femur</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>ILC2 cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:38:48.116573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>27455317</td>\n",
       "      <td>ERX10668818</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>femur</td>\n",
       "      <td>unsure</td>\n",
       "      <td>Bcl11b regulatory networks</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:30:22.045996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>27694563</td>\n",
       "      <td>SRX20273630</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>not specified</td>\n",
       "      <td>infected</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:44:17.048627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sra</td>\n",
       "      <td>29919592</td>\n",
       "      <td>SRX22006379</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blastocyst</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>inner cell mass (ICM)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:43:57.602382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sra</td>\n",
       "      <td>30864469</td>\n",
       "      <td>SRX22822524</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>tumor</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>OT1 CD8+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:18:04.534748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sra</td>\n",
       "      <td>30865352</td>\n",
       "      <td>SRX22823407</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>tumor</td>\n",
       "      <td>phaeohyphomycosis</td>\n",
       "      <td>anti-PD1 treatment</td>\n",
       "      <td>endogenous CD8+ T cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:48.003195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sra</td>\n",
       "      <td>30865785</td>\n",
       "      <td>SRX22823840</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>tumor</td>\n",
       "      <td>cancer immunotherapy</td>\n",
       "      <td>none specified</td>\n",
       "      <td>B16-OVA, MC38, LLC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:18:33.867744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sra</td>\n",
       "      <td>32469529</td>\n",
       "      <td>SRX24158623</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>glaucoma</td>\n",
       "      <td>primary hTM cells isolated from human donor co...</td>\n",
       "      <td>healthy</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:40:04.684759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sra</td>\n",
       "      <td>32702534</td>\n",
       "      <td>SRX24387553</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>bone marrow 90d post treatment (T20)</td>\n",
       "      <td>multiple myeloma, early relapse</td>\n",
       "      <td>BCMA CAR-T cell therapy</td>\n",
       "      <td>T-cells (from Myeloma patient 07)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:31:18.187839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sra</td>\n",
       "      <td>32719383</td>\n",
       "      <td>SRX24404397</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>multiple myeloma</td>\n",
       "      <td>BCMA CAR-T cell therapy</td>\n",
       "      <td>myeloid cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:38:18.491089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sra</td>\n",
       "      <td>32777790</td>\n",
       "      <td>SRX24461697</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>whole skin</td>\n",
       "      <td>not specified</td>\n",
       "      <td>tamoxifen</td>\n",
       "      <td>K14CreERTM/LSL-Sox2-IRES-GFP</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:38:39.671695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>sra</td>\n",
       "      <td>32927474</td>\n",
       "      <td>SRX24608712</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>embryo</td>\n",
       "      <td>none</td>\n",
       "      <td>Bmp4 KO Heterozygous (C57BL/6x129)</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:31:29.665074</td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sra</td>\n",
       "      <td>32927520</td>\n",
       "      <td>SRX24608758</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>mouse embryo</td>\n",
       "      <td>none</td>\n",
       "      <td>BMP4 signaling</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:17:18.538773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249216</td>\n",
       "      <td>SRX24914478</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>other</td>\n",
       "      <td>exercised</td>\n",
       "      <td>memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:17:29.575513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249222</td>\n",
       "      <td>SRX24914484</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>hepatitis C virus (HCV)</td>\n",
       "      <td>E2-specific B cells sorted from PBMCs</td>\n",
       "      <td>memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:43:48.934750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249325</td>\n",
       "      <td>SRX24914587</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:29:29.430436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249385</td>\n",
       "      <td>SRX24914647</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>hepatitis C virus (HCV) reinfection</td>\n",
       "      <td>unsure</td>\n",
       "      <td>Memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:40:07.232871</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40846 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession is_illumina is_single_cell  \\\n",
       "0          sra    8732970    SRX6621276         yes            yes   \n",
       "1          sra    8732971    SRX6621277         yes            yes   \n",
       "2          sra   10263497    SRX7845469         yes            yes   \n",
       "3          sra   10855787    SRX8344829         yes            yes   \n",
       "4          sra   10855906    SRX8344948         yes            yes   \n",
       "...        ...        ...           ...         ...            ...   \n",
       "40841      sra  100000005          None        None           None   \n",
       "40842      sra  100000006          None        None           None   \n",
       "40843      sra  100000007          None        None           None   \n",
       "40844      sra  100000008          None        None           None   \n",
       "40845      sra  100000009   SRX11094056         yes             no   \n",
       "\n",
       "      is_paired_end        lib_prep        tech_10x       cell_prep organism  \\\n",
       "0               yes    10x_Genomics     3_prime_gex     single_cell    mouse   \n",
       "1               yes    10x_Genomics           other     single_cell    mouse   \n",
       "2               yes    10x_Genomics     5_prime_gex     single_cell    human   \n",
       "3               yes      Smart-seq2  not_applicable     single_cell    mouse   \n",
       "4               yes      Smart-seq2  not_applicable     single_cell    mouse   \n",
       "...             ...             ...             ...             ...      ...   \n",
       "40841          None            None            None            None     None   \n",
       "40842          None            None            None            None     None   \n",
       "40843          None            None            None            None     None   \n",
       "40844          None            None            None            None     None   \n",
       "40845           yes  not_applicable  not_applicable  not_applicable    other   \n",
       "\n",
       "                                          tissue        disease  \\\n",
       "0                       RPE/choroid single cells  not specified   \n",
       "1                       RPE/choroid single cells  not specified   \n",
       "2      Peripheral Blood Mononuclear Cells (PBMC)  not specified   \n",
       "3                                           lung         unsure   \n",
       "4                                           lung         unsure   \n",
       "...                                          ...            ...   \n",
       "40841                                       None           None   \n",
       "40842                                       None           None   \n",
       "40843                                       None           None   \n",
       "40844                                       None           None   \n",
       "40845                                   mycelium         unsure   \n",
       "\n",
       "                                            purturbation  \\\n",
       "0      cultured choroidal GLI1+ cells treated with DM...   \n",
       "1      Cultured choroidal GLI1+ cells treated with DM...   \n",
       "2                                          not specified   \n",
       "3                                                 unsure   \n",
       "4                                                 unsure   \n",
       "...                                                  ...   \n",
       "40841                                               None   \n",
       "40842                                               None   \n",
       "40843                                               None   \n",
       "40844                                               None   \n",
       "40845  incubation substrates: xylan; sampling time: 4...   \n",
       "\n",
       "                                               cell_line czi_collection_id  \\\n",
       "0      B6129PF1/J, 3 month-old, RPE/choroid single cells              None   \n",
       "1      Strain background: C57BL/6J, Age: 3 month-old,...              None   \n",
       "2                                          not specified              None   \n",
       "3                                                 unsure              None   \n",
       "4                                                 unsure              None   \n",
       "...                                                  ...               ...   \n",
       "40841                                               None              None   \n",
       "40842                                               None              None   \n",
       "40843                                               None              None   \n",
       "40844                                               None              None   \n",
       "40845                               strain: delta_Tgace2              None   \n",
       "\n",
       "      czi_collection_name                                     notes  \\\n",
       "0                    None              Metadata obtained by SRAgent   \n",
       "1                    None              Metadata obtained by SRAgent   \n",
       "2                    None              Metadata obtained by SRAgent   \n",
       "3                    None              Metadata obtained by SRAgent   \n",
       "4                    None              Metadata obtained by SRAgent   \n",
       "...                   ...                                       ...   \n",
       "40841                None  New dataset found by Find-Datasets agent   \n",
       "40842                None  New dataset found by Find-Datasets agent   \n",
       "40843                None  New dataset found by Find-Datasets agent   \n",
       "40844                None  New dataset found by Find-Datasets agent   \n",
       "40845                None              Metadata obtained by SRAgent   \n",
       "\n",
       "                      created_at                 updated_at  \n",
       "0     2025-01-23 17:20:30.662485 2025-01-23 17:21:51.385908  \n",
       "1     2025-01-22 22:45:26.805249 2025-01-22 22:47:13.950104  \n",
       "2     2025-01-27 07:30:27.589695 2025-01-27 07:32:05.085304  \n",
       "3     2025-01-21 17:28:27.920468 2025-01-21 17:29:55.383231  \n",
       "4     2025-01-21 17:28:27.920468 2025-01-21 17:36:38.159210  \n",
       "...                          ...                        ...  \n",
       "40841 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40842 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40843 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40844 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40845 2025-01-16 13:20:31.752503 2025-01-16 13:26:41.517420  \n",
       "\n",
       "[40846 rows x 19 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \\\n",
    "    .distinct() \\\n",
    "    .where(tbl.notes != \"Processed by Chris Carpenter\")\n",
    "srx_metadata = pd.read_sql(str(stmt), conn_prod)\n",
    "srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [database, entrez_id, srx_accession, is_illumina, is_single_cell, is_paired_end, lib_prep, tech_10x, cell_prep, organism, tissue, disease, purturbation, cell_line, czi_collection_id, czi_collection_name, notes, created_at, updated_at]\n",
       "Index: []"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicate srx_accession values?\n",
    "x = srx_metadata[srx_metadata.srx_accession.notnull()]\n",
    "x[x.duplicated(subset=\"srx_accession\", keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    40082.0\n",
       "mean         1.0\n",
       "std          0.0\n",
       "min          1.0\n",
       "25%          1.0\n",
       "50%          1.0\n",
       "75%          1.0\n",
       "max          1.0\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicate srx_accession values?\n",
    "x[\"srx_accession\"].value_counts().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    14630.0\n",
       "mean         1.0\n",
       "std          0.0\n",
       "min          1.0\n",
       "25%          1.0\n",
       "50%          1.0\n",
       "75%          1.0\n",
       "max          1.0\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicate srx_accession values?\n",
    "x[\"entrez_id\"].value_counts().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8434"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of \"10X_Genomics\" samples\n",
    "srx_metadata[srx_metadata[\"lib_prep\"] == \"10x_Genomics\"].shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STAR results table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>feature</th>\n",
       "      <th>estimated_number_of_cells</th>\n",
       "      <th>fraction_of_unique_reads_in_cells</th>\n",
       "      <th>mean_gene_per_cell</th>\n",
       "      <th>mean_umi_per_cell</th>\n",
       "      <th>mean_feature_per_cell</th>\n",
       "      <th>median_gene_per_cell</th>\n",
       "      <th>median_umi_per_cell</th>\n",
       "      <th>median_feature_per_cell</th>\n",
       "      <th>...</th>\n",
       "      <th>reads_with_valid_barcodes</th>\n",
       "      <th>sequencing_saturation</th>\n",
       "      <th>total_feature_detected</th>\n",
       "      <th>umis_in_cells</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_gene</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_genefull</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_genefull_ex50pas</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_genefull_exonoverintron</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249739</td>\n",
       "      <td>SRX24915001</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>Hepatitis C Virus (HCV) reinfection outcomes</td>\n",
       "      <td>Botulinum toxin A injection</td>\n",
       "      <td>Memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:39:12.479315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249812</td>\n",
       "      <td>SRX24915074</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>HCV reinfection</td>\n",
       "      <td>mobile health intervention post-ACL reconstruc...</td>\n",
       "      <td>Memory B cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:39:14.822953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249944</td>\n",
       "      <td>SRX24915206</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>not specified</td>\n",
       "      <td>exercised, E2-specific B cells</td>\n",
       "      <td>human samples (Memory B cells)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:17:04.727422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>sra</td>\n",
       "      <td>34280289</td>\n",
       "      <td>ERX11662340</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>lung</td>\n",
       "      <td>non-small cell lung cancer</td>\n",
       "      <td>unsure</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:17:23.956995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sra</td>\n",
       "      <td>34602736</td>\n",
       "      <td>SRX25601950</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Peripheral Blood Mononuclear Cells (PBMCs)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:30:58.868003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   srx_accession organism\n",
       "0     SRX9556667    human\n",
       "1    ERX12060278    human\n",
       "2    ERX12060274    human\n",
       "3    ERX12060276    human\n",
       "4    ERX12060273    human\n",
       "5    ERX12060297    human\n",
       "6    ERX12060299    human\n",
       "7    ERX12060275    human\n",
       "8    SRX13549222    other\n",
       "9     SRX9556651    human\n",
       "10   ERX12060277    human"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # update \"other\" organisms to correct\n",
    "# idx = [\n",
    "#     ['SRX9556667', 'human'], \n",
    "#     ['ERX12060278', 'human'], \n",
    "#     ['ERX12060274', 'human'], \n",
    "#     ['ERX12060276', 'human'], \n",
    "#     ['ERX12060273', 'human'], \n",
    "#     ['ERX12060297', 'human'], \n",
    "#     ['ERX12060299', 'human'], \n",
    "#     ['ERX12060275', 'human'], \n",
    "#     ['SRX13549222', 'other'],\n",
    "#     ['SRX9556651', 'human'], \n",
    "#     ['ERX12060277', 'human'], \n",
    "# ]\n",
    "# idx = pd.DataFrame(idx, columns=[\"srx_accession\", \"organism\"])\n",
    "# idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sra</td>\n",
       "      <td>34602788</td>\n",
       "      <td>SRX25602002</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>Type 2 Diabetes Mellitus</td>\n",
       "      <td>Randomized controlled study on salt intake eff...</td>\n",
       "      <td>other</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:27.430411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>sra</td>\n",
       "      <td>34602910</td>\n",
       "      <td>SRX25602124</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Peripheral Blood Mononuclear Cells (PBMCs)</td>\n",
       "      <td>not specified</td>\n",
       "      <td>not specified</td>\n",
       "      <td>not specified</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:30:39.316203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>sra</td>\n",
       "      <td>34602926</td>\n",
       "      <td>SRX25602140</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>PBMCs (Peripheral Blood Mononuclear Cells)</td>\n",
       "      <td>none</td>\n",
       "      <td>effects of salt intake in healthy humans</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:38:49.139233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>sra</td>\n",
       "      <td>34648698</td>\n",
       "      <td>SRX25634198</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>prostate</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>anti-PD1 + Vehicle</td>\n",
       "      <td>Myc-CaP</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:44:30.489428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>sra</td>\n",
       "      <td>34914388</td>\n",
       "      <td>SRX25852474</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>epidermis</td>\n",
       "      <td>none</td>\n",
       "      <td>untreated</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:39:43.424840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sra</td>\n",
       "      <td>36086779</td>\n",
       "      <td>SRX26710377</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Duodenal Lymph Nodes)</td>\n",
       "      <td>Desmoid-type fibromatosis</td>\n",
       "      <td>Surgical resection</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:31:53.332445</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15945 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample      feature  estimated_number_of_cells  \\\n",
       "0      ERX10024831    gene_full                       2292   \n",
       "1      ERX10024831    gene_ex50                       2290   \n",
       "2      ERX10024831  gene_ex_int                       2290   \n",
       "3      ERX10024831     velocyto                          0   \n",
       "4      ERX10086874    gene_full                       2262   \n",
       "...            ...          ...                        ...   \n",
       "15940  ERX11899886         gene                      19391   \n",
       "15941  ERX11899886    gene_ex50                      20030   \n",
       "15942  ERX11899886  gene_ex_int                      20035   \n",
       "15943  ERX11899886    gene_full                      20016   \n",
       "15944  ERX11899886     velocyto                          0   \n",
       "\n",
       "       fraction_of_unique_reads_in_cells mean_gene_per_cell  \\\n",
       "0                               0.911745               None   \n",
       "1                               0.912572               None   \n",
       "2                               0.912561               None   \n",
       "3                                    NaN               None   \n",
       "4                               0.909667               None   \n",
       "...                                  ...                ...   \n",
       "15940                           0.903745               None   \n",
       "15941                           0.914242               None   \n",
       "15942                           0.914298               None   \n",
       "15943                           0.914711               None   \n",
       "15944                                NaN               None   \n",
       "\n",
       "       mean_umi_per_cell  mean_feature_per_cell median_gene_per_cell  \\\n",
       "0                12557.0                 2730.0                 None   \n",
       "1                17603.0                 2756.0                 None   \n",
       "2                17625.0                 2767.0                 None   \n",
       "3                    NaN                    NaN                 None   \n",
       "4                 4977.0                 2018.0                 None   \n",
       "...                  ...                    ...                  ...   \n",
       "15940             6698.0                 1871.0                 None   \n",
       "15941             7989.0                 2432.0                 None   \n",
       "15942             8012.0                 2447.0                 None   \n",
       "15943             7865.0                 2423.0                 None   \n",
       "15944                NaN                    NaN                 None   \n",
       "\n",
       "       median_umi_per_cell  median_feature_per_cell  ...  \\\n",
       "0                  10513.0                   2775.0  ...   \n",
       "1                  11784.0                   2799.0  ...   \n",
       "2                  11803.0                   2809.0  ...   \n",
       "3                      NaN                      NaN  ...   \n",
       "4                   4615.0                   2116.0  ...   \n",
       "...                    ...                      ...  ...   \n",
       "15940               4492.0                   1617.0  ...   \n",
       "15941               5618.0                   2163.0  ...   \n",
       "15942               5639.0                   2178.0  ...   \n",
       "15943               5565.0                   2159.0  ...   \n",
       "15944                  NaN                      NaN  ...   \n",
       "\n",
       "       reads_with_valid_barcodes  sequencing_saturation  \\\n",
       "0                       0.983305               0.593299   \n",
       "1                       0.983334               0.601129   \n",
       "2                       0.983322               0.601073   \n",
       "3                       0.984647                   -inf   \n",
       "4                       0.967224               0.887163   \n",
       "...                          ...                    ...   \n",
       "15940                   0.984381               0.761567   \n",
       "15941                   0.983904               0.765714   \n",
       "15942                   0.983833               0.765700   \n",
       "15943                   0.983820               0.765057   \n",
       "15944                   0.985624                   -inf   \n",
       "\n",
       "       total_feature_detected  umis_in_cells  \\\n",
       "0                     21731.0       28780763   \n",
       "1                     21911.0       40311609   \n",
       "2                     22001.0       40362108   \n",
       "3                         NaN              0   \n",
       "4                     21453.0       11258108   \n",
       "...                       ...            ...   \n",
       "15940                 29240.0      129894014   \n",
       "15941                 32603.0      160037142   \n",
       "15942                 32621.0      160539579   \n",
       "15943                 31181.0      157427372   \n",
       "15944                     NaN              0   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_gene  \\\n",
       "0                                       NaN   \n",
       "1                                       NaN   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4                                       NaN   \n",
       "...                                     ...   \n",
       "15940                           576267904.0   \n",
       "15941                                   NaN   \n",
       "15942                                   NaN   \n",
       "15943                                   NaN   \n",
       "15944                                   NaN   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_genefull  \\\n",
       "0                                    71181072.0   \n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3                                           NaN   \n",
       "4                                   108541887.0   \n",
       "...                                         ...   \n",
       "15940                                       NaN   \n",
       "15941                                       NaN   \n",
       "15942                                       NaN   \n",
       "15943                               712975893.0   \n",
       "15944                                       NaN   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_genefull_ex50pas  \\\n",
       "0                                                   NaN   \n",
       "1                                           101599277.0   \n",
       "2                                                   NaN   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "15940                                               NaN   \n",
       "15941                                       726711495.0   \n",
       "15942                                               NaN   \n",
       "15943                                               NaN   \n",
       "15944                                               NaN   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_genefull_exonoverintron  \\\n",
       "0                                                    NaN         \n",
       "1                                                    NaN         \n",
       "2                                            101712901.0         \n",
       "3                                                    NaN         \n",
       "4                                                    NaN         \n",
       "...                                                  ...         \n",
       "15940                                                NaN         \n",
       "15941                                                NaN         \n",
       "15942                                        729034278.0         \n",
       "15943                                                NaN         \n",
       "15944                                                NaN         \n",
       "\n",
       "                      created_at                 updated_at  \n",
       "0     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "1     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "2     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "3     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "4     2025-01-13 15:17:54.767872 2025-01-14 20:06:25.557352  \n",
       "...                          ...                        ...  \n",
       "15940 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15941 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15942 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15943 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15944 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "\n",
       "[15945 rows x 36 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"screcounter_star_results\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \n",
    "star_results = pd.read_sql(str(stmt), conn_prod)\n",
    "star_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3189.0\n",
       "mean        5.0\n",
       "std         0.0\n",
       "min         5.0\n",
       "25%         5.0\n",
       "50%         5.0\n",
       "75%         5.0\n",
       "max         5.0\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "star_results[\"sample\"].value_counts().describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameters table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>star_index</th>\n",
       "      <th>cell_barcode_length</th>\n",
       "      <th>umi_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample, barcodes, star_index, cell_barcode_length, umi_length, strand, created_at, updated_at]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"screcounter_star_params\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \\\n",
    "    .where(tbl.star_index == \"\")\n",
    "params = pd.read_sql(str(stmt), conn_prod)\n",
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>sra</td>\n",
       "      <td>36086788</td>\n",
       "      <td>SRX26710386</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Dendritic Lymph Nodes)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>SV (specific treatment not detailed)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:15:53.177758</td>\n",
       "      <td>2025-01-06 22:17:31.151426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>sra</td>\n",
       "      <td>36086856</td>\n",
       "      <td>SRX26710454</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Duodenal Lymph Nodes)</td>\n",
       "      <td>none</td>\n",
       "      <td>HP (Helminth Infection)</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:42:38.083215</td>\n",
       "      <td>2025-01-03 18:45:29.642017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>sra</td>\n",
       "      <td>36086915</td>\n",
       "      <td>SRX26710513</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Duodenal Lymph Nodes)</td>\n",
       "      <td>not specified</td>\n",
       "      <td>OV (presumably refers to a treatment condition)</td>\n",
       "      <td>not specified</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:30:10.729039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>sra</td>\n",
       "      <td>36087069</td>\n",
       "      <td>SRX26710667</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Duodenal Lymph Node)</td>\n",
       "      <td>related to cancerous mouse prostate samples</td>\n",
       "      <td>OV treatment</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-06 22:37:10.538238</td>\n",
       "      <td>2025-01-06 22:39:46.008134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>sra</td>\n",
       "      <td>36087118</td>\n",
       "      <td>SRX26710716</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>D-LN (Duodenal Lymph Nodes)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>OV</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:30:26.053848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>sra</td>\n",
       "      <td>36087317</td>\n",
       "      <td>SRX26710915</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>duodenal lymph nodes</td>\n",
       "      <td>unsure</td>\n",
       "      <td>SV treatment</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-07 00:28:36.611049</td>\n",
       "      <td>2025-01-07 00:30:23.602324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>647 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   database  entrez_id srx_accession is_illumina is_single_cell is_paired_end  \\\n",
       "0       sra   26659536   SRX19392127         yes            yes           yes   \n",
       "1       sra   27454961   ERX10668462         yes            yes           yes   \n",
       "2       sra   27454971   ERX10668472         yes            yes           yes   \n",
       "3       sra   27455317   ERX10668818         yes            yes           yes   \n",
       "4       sra   27694563   SRX20273630         yes            yes           yes   \n",
       "5       sra   29919592   SRX22006379         yes            yes           yes   \n",
       "6       sra   30864469   SRX22822524         yes            yes           yes   \n",
       "7       sra   30865352   SRX22823407         yes            yes           yes   \n",
       "8       sra   30865785   SRX22823840         yes            yes           yes   \n",
       "9       sra   32469529   SRX24158623         yes            yes           yes   \n",
       "10      sra   32702534   SRX24387553         yes            yes           yes   \n",
       "11      sra   32719383   SRX24404397         yes            yes           yes   \n",
       "12      sra   32777790   SRX24461697         yes            yes           yes   \n",
       "13      sra   32927474   SRX24608712         yes            yes           yes   \n",
       "14      sra   32927520   SRX24608758         yes            yes           yes   \n",
       "15      sra   33249216   SRX24914478         yes            yes           yes   \n",
       "16      sra   33249222   SRX24914484         yes            yes           yes   \n",
       "17      sra   33249325   SRX24914587         yes            yes           yes   \n",
       "18      sra   33249385   SRX24914647         yes            yes           yes   \n",
       "19      sra   33249739   SRX24915001         yes            yes           yes   \n",
       "20      sra   33249812   SRX24915074         yes            yes           yes   \n",
       "21      sra   33249944   SRX24915206         yes            yes           yes   \n",
       "22      sra   34280289   ERX11662340         yes            yes           yes   \n",
       "23      sra   34602736   SRX25601950         yes            yes           yes   \n",
       "24      sra   34602788   SRX25602002         yes            yes           yes   \n",
       "25      sra   34602910   SRX25602124         yes            yes           yes   \n",
       "26      sra   34602926   SRX25602140         yes            yes           yes   \n",
       "27      sra   34648698   SRX25634198         yes            yes           yes   \n",
       "28      sra   34914388   SRX25852474         yes            yes           yes   \n",
       "29      sra   36086779   SRX26710377         yes            yes           yes   \n",
       "30      sra   36086788   SRX26710386         yes            yes           yes   \n",
       "31      sra   36086856   SRX26710454         yes            yes           yes   \n",
       "32      sra   36086915   SRX26710513         yes            yes           yes   \n",
       "33      sra   36087069   SRX26710667         yes            yes           yes   \n",
       "34      sra   36087118   SRX26710716         yes            yes           yes   \n",
       "35      sra   36087317   SRX26710915         yes            yes           yes   \n",
       "\n",
       "        lib_prep        tech_10x    cell_prep organism  \\\n",
       "0   10x_Genomics     3_prime_gex  single_cell    mouse   \n",
       "1   10x_Genomics           other  single_cell    mouse   \n",
       "2          other  not_applicable  single_cell    mouse   \n",
       "3          other  not_applicable  single_cell    mouse   \n",
       "4          other  not_applicable  single_cell    human   \n",
       "5     Smart-seq2  not_applicable  single_cell    human   \n",
       "6       MARS-seq  not_applicable  single_cell    mouse   \n",
       "7          other  not_applicable  single_cell    mouse   \n",
       "8       MARS-seq  not_applicable  single_cell    mouse   \n",
       "9   10x_Genomics     3_prime_gex  single_cell    human   \n",
       "10  10x_Genomics     5_prime_gex  single_cell    human   \n",
       "11  10x_Genomics     5_prime_gex  single_cell    human   \n",
       "12  10x_Genomics     3_prime_gex  single_cell    mouse   \n",
       "13      MARS-seq  not_applicable  single_cell    mouse   \n",
       "14      MARS-seq  not_applicable  single_cell    mouse   \n",
       "15    Smart-seq2  not_applicable  single_cell    human   \n",
       "16    Smart-seq2  not_applicable  single_cell    human   \n",
       "17    Smart-seq2  not_applicable  single_cell    human   \n",
       "18    Smart-seq2  not_applicable  single_cell    human   \n",
       "19    Smart-seq2  not_applicable  single_cell    human   \n",
       "20    Smart-seq2  not_applicable  single_cell    human   \n",
       "21    Smart-seq2  not_applicable  single_cell    human   \n",
       "22  10x_Genomics     3_prime_gex  single_cell    human   \n",
       "23  10x_Genomics           other  single_cell    human   \n",
       "24         other  not_applicable  single_cell    human   \n",
       "25  10x_Genomics  not_applicable  single_cell    human   \n",
       "26         other  not_applicable  single_cell    human   \n",
       "27         other  not_applicable  single_cell    mouse   \n",
       "28  10x_Genomics     3_prime_gex  single_cell    human   \n",
       "29    Smart-seq2  not_applicable  single_cell    mouse   \n",
       "30  10x_Genomics           other  single_cell    mouse   \n",
       "31    Smart-seq2           other  single_cell    mouse   \n",
       "32  10x_Genomics           other  single_cell    mouse   \n",
       "33  10x_Genomics     3_prime_gex  single_cell    mouse   \n",
       "34  10x_Genomics           other  single_cell    mouse   \n",
       "35  10x_Genomics  not_applicable  single_cell    mouse   \n",
       "\n",
       "                                        tissue  \\\n",
       "0                                    S1 cortex   \n",
       "1                   ILC progenitors from femur   \n",
       "2                   ILC progenitors from femur   \n",
       "3                                        femur   \n",
       "4                                       HCT116   \n",
       "5                                   blastocyst   \n",
       "6                                        tumor   \n",
       "7                                        tumor   \n",
       "8                                        tumor   \n",
       "9                         primary cell culture   \n",
       "10        bone marrow 90d post treatment (T20)   \n",
       "11                                 bone marrow   \n",
       "12                                  whole skin   \n",
       "13                                      embryo   \n",
       "14                                mouse embryo   \n",
       "15                                       blood   \n",
       "16                                       blood   \n",
       "17                                       blood   \n",
       "18                                       blood   \n",
       "19                                       blood   \n",
       "20                                       blood   \n",
       "21                                       blood   \n",
       "22                                        lung   \n",
       "23  Peripheral Blood Mononuclear Cells (PBMCs)   \n",
       "24                                       PBMCs   \n",
       "25  Peripheral Blood Mononuclear Cells (PBMCs)   \n",
       "26  PBMCs (Peripheral Blood Mononuclear Cells)   \n",
       "27                                    prostate   \n",
       "28                                   epidermis   \n",
       "29                 D-LN (Duodenal Lymph Nodes)   \n",
       "30                D-LN (Dendritic Lymph Nodes)   \n",
       "31                 D-LN (Duodenal Lymph Nodes)   \n",
       "32                 D-LN (Duodenal Lymph Nodes)   \n",
       "33                  D-LN (Duodenal Lymph Node)   \n",
       "34                 D-LN (Duodenal Lymph Nodes)   \n",
       "35                        duodenal lymph nodes   \n",
       "\n",
       "                                         disease  \\\n",
       "0                      neuropsychiatric outcomes   \n",
       "1                                         unsure   \n",
       "2                                         unsure   \n",
       "3                                         unsure   \n",
       "4                                  not specified   \n",
       "5                                         unsure   \n",
       "6                                         unsure   \n",
       "7                              phaeohyphomycosis   \n",
       "8                           cancer immunotherapy   \n",
       "9                                       glaucoma   \n",
       "10               multiple myeloma, early relapse   \n",
       "11                              multiple myeloma   \n",
       "12                                 not specified   \n",
       "13                                          none   \n",
       "14                                          none   \n",
       "15                                         other   \n",
       "16                       hepatitis C virus (HCV)   \n",
       "17                                        unsure   \n",
       "18           hepatitis C virus (HCV) reinfection   \n",
       "19  Hepatitis C Virus (HCV) reinfection outcomes   \n",
       "20                               HCV reinfection   \n",
       "21                                 not specified   \n",
       "22                    non-small cell lung cancer   \n",
       "23                                        unsure   \n",
       "24                      Type 2 Diabetes Mellitus   \n",
       "25                                 not specified   \n",
       "26                                          none   \n",
       "27                               prostate cancer   \n",
       "28                                          none   \n",
       "29                     Desmoid-type fibromatosis   \n",
       "30                                        unsure   \n",
       "31                                          none   \n",
       "32                                 not specified   \n",
       "33   related to cancerous mouse prostate samples   \n",
       "34                                        unsure   \n",
       "35                                        unsure   \n",
       "\n",
       "                                         purturbation  \\\n",
       "0                          maternal immune activation   \n",
       "1                                              unsure   \n",
       "2                                              unsure   \n",
       "3                          Bcl11b regulatory networks   \n",
       "4                                            infected   \n",
       "5                                              unsure   \n",
       "6                                              unsure   \n",
       "7                                  anti-PD1 treatment   \n",
       "8                                      none specified   \n",
       "9   primary hTM cells isolated from human donor co...   \n",
       "10                            BCMA CAR-T cell therapy   \n",
       "11                            BCMA CAR-T cell therapy   \n",
       "12                                          tamoxifen   \n",
       "13                 Bmp4 KO Heterozygous (C57BL/6x129)   \n",
       "14                                     BMP4 signaling   \n",
       "15                                          exercised   \n",
       "16              E2-specific B cells sorted from PBMCs   \n",
       "17                                             unsure   \n",
       "18                                             unsure   \n",
       "19                        Botulinum toxin A injection   \n",
       "20  mobile health intervention post-ACL reconstruc...   \n",
       "21                     exercised, E2-specific B cells   \n",
       "22                                             unsure   \n",
       "23                                             unsure   \n",
       "24  Randomized controlled study on salt intake eff...   \n",
       "25                                      not specified   \n",
       "26           effects of salt intake in healthy humans   \n",
       "27                                 anti-PD1 + Vehicle   \n",
       "28                                          untreated   \n",
       "29                                 Surgical resection   \n",
       "30               SV (specific treatment not detailed)   \n",
       "31                            HP (Helminth Infection)   \n",
       "32    OV (presumably refers to a treatment condition)   \n",
       "33                                       OV treatment   \n",
       "34                                                 OV   \n",
       "35                                       SV treatment   \n",
       "\n",
       "                             cell_line czi_collection_id czi_collection_name  \\\n",
       "0   microglia, P14, male, C57Bl/6N, WT              None                None   \n",
       "1                               unsure              None                None   \n",
       "2                           ILC2 cells              None                None   \n",
       "3                               unsure              None                None   \n",
       "4                               HCT116              None                None   \n",
       "5                inner cell mass (ICM)              None                None   \n",
       "6                             OT1 CD8+              None                None   \n",
       "7              endogenous CD8+ T cells              None                None   \n",
       "8                   B16-OVA, MC38, LLC              None                None   \n",
       "9                              healthy              None                None   \n",
       "10   T-cells (from Myeloma patient 07)              None                None   \n",
       "11                       myeloid cells              None                None   \n",
       "12        K14CreERTM/LSL-Sox2-IRES-GFP              None                None   \n",
       "13                                none              None                None   \n",
       "14                                none              None                None   \n",
       "15                      memory B cells              None                None   \n",
       "16                      memory B cells              None                None   \n",
       "17                      memory B cells              None                None   \n",
       "18                      Memory B cells              None                None   \n",
       "19                      Memory B cells              None                None   \n",
       "20                      Memory B cells              None                None   \n",
       "21      human samples (Memory B cells)              None                None   \n",
       "22                      not_applicable              None                None   \n",
       "23                      not_applicable              None                None   \n",
       "24                               other              None                None   \n",
       "25                       not specified              None                None   \n",
       "26                                none              None                None   \n",
       "27                             Myc-CaP              None                None   \n",
       "28                                none              None                None   \n",
       "29                      not_applicable              None                None   \n",
       "30                              unsure              None                None   \n",
       "31                                none              None                None   \n",
       "32                       not specified              None                None   \n",
       "33                              unsure              None                None   \n",
       "34                              unsure              None                None   \n",
       "35                      not_applicable              None                None   \n",
       "\n",
       "                           notes                 created_at  \\\n",
       "0   Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "1   Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "2   Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "3   Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "4   Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "5   Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "6   Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "7   Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "8   Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "9   Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "10  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "11  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "12  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "13  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "14  Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "15  Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "16  Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "17  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "18  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "19  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "20  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "21  Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "22  Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "23  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "24  Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "25  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "26  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "27  Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "28  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "29  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "30  Metadata obtained by SRAgent 2025-01-06 22:15:53.177758   \n",
       "31  Metadata obtained by SRAgent 2025-01-03 18:42:38.083215   \n",
       "32  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "33  Metadata obtained by SRAgent 2025-01-06 22:37:10.538238   \n",
       "34  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "35  Metadata obtained by SRAgent 2025-01-07 00:28:36.611049   \n",
       "\n",
       "                   updated_at  \n",
       "0  2025-01-06 22:19:05.896057  \n",
       "1  2025-01-06 22:19:02.038324  \n",
       "2  2025-01-06 22:38:48.116573  \n",
       "3  2025-01-07 00:30:22.045996  \n",
       "4  2025-01-03 18:44:17.048627  \n",
       "5  2025-01-03 18:43:57.602382  \n",
       "6  2025-01-06 22:18:04.534748  \n",
       "7  2025-01-03 18:45:48.003195  \n",
       "8  2025-01-06 22:18:33.867744  \n",
       "9  2025-01-06 22:40:04.684759  \n",
       "10 2025-01-07 00:31:18.187839  \n",
       "11 2025-01-06 22:38:18.491089  \n",
       "12 2025-01-06 22:38:39.671695  \n",
       "13 2025-01-07 00:31:29.665074  \n",
       "14 2025-01-06 22:17:18.538773  \n",
       "15 2025-01-06 22:17:29.575513  \n",
       "16 2025-01-03 18:43:48.934750  \n",
       "17 2025-01-07 00:29:29.430436  \n",
       "18 2025-01-06 22:40:07.232871  \n",
       "19 2025-01-06 22:39:12.479315  \n",
       "20 2025-01-06 22:39:14.822953  \n",
       "21 2025-01-06 22:17:04.727422  \n",
       "22 2025-01-06 22:17:23.956995  \n",
       "23 2025-01-07 00:30:58.868003  \n",
       "24 2025-01-03 18:45:27.430411  \n",
       "25 2025-01-07 00:30:39.316203  \n",
       "26 2025-01-06 22:38:49.139233  \n",
       "27 2025-01-03 18:44:30.489428  \n",
       "28 2025-01-06 22:39:43.424840  \n",
       "29 2025-01-07 00:31:53.332445  \n",
       "30 2025-01-06 22:17:31.151426  \n",
       "31 2025-01-03 18:45:29.642017  \n",
       "32 2025-01-07 00:30:10.729039  \n",
       "33 2025-01-06 22:39:46.008134  \n",
       "34 2025-01-07 00:30:26.053848  \n",
       "35 2025-01-07 00:30:23.602324  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \\\n",
    "    .distinct() \\\n",
    "    .where(tbl.notes != \"Processed by Chris Carpenter\")\n",
    "srx_metadata = pd.read_sql(str(stmt), conn_prod)\n",
    "srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>star_index</th>\n",
       "      <th>cell_barcode_length</th>\n",
       "      <th>umi_length</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX11662340</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX19392127</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX24158623</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX22159680</td>\n",
       "      <td>737K-august-2016.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>Reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX24228136</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRX24387553</td>\n",
       "      <td>737K-august-2016.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>Reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRX24461697</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRX24523046</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRX24585614</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRX25852474</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRX21638105</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample              barcodes              star_index  \\\n",
       "0   ERX11662340  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "1   SRX19392127  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "2   SRX24158623  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "3   SRX22159680  737K-august-2016.txt  star_refData_2020_hg38   \n",
       "4   SRX24228136  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "5   SRX24387553  737K-august-2016.txt  star_refData_2020_hg38   \n",
       "6   SRX24461697  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "7   SRX24523046  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "8   SRX24585614  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "9   SRX25852474  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "10  SRX21638105  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "\n",
       "    cell_barcode_length  umi_length   strand  \n",
       "0                    16          12  Forward  \n",
       "1                    16          12  Forward  \n",
       "2                    16          12  Forward  \n",
       "3                    16          10  Reverse  \n",
       "4                    16          12  Forward  \n",
       "5                    16          10  Reverse  \n",
       "6                    16          12  Forward  \n",
       "7                    16          12  Forward  \n",
       "8                    16          12  Forward  \n",
       "9                    16          12  Forward  \n",
       "10                   16          12  Forward  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge by srx_accession\n",
    "def set_star_ref(row):\n",
    "    if row.organism == \"human\":\n",
    "        return \"star_refData_2020_hg38\"\n",
    "    elif row.organism == \"mouse\":\n",
    "        return \"star_refData_2020_mm10\"\n",
    "    else:\n",
    "        raise ValueError(f\"organism {row.organism} not supported\")\n",
    "\n",
    "params = params.merge(srx_metadata, left_on=\"sample\", right_on=\"srx_accession\") \n",
    "params[\"star_index\"] = params.apply(lambda x: set_star_ref(x), axis=1)\n",
    "params = params.drop(columns=[\"created_at\", \"updated_at\", \"srx_accession\", \"organism\"], axis=1)\n",
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# upsert\n",
    "db_upsert(params, \"screcounter_star_params\", conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# --OLD--"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out records with no SRX accessions\n",
    "\n",
    "`New dataset found by Find-Datasets agent`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entrez_ids = srx_metadata[srx_metadata[\"srx_accession\"].isna()][\"entrez_id\"].unique().tolist()\n",
    "entrez_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete extrez_ids from srx_metadata\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .delete() \\\n",
    "    .where(tbl.entrez_id.isin(entrez_ids))\n",
    "with conn_prod.cursor() as cur:\n",
    "    cur.execute(str(stmt))\n",
    "    conn_prod.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete from log\n",
    "tbl = Table(\"screcounter_log\")\n",
    "to_rm = [\"ERX11146221\"]\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .delete() \\\n",
    "    .where(tbl.sample.isin(to_rm))\n",
    "with conn_prod.cursor() as cur:\n",
    "    cur.execute(str(stmt))\n",
    "    conn_prod.commit()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
