{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Check on the test and/or production databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv(override=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pypika import Query, Table, Field, Column, Criterion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.upsert import db_upsert\n",
    "from SRAgent.db.utils import db_list_tables, db_glimpse_tables, db_get_table, execute_query\n",
    "from SRAgent.db.get import db_find_srx\n",
    "from SRAgent.db.create import create_table, create_table_router"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "screcounter_star_results\n",
      "eval\n",
      "screcounter_trace\n",
      "srx_srr\n",
      "srx_metadata\n",
      "screcounter_log\n",
      "screcounter_star_params\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in prod\n",
    "os.environ['DYNACONF'] = 'prod'\n",
    "conn_prod = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_prod)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['screcounter_star_results',\n",
       " 'eval',\n",
       " 'screcounter_trace',\n",
       " 'srx_srr',\n",
       " 'srx_metadata',\n",
       " 'screcounter_log',\n",
       " 'screcounter_star_params']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_list_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: screcounter_star_results --#\n",
      "sample\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\tmedian_reads_per_cell\tq30_bases_in_cb_umi\tq30_bases_in_rna_read\treads_mapped_to_gene__unique_gene\treads_mapped_to_gene__unique_multiple_gene\treads_mapped_to_genefull__unique_genefull\treads_mapped_to_genefull__unique_multiple_genefull\treads_mapped_to_genefull_ex50pas__unique_genefull_ex50pas\treads_mapped_to_genefull_ex50pas__unique_multiple_genefull_ex50\treads_mapped_to_genefull_exonoverintron__unique_genefull_exonov\treads_mapped_to_genefull_exonoverintron__unique_multiple_genefu\treads_mapped_to_genome__unique\treads_mapped_to_genome__unique_multiple\treads_mapped_to_velocyto__unique_velocyto\treads_mapped_to_velocyto__unique_multiple_velocyto\treads_with_valid_barcodes\tsequencing_saturation\ttotal_feature_detected\tumis_in_cells\tunique_reads_in_cells_mapped_to_gene\tunique_reads_in_cells_mapped_to_genefull\tunique_reads_in_cells_mapped_to_genefull_ex50pas\tunique_reads_in_cells_mapped_to_genefull_exonoverintron\tcreated_at\tupdated_at\n",
      "ERX10024831\tgene_full\t2292\t0.911745\t\t12557.0\t2730.0\t\t10513.0\t2775.0\t146018831\t25399.0\t0.968155\t0.918635\t\t\t0.534665\t0.788618\t\t\t\t\t0.734504\t0.87767\t\t\t0.983305\t0.593299\t21731.0\t28780763\t\t71181072.0\t\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tgene_ex50\t2290\t0.912572\t\t17603.0\t2756.0\t\t11784.0\t2799.0\t146018831\t28689.0\t0.968155\t0.918635\t\t\t\t\t0.762455\t\t\t\t0.734504\t0.87767\t\t\t0.983334\t0.601129\t21911.0\t40311609\t\t\t101599277.0\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tgene_ex_int\t2290\t0.912561\t\t17625.0\t2767.0\t\t11803.0\t2809.0\t146018831\t28735.0\t0.968155\t0.918635\t\t\t\t\t\t\t\t\t0.734504\t0.87767\t\t\t0.983322\t0.601073\t22001.0\t40362108\t\t\t\t101712901.0\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tvelocyto\t0\t\t\t\t\t\t\t\t146018831\t\t0.968155\t0.918635\t\t\t\t\t\t\t\t\t0.734504\t0.87767\t0.0\t0.0\t0.984647\t-inf\t\t0\t\t\t\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10086874\tgene_full\t2262\t0.909667\t\t4977.0\t2018.0\t\t4615.0\t2116.0\t160000000\t43880.0\t0.952429\t0.920767\t\t\t0.745753\t0.801599\t\t\t\t\t0.800461\t0.924752\t\t\t0.967224\t0.887163\t21453.0\t11258108\t\t108541887.0\t\t\t2025-01-13 15:17:54.767872\t2025-01-14 20:06:25.557352\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "\n",
      "#-- Table: screcounter_trace --#\n",
      "task_id\thash\tnative_id\tname\tstatus\texit\tsubmit\tcontainer\tcpus\ttime\tdisk\tmemory\tattempt\tduration\trealtime\tcpu_percent\tpeak_rss\tpeak_vmem\trchar\twchar\tworkdir\tscratch\tcreated_at\tupdated_at\n",
      "1\t56/d8d487\tnf-56d8d487-1736775758365/0/nf-56d8d487-173677-6a5e0c92-ed1e-4c820\tDB_ACC_WF:GET_DB_ACCESSIONS\tCOMPLETED\t0\t2025-01-13 05:42:40.879\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 53s\t1.1s\t66.4%\t3 MB\t4.2 MB\t20.1 MB\t91.6 KB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/56/d8d4879ac276a9d30ec850f1ecf20a\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "2\te8/3237c0\tnf-e83237c0-1736775874725/0/nf-e83237c0-173677-e84d9f6b-20f7-48380\tSRA_STAT (1)\tCOMPLETED\t0\t2025-01-13 05:44:35.213\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t40.2%\t79.2 MB\t543.5 MB\t14.8 MB\t1.8 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/e8/3237c03758be58eca921e818fa3e0e\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "3\t64/3df771\tnf-643df771-1736775874726/0/nf-643df771-173677-54f5b596-489b-43870\tSRA_STAT (2)\tCOMPLETED\t0\t2025-01-13 05:44:35.674\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t37.7%\t79.5 MB\t543.5 MB\t14.8 MB\t755 KB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/64/3df7716232b892d0d8a7fa72ece0b8\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "4\t4e/793c08\tnf-4e793c08-1736775874789/0/nf-4e793c08-173677-239dfe74-6cc5-41130\tSRA_STAT (3)\tCOMPLETED\t0\t2025-01-13 05:44:36.077\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t39.1%\t79.4 MB\t543.5 MB\t14.8 MB\t1.7 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/4e/793c0898e1f5b974d97dca646c8fa2\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "5\tce/e8e21e\tnf-cee8e21e-1736775874795/0/nf-cee8e21e-173677-f6013cb6-ca08-40df0\tSRA_STAT (4)\tCOMPLETED\t0\t2025-01-13 05:44:36.533\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 38s\t1.3s\t39.7%\t79.1 MB\t543.5 MB\t14.8 MB\t1.7 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/ce/e8e21e2d91358d48f9b67118002b20\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX24914484\tSRR29400843\t2025-01-03 18:43:49.275138\t2025-01-03 18:43:49.275138\n",
      "SRX22006379\tSRR26297925\t2025-01-03 18:43:57.864762\t2025-01-03 18:43:57.864762\n",
      "SRX20273630\tSRR24488012\t2025-01-03 18:44:17.303662\t2025-01-03 18:44:17.303662\n",
      "SRX25634198\tSRR30167498\t2025-01-03 18:44:30.759838\t2025-01-03 18:44:30.759838\n",
      "SRX25602002\tSRR30132651\t2025-01-03 18:45:27.695421\t2025-01-03 18:45:27.695421\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t29110064\tERX11148781\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t73f82ac8-15cc-4fcd-87f8-5683723fce7f\tDevelopmental cell programs are co-opted in inflammatory skin disease\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t19007785\tSRX13670569\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t91c8e321-566f-4f9d-b89e-3a164be654d5\tNeuron type-specific effects of human aging and sex on DNA methylation and transcription\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t12488012\tSRX9556597\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\ted9185e3-5b82-40c7-9824-b2141590c7f0\tTime-resolved Systems Immunology Reveals a Late Juncture Linked to Fatal COVID-19\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t21270435\tERX8791959\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t62ef75e4-cbea-454e-a0ce-998ec40223d3\tCross-tissue immune cell analysis reveals tissue-specific features in humans\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t34493113\tERX12558784\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\tec329aed-22bc-4d6e-8935-8282dcb1acac\tEarly human lung immune cell development and its role in epithelial cell fate\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "ERX10024831\tERR10509460\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10024831\tERR10509461\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10086874\tERR10614597\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10086874\tERR10614598\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10335126\tERR10891936\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "\n",
      "#-- Table: screcounter_star_params --#\n",
      "sample\tbarcodes\tstar_index\tcell_barcode_length\tumi_length\tstrand\tcreated_at\tupdated_at\n",
      "ERX10086874\t737K-august-2016.txt\tstar_refData_2020_mm10\t16\t10\tForward\t2025-01-13 13:58:53.927797\t2025-01-13 13:58:53.927797\n",
      "ERX10396585\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:56.112918\t2025-01-13 13:58:56.112918\n",
      "ERX10396584\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:58.510780\t2025-01-13 13:58:58.510780\n",
      "ERX10690512\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:58.640112\t2025-01-13 13:58:58.640112\n",
      "ERX10855739\t737K-august-2016.txt\tstar_refData_2020_hg38\t16\t10\tForward\t2025-01-13 13:58:59.433825\t2025-01-13 13:58:59.433825\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>8732970</td>\n",
       "      <td>SRX6621276</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>RPE/choroid single cells</td>\n",
       "      <td>not specified</td>\n",
       "      <td>cultured choroidal GLI1+ cells treated with DM...</td>\n",
       "      <td>B6129PF1/J, 3 month-old, RPE/choroid single cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-23 17:20:30.662485</td>\n",
       "      <td>2025-01-23 17:21:51.385908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>8732971</td>\n",
       "      <td>SRX6621277</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>RPE/choroid single cells</td>\n",
       "      <td>not specified</td>\n",
       "      <td>Cultured choroidal GLI1+ cells treated with DM...</td>\n",
       "      <td>Strain background: C57BL/6J, Age: 3 month-old,...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-22 22:45:26.805249</td>\n",
       "      <td>2025-01-22 22:47:13.950104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>10263497</td>\n",
       "      <td>SRX7845469</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Peripheral Blood Mononuclear Cells (PBMC)</td>\n",
       "      <td>not specified</td>\n",
       "      <td>not specified</td>\n",
       "      <td>not specified</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-27 07:30:27.589695</td>\n",
       "      <td>2025-01-27 07:32:05.085304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>10855787</td>\n",
       "      <td>SRX8344829</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>lung</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-21 17:28:27.920468</td>\n",
       "      <td>2025-01-21 17:29:55.383231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>10855906</td>\n",
       "      <td>SRX8344948</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>lung</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-21 17:28:27.920468</td>\n",
       "      <td>2025-01-21 17:36:38.159210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40841</th>\n",
       "      <td>sra</td>\n",
       "      <td>100000005</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>New dataset found by Find-Datasets agent</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40842</th>\n",
       "      <td>sra</td>\n",
       "      <td>100000006</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>New dataset found by Find-Datasets agent</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40843</th>\n",
       "      <td>sra</td>\n",
       "      <td>100000007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>New dataset found by Find-Datasets agent</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40844</th>\n",
       "      <td>sra</td>\n",
       "      <td>100000008</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>New dataset found by Find-Datasets agent</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40845</th>\n",
       "      <td>sra</td>\n",
       "      <td>100000009</td>\n",
       "      <td>SRX11094056</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>other</td>\n",
       "      <td>mycelium</td>\n",
       "      <td>unsure</td>\n",
       "      <td>incubation substrates: xylan; sampling time: 4...</td>\n",
       "      <td>strain: delta_Tgace2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-16 13:20:31.752503</td>\n",
       "      <td>2025-01-16 13:26:41.517420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40846 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession is_illumina is_single_cell  \\\n",
       "0          sra    8732970    SRX6621276         yes            yes   \n",
       "1          sra    8732971    SRX6621277         yes            yes   \n",
       "2          sra   10263497    SRX7845469         yes            yes   \n",
       "3          sra   10855787    SRX8344829         yes            yes   \n",
       "4          sra   10855906    SRX8344948         yes            yes   \n",
       "...        ...        ...           ...         ...            ...   \n",
       "40841      sra  100000005          None        None           None   \n",
       "40842      sra  100000006          None        None           None   \n",
       "40843      sra  100000007          None        None           None   \n",
       "40844      sra  100000008          None        None           None   \n",
       "40845      sra  100000009   SRX11094056         yes             no   \n",
       "\n",
       "      is_paired_end        lib_prep        tech_10x       cell_prep organism  \\\n",
       "0               yes    10x_Genomics     3_prime_gex     single_cell    mouse   \n",
       "1               yes    10x_Genomics           other     single_cell    mouse   \n",
       "2               yes    10x_Genomics     5_prime_gex     single_cell    human   \n",
       "3               yes      Smart-seq2  not_applicable     single_cell    mouse   \n",
       "4               yes      Smart-seq2  not_applicable     single_cell    mouse   \n",
       "...             ...             ...             ...             ...      ...   \n",
       "40841          None            None            None            None     None   \n",
       "40842          None            None            None            None     None   \n",
       "40843          None            None            None            None     None   \n",
       "40844          None            None            None            None     None   \n",
       "40845           yes  not_applicable  not_applicable  not_applicable    other   \n",
       "\n",
       "                                          tissue        disease  \\\n",
       "0                       RPE/choroid single cells  not specified   \n",
       "1                       RPE/choroid single cells  not specified   \n",
       "2      Peripheral Blood Mononuclear Cells (PBMC)  not specified   \n",
       "3                                           lung         unsure   \n",
       "4                                           lung         unsure   \n",
       "...                                          ...            ...   \n",
       "40841                                       None           None   \n",
       "40842                                       None           None   \n",
       "40843                                       None           None   \n",
       "40844                                       None           None   \n",
       "40845                                   mycelium         unsure   \n",
       "\n",
       "                                            purturbation  \\\n",
       "0      cultured choroidal GLI1+ cells treated with DM...   \n",
       "1      Cultured choroidal GLI1+ cells treated with DM...   \n",
       "2                                          not specified   \n",
       "3                                                 unsure   \n",
       "4                                                 unsure   \n",
       "...                                                  ...   \n",
       "40841                                               None   \n",
       "40842                                               None   \n",
       "40843                                               None   \n",
       "40844                                               None   \n",
       "40845  incubation substrates: xylan; sampling time: 4...   \n",
       "\n",
       "                                               cell_line czi_collection_id  \\\n",
       "0      B6129PF1/J, 3 month-old, RPE/choroid single cells              None   \n",
       "1      Strain background: C57BL/6J, Age: 3 month-old,...              None   \n",
       "2                                          not specified              None   \n",
       "3                                                 unsure              None   \n",
       "4                                                 unsure              None   \n",
       "...                                                  ...               ...   \n",
       "40841                                               None              None   \n",
       "40842                                               None              None   \n",
       "40843                                               None              None   \n",
       "40844                                               None              None   \n",
       "40845                               strain: delta_Tgace2              None   \n",
       "\n",
       "      czi_collection_name                                     notes  \\\n",
       "0                    None              Metadata obtained by SRAgent   \n",
       "1                    None              Metadata obtained by SRAgent   \n",
       "2                    None              Metadata obtained by SRAgent   \n",
       "3                    None              Metadata obtained by SRAgent   \n",
       "4                    None              Metadata obtained by SRAgent   \n",
       "...                   ...                                       ...   \n",
       "40841                None  New dataset found by Find-Datasets agent   \n",
       "40842                None  New dataset found by Find-Datasets agent   \n",
       "40843                None  New dataset found by Find-Datasets agent   \n",
       "40844                None  New dataset found by Find-Datasets agent   \n",
       "40845                None              Metadata obtained by SRAgent   \n",
       "\n",
       "                      created_at                 updated_at  \n",
       "0     2025-01-23 17:20:30.662485 2025-01-23 17:21:51.385908  \n",
       "1     2025-01-22 22:45:26.805249 2025-01-22 22:47:13.950104  \n",
       "2     2025-01-27 07:30:27.589695 2025-01-27 07:32:05.085304  \n",
       "3     2025-01-21 17:28:27.920468 2025-01-21 17:29:55.383231  \n",
       "4     2025-01-21 17:28:27.920468 2025-01-21 17:36:38.159210  \n",
       "...                          ...                        ...  \n",
       "40841 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40842 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40843 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40844 2025-01-16 13:20:31.752503 2025-01-16 13:20:31.752503  \n",
       "40845 2025-01-16 13:20:31.752503 2025-01-16 13:26:41.517420  \n",
       "\n",
       "[40846 rows x 19 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \\\n",
    "    .distinct() \\\n",
    "    .where(tbl.notes != \"Processed by Chris Carpenter\")\n",
    "srx_metadata = pd.read_sql(str(stmt), conn_prod)\n",
    "srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [database, entrez_id, srx_accession, is_illumina, is_single_cell, is_paired_end, lib_prep, tech_10x, cell_prep, organism, tissue, disease, purturbation, cell_line, czi_collection_id, czi_collection_name, notes, created_at, updated_at]\n",
       "Index: []"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicate srx_accession values?\n",
    "x = srx_metadata[srx_metadata.srx_accession.notnull()]\n",
    "x[x.duplicated(subset=\"srx_accession\", keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    40082.0\n",
       "mean         1.0\n",
       "std          0.0\n",
       "min          1.0\n",
       "25%          1.0\n",
       "50%          1.0\n",
       "75%          1.0\n",
       "max          1.0\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicate srx_accession values?\n",
    "x[\"srx_accession\"].value_counts().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    14630.0\n",
       "mean         1.0\n",
       "std          0.0\n",
       "min          1.0\n",
       "25%          1.0\n",
       "50%          1.0\n",
       "75%          1.0\n",
       "max          1.0\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicate srx_accession values?\n",
    "x[\"entrez_id\"].value_counts().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8434"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of \"10X_Genomics\" samples\n",
    "srx_metadata[srx_metadata[\"lib_prep\"] == \"10x_Genomics\"].shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STAR results table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>feature</th>\n",
       "      <th>estimated_number_of_cells</th>\n",
       "      <th>fraction_of_unique_reads_in_cells</th>\n",
       "      <th>mean_gene_per_cell</th>\n",
       "      <th>mean_umi_per_cell</th>\n",
       "      <th>mean_feature_per_cell</th>\n",
       "      <th>median_gene_per_cell</th>\n",
       "      <th>median_umi_per_cell</th>\n",
       "      <th>median_feature_per_cell</th>\n",
       "      <th>...</th>\n",
       "      <th>reads_with_valid_barcodes</th>\n",
       "      <th>sequencing_saturation</th>\n",
       "      <th>total_feature_detected</th>\n",
       "      <th>umis_in_cells</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_gene</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_genefull</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_genefull_ex50pas</th>\n",
       "      <th>unique_reads_in_cells_mapped_to_genefull_exonoverintron</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX10024831</td>\n",
       "      <td>gene_full</td>\n",
       "      <td>2292</td>\n",
       "      <td>0.911745</td>\n",
       "      <td>None</td>\n",
       "      <td>12557.0</td>\n",
       "      <td>2730.0</td>\n",
       "      <td>None</td>\n",
       "      <td>10513.0</td>\n",
       "      <td>2775.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983305</td>\n",
       "      <td>0.593299</td>\n",
       "      <td>21731.0</td>\n",
       "      <td>28780763</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71181072.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-13 15:27:29.589416</td>\n",
       "      <td>2025-01-14 20:06:25.557352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX10024831</td>\n",
       "      <td>gene_ex50</td>\n",
       "      <td>2290</td>\n",
       "      <td>0.912572</td>\n",
       "      <td>None</td>\n",
       "      <td>17603.0</td>\n",
       "      <td>2756.0</td>\n",
       "      <td>None</td>\n",
       "      <td>11784.0</td>\n",
       "      <td>2799.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983334</td>\n",
       "      <td>0.601129</td>\n",
       "      <td>21911.0</td>\n",
       "      <td>40311609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101599277.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-13 15:27:29.589416</td>\n",
       "      <td>2025-01-14 20:06:25.557352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERX10024831</td>\n",
       "      <td>gene_ex_int</td>\n",
       "      <td>2290</td>\n",
       "      <td>0.912561</td>\n",
       "      <td>None</td>\n",
       "      <td>17625.0</td>\n",
       "      <td>2767.0</td>\n",
       "      <td>None</td>\n",
       "      <td>11803.0</td>\n",
       "      <td>2809.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983322</td>\n",
       "      <td>0.601073</td>\n",
       "      <td>22001.0</td>\n",
       "      <td>40362108</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101712901.0</td>\n",
       "      <td>2025-01-13 15:27:29.589416</td>\n",
       "      <td>2025-01-14 20:06:25.557352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERX10024831</td>\n",
       "      <td>velocyto</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.984647</td>\n",
       "      <td>-inf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-13 15:27:29.589416</td>\n",
       "      <td>2025-01-14 20:06:25.557352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERX10086874</td>\n",
       "      <td>gene_full</td>\n",
       "      <td>2262</td>\n",
       "      <td>0.909667</td>\n",
       "      <td>None</td>\n",
       "      <td>4977.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>None</td>\n",
       "      <td>4615.0</td>\n",
       "      <td>2116.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.967224</td>\n",
       "      <td>0.887163</td>\n",
       "      <td>21453.0</td>\n",
       "      <td>11258108</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108541887.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-13 15:17:54.767872</td>\n",
       "      <td>2025-01-14 20:06:25.557352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15940</th>\n",
       "      <td>ERX11899886</td>\n",
       "      <td>gene</td>\n",
       "      <td>19391</td>\n",
       "      <td>0.903745</td>\n",
       "      <td>None</td>\n",
       "      <td>6698.0</td>\n",
       "      <td>1871.0</td>\n",
       "      <td>None</td>\n",
       "      <td>4492.0</td>\n",
       "      <td>1617.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.984381</td>\n",
       "      <td>0.761567</td>\n",
       "      <td>29240.0</td>\n",
       "      <td>129894014</td>\n",
       "      <td>576267904.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15941</th>\n",
       "      <td>ERX11899886</td>\n",
       "      <td>gene_ex50</td>\n",
       "      <td>20030</td>\n",
       "      <td>0.914242</td>\n",
       "      <td>None</td>\n",
       "      <td>7989.0</td>\n",
       "      <td>2432.0</td>\n",
       "      <td>None</td>\n",
       "      <td>5618.0</td>\n",
       "      <td>2163.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983904</td>\n",
       "      <td>0.765714</td>\n",
       "      <td>32603.0</td>\n",
       "      <td>160037142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>726711495.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15942</th>\n",
       "      <td>ERX11899886</td>\n",
       "      <td>gene_ex_int</td>\n",
       "      <td>20035</td>\n",
       "      <td>0.914298</td>\n",
       "      <td>None</td>\n",
       "      <td>8012.0</td>\n",
       "      <td>2447.0</td>\n",
       "      <td>None</td>\n",
       "      <td>5639.0</td>\n",
       "      <td>2178.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983833</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>32621.0</td>\n",
       "      <td>160539579</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>729034278.0</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15943</th>\n",
       "      <td>ERX11899886</td>\n",
       "      <td>gene_full</td>\n",
       "      <td>20016</td>\n",
       "      <td>0.914711</td>\n",
       "      <td>None</td>\n",
       "      <td>7865.0</td>\n",
       "      <td>2423.0</td>\n",
       "      <td>None</td>\n",
       "      <td>5565.0</td>\n",
       "      <td>2159.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983820</td>\n",
       "      <td>0.765057</td>\n",
       "      <td>31181.0</td>\n",
       "      <td>157427372</td>\n",
       "      <td>NaN</td>\n",
       "      <td>712975893.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15944</th>\n",
       "      <td>ERX11899886</td>\n",
       "      <td>velocyto</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.985624</td>\n",
       "      <td>-inf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "      <td>2025-01-21 17:33:06.489468</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15945 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample      feature  estimated_number_of_cells  \\\n",
       "0      ERX10024831    gene_full                       2292   \n",
       "1      ERX10024831    gene_ex50                       2290   \n",
       "2      ERX10024831  gene_ex_int                       2290   \n",
       "3      ERX10024831     velocyto                          0   \n",
       "4      ERX10086874    gene_full                       2262   \n",
       "...            ...          ...                        ...   \n",
       "15940  ERX11899886         gene                      19391   \n",
       "15941  ERX11899886    gene_ex50                      20030   \n",
       "15942  ERX11899886  gene_ex_int                      20035   \n",
       "15943  ERX11899886    gene_full                      20016   \n",
       "15944  ERX11899886     velocyto                          0   \n",
       "\n",
       "       fraction_of_unique_reads_in_cells mean_gene_per_cell  \\\n",
       "0                               0.911745               None   \n",
       "1                               0.912572               None   \n",
       "2                               0.912561               None   \n",
       "3                                    NaN               None   \n",
       "4                               0.909667               None   \n",
       "...                                  ...                ...   \n",
       "15940                           0.903745               None   \n",
       "15941                           0.914242               None   \n",
       "15942                           0.914298               None   \n",
       "15943                           0.914711               None   \n",
       "15944                                NaN               None   \n",
       "\n",
       "       mean_umi_per_cell  mean_feature_per_cell median_gene_per_cell  \\\n",
       "0                12557.0                 2730.0                 None   \n",
       "1                17603.0                 2756.0                 None   \n",
       "2                17625.0                 2767.0                 None   \n",
       "3                    NaN                    NaN                 None   \n",
       "4                 4977.0                 2018.0                 None   \n",
       "...                  ...                    ...                  ...   \n",
       "15940             6698.0                 1871.0                 None   \n",
       "15941             7989.0                 2432.0                 None   \n",
       "15942             8012.0                 2447.0                 None   \n",
       "15943             7865.0                 2423.0                 None   \n",
       "15944                NaN                    NaN                 None   \n",
       "\n",
       "       median_umi_per_cell  median_feature_per_cell  ...  \\\n",
       "0                  10513.0                   2775.0  ...   \n",
       "1                  11784.0                   2799.0  ...   \n",
       "2                  11803.0                   2809.0  ...   \n",
       "3                      NaN                      NaN  ...   \n",
       "4                   4615.0                   2116.0  ...   \n",
       "...                    ...                      ...  ...   \n",
       "15940               4492.0                   1617.0  ...   \n",
       "15941               5618.0                   2163.0  ...   \n",
       "15942               5639.0                   2178.0  ...   \n",
       "15943               5565.0                   2159.0  ...   \n",
       "15944                  NaN                      NaN  ...   \n",
       "\n",
       "       reads_with_valid_barcodes  sequencing_saturation  \\\n",
       "0                       0.983305               0.593299   \n",
       "1                       0.983334               0.601129   \n",
       "2                       0.983322               0.601073   \n",
       "3                       0.984647                   -inf   \n",
       "4                       0.967224               0.887163   \n",
       "...                          ...                    ...   \n",
       "15940                   0.984381               0.761567   \n",
       "15941                   0.983904               0.765714   \n",
       "15942                   0.983833               0.765700   \n",
       "15943                   0.983820               0.765057   \n",
       "15944                   0.985624                   -inf   \n",
       "\n",
       "       total_feature_detected  umis_in_cells  \\\n",
       "0                     21731.0       28780763   \n",
       "1                     21911.0       40311609   \n",
       "2                     22001.0       40362108   \n",
       "3                         NaN              0   \n",
       "4                     21453.0       11258108   \n",
       "...                       ...            ...   \n",
       "15940                 29240.0      129894014   \n",
       "15941                 32603.0      160037142   \n",
       "15942                 32621.0      160539579   \n",
       "15943                 31181.0      157427372   \n",
       "15944                     NaN              0   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_gene  \\\n",
       "0                                       NaN   \n",
       "1                                       NaN   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4                                       NaN   \n",
       "...                                     ...   \n",
       "15940                           576267904.0   \n",
       "15941                                   NaN   \n",
       "15942                                   NaN   \n",
       "15943                                   NaN   \n",
       "15944                                   NaN   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_genefull  \\\n",
       "0                                    71181072.0   \n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3                                           NaN   \n",
       "4                                   108541887.0   \n",
       "...                                         ...   \n",
       "15940                                       NaN   \n",
       "15941                                       NaN   \n",
       "15942                                       NaN   \n",
       "15943                               712975893.0   \n",
       "15944                                       NaN   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_genefull_ex50pas  \\\n",
       "0                                                   NaN   \n",
       "1                                           101599277.0   \n",
       "2                                                   NaN   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "15940                                               NaN   \n",
       "15941                                       726711495.0   \n",
       "15942                                               NaN   \n",
       "15943                                               NaN   \n",
       "15944                                               NaN   \n",
       "\n",
       "       unique_reads_in_cells_mapped_to_genefull_exonoverintron  \\\n",
       "0                                                    NaN         \n",
       "1                                                    NaN         \n",
       "2                                            101712901.0         \n",
       "3                                                    NaN         \n",
       "4                                                    NaN         \n",
       "...                                                  ...         \n",
       "15940                                                NaN         \n",
       "15941                                                NaN         \n",
       "15942                                        729034278.0         \n",
       "15943                                                NaN         \n",
       "15944                                                NaN         \n",
       "\n",
       "                      created_at                 updated_at  \n",
       "0     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "1     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "2     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "3     2025-01-13 15:27:29.589416 2025-01-14 20:06:25.557352  \n",
       "4     2025-01-13 15:17:54.767872 2025-01-14 20:06:25.557352  \n",
       "...                          ...                        ...  \n",
       "15940 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15941 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15942 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15943 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "15944 2025-01-21 17:33:06.489468 2025-01-21 17:33:06.489468  \n",
       "\n",
       "[15945 rows x 36 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"screcounter_star_results\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \n",
    "star_results = pd.read_sql(str(stmt), conn_prod)\n",
    "star_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3189.0\n",
       "mean        5.0\n",
       "std         0.0\n",
       "min         5.0\n",
       "25%         5.0\n",
       "50%         5.0\n",
       "75%         5.0\n",
       "max         5.0\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "star_results[\"sample\"].value_counts().describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameters table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>star_index</th>\n",
       "      <th>cell_barcode_length</th>\n",
       "      <th>umi_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample, barcodes, star_index, cell_barcode_length, umi_length, strand, created_at, updated_at]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"screcounter_star_params\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(\"*\") \\\n",
    "    .where(tbl.star_index == \"\")\n",
    "params = pd.read_sql(str(stmt), conn_prod)\n",
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX22822724</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX20516510</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX26710136</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX24387545</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>ERX10668489</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>SRX19719137</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>SRX22006554</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>SRX20274306</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>ERX10991228</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>647 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    srx_accession organism\n",
       "0     SRX22822724    mouse\n",
       "1            None     None\n",
       "2     SRX20516510    mouse\n",
       "3     SRX26710136    mouse\n",
       "4     SRX24387545    human\n",
       "..            ...      ...\n",
       "642   ERX10668489    mouse\n",
       "643   SRX19719137    human\n",
       "644   SRX22006554    human\n",
       "645   SRX20274306    mouse\n",
       "646   ERX10991228    mouse\n",
       "\n",
       "[647 rows x 2 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(tbl.srx_accession, tbl.organism) \\\n",
    "    .distinct() \\\n",
    "    .where(tbl.notes != \"Processed by Chris Carpenter\")\n",
    "srx_metadata = pd.read_sql(str(stmt), conn_prod)\n",
    "srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>star_index</th>\n",
       "      <th>cell_barcode_length</th>\n",
       "      <th>umi_length</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX11662340</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX19392127</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX24158623</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX22159680</td>\n",
       "      <td>737K-august-2016.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>Reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX24228136</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRX24387553</td>\n",
       "      <td>737K-august-2016.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>Reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRX24461697</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRX24523046</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRX24585614</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRX25852474</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_hg38</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRX21638105</td>\n",
       "      <td>3M-february-2018.txt</td>\n",
       "      <td>star_refData_2020_mm10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>Forward</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample              barcodes              star_index  \\\n",
       "0   ERX11662340  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "1   SRX19392127  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "2   SRX24158623  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "3   SRX22159680  737K-august-2016.txt  star_refData_2020_hg38   \n",
       "4   SRX24228136  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "5   SRX24387553  737K-august-2016.txt  star_refData_2020_hg38   \n",
       "6   SRX24461697  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "7   SRX24523046  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "8   SRX24585614  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "9   SRX25852474  3M-february-2018.txt  star_refData_2020_hg38   \n",
       "10  SRX21638105  3M-february-2018.txt  star_refData_2020_mm10   \n",
       "\n",
       "    cell_barcode_length  umi_length   strand  \n",
       "0                    16          12  Forward  \n",
       "1                    16          12  Forward  \n",
       "2                    16          12  Forward  \n",
       "3                    16          10  Reverse  \n",
       "4                    16          12  Forward  \n",
       "5                    16          10  Reverse  \n",
       "6                    16          12  Forward  \n",
       "7                    16          12  Forward  \n",
       "8                    16          12  Forward  \n",
       "9                    16          12  Forward  \n",
       "10                   16          12  Forward  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge by srx_accession\n",
    "def set_star_ref(row):\n",
    "    if row.organism == \"human\":\n",
    "        return \"star_refData_2020_hg38\"\n",
    "    elif row.organism == \"mouse\":\n",
    "        return \"star_refData_2020_mm10\"\n",
    "    else:\n",
    "        raise ValueError(f\"organism {row.organism} not supported\")\n",
    "\n",
    "params = params.merge(srx_metadata, left_on=\"sample\", right_on=\"srx_accession\") \n",
    "params[\"star_index\"] = params.apply(lambda x: set_star_ref(x), axis=1)\n",
    "params = params.drop(columns=[\"created_at\", \"updated_at\", \"srx_accession\", \"organism\"], axis=1)\n",
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# upsert\n",
    "db_upsert(params, \"screcounter_star_params\", conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# --OLD--"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out records with no SRX accessions\n",
    "\n",
    "`New dataset found by Find-Datasets agent`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entrez_ids = srx_metadata[srx_metadata[\"srx_accession\"].isna()][\"entrez_id\"].unique().tolist()\n",
    "entrez_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete extrez_ids from srx_metadata\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .delete() \\\n",
    "    .where(tbl.entrez_id.isin(entrez_ids))\n",
    "with conn_prod.cursor() as cur:\n",
    "    cur.execute(str(stmt))\n",
    "    conn_prod.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete from log\n",
    "tbl = Table(\"screcounter_log\")\n",
    "to_rm = [\"ERX11146221\"]\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .delete() \\\n",
    "    .where(tbl.sample.isin(to_rm))\n",
    "with conn_prod.cursor() as cur:\n",
    "    cur.execute(str(stmt))\n",
    "    conn_prod.commit()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
