{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Migrate data from test to production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.upsert import db_upsert\n",
    "from SRAgent.db.utils import db_list_tables, db_glimpse_tables, db_get_table, execute_query\n",
    "from SRAgent.db.get import db_find_srx\n",
    "from SRAgent.db.create import create_table, create_table_router"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "srx_srr\n",
      "screcounter_star\n",
      "eval\n",
      "screcounter_log\n",
      "srx_metadata\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in test\n",
    "os.environ['DYNACONF'] = 'test'\n",
    "conn_test = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_test)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX26727599\tSRR31350667\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23538581\tSRR27876733\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23261451\tSRR27592690\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23261451\tSRR27592688\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23261451\tSRR27592689\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "SRX19729045\tSRR23917610\tgene\t139\t1.0\t\t4.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.171386\t556\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tgene_ex50\t197\t1.0\t\t3.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.149675\t784\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tgene_ex_int\t198\t1.0\t\t3.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.1507\t789\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tgene_full\t188\t1.0\t\t4.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.147486\t763\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tvelocyto\t0\t\t\t\t\t\t\t\t100000\t0\t-inf\t0\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "ERX10987225\tERR11583756\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583807\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583914\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583981\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583986\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tnotes\tcreated_at\tupdated_at\n",
      "sra\t33249542\tSRX24914804\tyes\tyes\tyes\tSmart-seq2\tnot_applicable\tsingle_cell\thuman\tblood\tHepatitis C Virus (HCV) reinfection outcomes\tAkebia Saponin D (ASD) and TNF-induced optic nerve damage\tMemory B cells\tMetadata obtained by SRAgent\t2024-12-18 18:23:36.181001\t2024-12-18 18:25:14.322667\n",
      "sra\t34603139\tSRX25602353\tyes\tyes\tyes\t10x_Genomics\tother\tsingle_cell\thuman\tPBMCs\tunsure\teffects of salt intake in healthy humans\tunsure\tMetadata obtained by SRAgent\t2024-12-18 18:23:36.181001\t2024-12-18 18:25:33.937056\n",
      "sra\t30864944\t\t\t\t\t\t\t\t\t\t\t\t\tNew dataset found by Find-Datasets agent\t2025-01-03 17:29:09.649591\t2025-01-03 17:29:09.649591\n",
      "sra\t34602541\t\t\t\t\t\t\t\t\t\t\t\t\tNew dataset found by Find-Datasets agent\t2025-01-03 17:29:09.649591\t2025-01-03 17:29:09.649591\n",
      "sra\t31225478\tSRX23114342\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tsingle_cell\tmouse\tsciatic nerve\tspontaneous autoimmune neuropathy (CIDP)\tunsure\timmune cells (CD45+)\tMetadata obtained by SRAgent\t2024-12-18 18:40:50.574099\t2024-12-18 18:42:11.558669\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eval\n",
      "screcounter_log\n",
      "screcounter_star\n",
      "srx_srr\n",
      "srx_metadata\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in prod\n",
    "os.environ['DYNACONF'] = 'prod'\n",
    "conn_prod = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_prod)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SRX Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chris' processed records\n",
    "\n",
    "* Add to production database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "      <th>experiment</th>\n",
       "      <th>CZI_collection_name</th>\n",
       "      <th>CZI_collection_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110064</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>73f82ac8-15cc-4fcd-87f8-5683723fce7f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>19007785</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488012</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "      <td>ed9185e3-5b82-40c7-9824-b2141590c7f0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270435</td>\n",
       "      <td>ERX8791959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX8791959</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sra</td>\n",
       "      <td>34493113</td>\n",
       "      <td>ERX12558784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX12558784</td>\n",
       "      <td>Early human lung immune cell development and i...</td>\n",
       "      <td>ec329aed-22bc-4d6e-8935-8282dcb1acac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52899</th>\n",
       "      <td>sra</td>\n",
       "      <td>19008576</td>\n",
       "      <td>SRX13671360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX13671360</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52900</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270500</td>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52901</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110088</td>\n",
       "      <td>ERX11148805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX11148805</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>73f82ac8-15cc-4fcd-87f8-5683723fce7f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52902</th>\n",
       "      <td>sra</td>\n",
       "      <td>10884261</td>\n",
       "      <td>ERX4126865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX4126865</td>\n",
       "      <td>scRNA-seq assessment of the human lung, spleen...</td>\n",
       "      <td>4d74781b-8186-4c9a-b659-ff4dc4601d91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52903</th>\n",
       "      <td>sra</td>\n",
       "      <td>12773490</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>3'v3</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "      <td>71f4bccf-53d4-4c12-9e80-e73bfb89e398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6060 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession  is_illumina  is_single_cell  \\\n",
       "0          sra   29110064   ERX11148781          NaN             NaN   \n",
       "1          sra   19007785   SRX13670569          NaN             NaN   \n",
       "2          sra   12488012    SRX9556597          NaN             NaN   \n",
       "14         sra   21270435    ERX8791959          NaN             NaN   \n",
       "15         sra   34493113   ERX12558784          NaN             NaN   \n",
       "...        ...        ...           ...          ...             ...   \n",
       "52899      sra   19008576   SRX13671360          NaN             NaN   \n",
       "52900      sra   21270500    ERX8792024          NaN             NaN   \n",
       "52901      sra   29110088   ERX11148805          NaN             NaN   \n",
       "52902      sra   10884261    ERX4126865          NaN             NaN   \n",
       "52903      sra   12773490    SRX9770752          NaN             NaN   \n",
       "\n",
       "       is_paired_end        lib_prep tech_10x    cell_prep  organism  tissue  \\\n",
       "0                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "1                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "2                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "14               NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "15               NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "...              ...             ...      ...          ...       ...     ...   \n",
       "52899            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "52900            NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "52901            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "52902            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "52903            NaN  \\t10x_Genomics     3'v3  single_cell       NaN     NaN   \n",
       "\n",
       "       disease  purturbation  cell_line                         notes  \\\n",
       "0          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "1          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "2          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "14         NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "15         NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "...        ...           ...        ...                           ...   \n",
       "52899      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52900      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52901      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52902      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52903      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "\n",
       "      created_at updated_at   experiment  \\\n",
       "0        31:16.4    31:35.1  ERX11148781   \n",
       "1        31:16.4    31:35.1  SRX13670569   \n",
       "2        31:16.4    31:35.1   SRX9556597   \n",
       "14       31:16.4    31:35.1   ERX8791959   \n",
       "15       31:16.4    31:35.1  ERX12558784   \n",
       "...          ...        ...          ...   \n",
       "52899    31:16.4    31:35.1  SRX13671360   \n",
       "52900    31:16.4    31:35.1   ERX8792024   \n",
       "52901    31:16.4    31:35.1  ERX11148805   \n",
       "52902    31:16.4    31:35.1   ERX4126865   \n",
       "52903    31:16.4    31:35.1   SRX9770752   \n",
       "\n",
       "                                     CZI_collection_name  \\\n",
       "0      Developmental cell programs are co-opted in in...   \n",
       "1      Neuron type-specific effects of human aging an...   \n",
       "2      Time-resolved Systems Immunology Reveals a Lat...   \n",
       "14     Cross-tissue immune cell analysis reveals tiss...   \n",
       "15     Early human lung immune cell development and i...   \n",
       "...                                                  ...   \n",
       "52899  Neuron type-specific effects of human aging an...   \n",
       "52900  Cross-tissue immune cell analysis reveals tiss...   \n",
       "52901  Developmental cell programs are co-opted in in...   \n",
       "52902  scRNA-seq assessment of the human lung, spleen...   \n",
       "52903  An Integrated Single Cell Atlas of Human Perio...   \n",
       "\n",
       "                          CZI_collection_id  \n",
       "0      73f82ac8-15cc-4fcd-87f8-5683723fce7f  \n",
       "1      91c8e321-566f-4f9d-b89e-3a164be654d5  \n",
       "2      ed9185e3-5b82-40c7-9824-b2141590c7f0  \n",
       "14     62ef75e4-cbea-454e-a0ce-998ec40223d3  \n",
       "15     ec329aed-22bc-4d6e-8935-8282dcb1acac  \n",
       "...                                     ...  \n",
       "52899  91c8e321-566f-4f9d-b89e-3a164be654d5  \n",
       "52900  62ef75e4-cbea-454e-a0ce-998ec40223d3  \n",
       "52901  73f82ac8-15cc-4fcd-87f8-5683723fce7f  \n",
       "52902  4d74781b-8186-4c9a-b659-ff4dc4601d91  \n",
       "52903  71f4bccf-53d4-4c12-9e80-e73bfb89e398  \n",
       "\n",
       "[6060 rows x 20 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "infile = \"../data/20241227_ChrisC_metadata_v1.tsv\"\n",
    "chris_cur_meta = pd.read_csv(infile, sep=\"\\t\").drop_duplicates()\n",
    "chris_cur_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the columns that are not in the database\n",
    "chris_cur_meta.drop([\"experiment\", \"created_at\", \"updated_at\"], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>notes</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>czi_collection_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110064</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>73f82ac8-15cc-4fcd-87f8-5683723fce7f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>19007785</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488012</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "      <td>ed9185e3-5b82-40c7-9824-b2141590c7f0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270435</td>\n",
       "      <td>ERX8791959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sra</td>\n",
       "      <td>34493113</td>\n",
       "      <td>ERX12558784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Early human lung immune cell development and i...</td>\n",
       "      <td>ec329aed-22bc-4d6e-8935-8282dcb1acac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52899</th>\n",
       "      <td>sra</td>\n",
       "      <td>19008576</td>\n",
       "      <td>SRX13671360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52900</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270500</td>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52901</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110088</td>\n",
       "      <td>ERX11148805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>73f82ac8-15cc-4fcd-87f8-5683723fce7f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52902</th>\n",
       "      <td>sra</td>\n",
       "      <td>10884261</td>\n",
       "      <td>ERX4126865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>scRNA-seq assessment of the human lung, spleen...</td>\n",
       "      <td>4d74781b-8186-4c9a-b659-ff4dc4601d91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52903</th>\n",
       "      <td>sra</td>\n",
       "      <td>12773490</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>3'v3</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "      <td>71f4bccf-53d4-4c12-9e80-e73bfb89e398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6060 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession  is_illumina  is_single_cell  \\\n",
       "0          sra   29110064   ERX11148781          NaN             NaN   \n",
       "1          sra   19007785   SRX13670569          NaN             NaN   \n",
       "2          sra   12488012    SRX9556597          NaN             NaN   \n",
       "14         sra   21270435    ERX8791959          NaN             NaN   \n",
       "15         sra   34493113   ERX12558784          NaN             NaN   \n",
       "...        ...        ...           ...          ...             ...   \n",
       "52899      sra   19008576   SRX13671360          NaN             NaN   \n",
       "52900      sra   21270500    ERX8792024          NaN             NaN   \n",
       "52901      sra   29110088   ERX11148805          NaN             NaN   \n",
       "52902      sra   10884261    ERX4126865          NaN             NaN   \n",
       "52903      sra   12773490    SRX9770752          NaN             NaN   \n",
       "\n",
       "       is_paired_end        lib_prep tech_10x    cell_prep  organism  tissue  \\\n",
       "0                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "1                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "2                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "14               NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "15               NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "...              ...             ...      ...          ...       ...     ...   \n",
       "52899            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "52900            NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "52901            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "52902            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "52903            NaN  \\t10x_Genomics     3'v3  single_cell       NaN     NaN   \n",
       "\n",
       "       disease  purturbation  cell_line                         notes  \\\n",
       "0          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "1          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "2          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "14         NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "15         NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "...        ...           ...        ...                           ...   \n",
       "52899      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52900      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52901      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52902      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "52903      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "\n",
       "                                     czi_collection_name  \\\n",
       "0      Developmental cell programs are co-opted in in...   \n",
       "1      Neuron type-specific effects of human aging an...   \n",
       "2      Time-resolved Systems Immunology Reveals a Lat...   \n",
       "14     Cross-tissue immune cell analysis reveals tiss...   \n",
       "15     Early human lung immune cell development and i...   \n",
       "...                                                  ...   \n",
       "52899  Neuron type-specific effects of human aging an...   \n",
       "52900  Cross-tissue immune cell analysis reveals tiss...   \n",
       "52901  Developmental cell programs are co-opted in in...   \n",
       "52902  scRNA-seq assessment of the human lung, spleen...   \n",
       "52903  An Integrated Single Cell Atlas of Human Perio...   \n",
       "\n",
       "                          czi_collection_id  \n",
       "0      73f82ac8-15cc-4fcd-87f8-5683723fce7f  \n",
       "1      91c8e321-566f-4f9d-b89e-3a164be654d5  \n",
       "2      ed9185e3-5b82-40c7-9824-b2141590c7f0  \n",
       "14     62ef75e4-cbea-454e-a0ce-998ec40223d3  \n",
       "15     ec329aed-22bc-4d6e-8935-8282dcb1acac  \n",
       "...                                     ...  \n",
       "52899  91c8e321-566f-4f9d-b89e-3a164be654d5  \n",
       "52900  62ef75e4-cbea-454e-a0ce-998ec40223d3  \n",
       "52901  73f82ac8-15cc-4fcd-87f8-5683723fce7f  \n",
       "52902  4d74781b-8186-4c9a-b659-ff4dc4601d91  \n",
       "52903  71f4bccf-53d4-4c12-9e80-e73bfb89e398  \n",
       "\n",
       "[6060 rows x 17 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lower case column names\n",
    "chris_cur_meta.columns = chris_cur_meta.columns.str.lower()\n",
    "chris_cur_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Format metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "## count values in each character column\n",
    "for col in [\"is_illumina\", \"is_single_cell\", \"is_paired_end\"]:\n",
    "    print(chris_cur_meta[col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format tech_10x\n",
    "idx = {\n",
    "    \"5'\": \"5_prime_gex\",\n",
    "    \"3'v3\": \"3_prime_gex\",\n",
    "    \"3'v2\": \"3_prime_gex\",\n",
    "    \"multiome GEX\": \"multiome\",\n",
    "}\n",
    "for k,v in idx.items():\n",
    "    chris_cur_meta[\"tech_10x\"] = chris_cur_meta[\"tech_10x\"].replace(k, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format lib_prep\n",
    "idx = {\n",
    "    \"\\t10x_Genomics\": \"10x_Genomics\",\n",
    "}\n",
    "for k,v in idx.items():\n",
    "    chris_cur_meta[\"lib_prep\"] = chris_cur_meta[\"lib_prep\"].replace(k, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lib_prep\n",
      "10x_Genomics    2223\n",
      "Name: count, dtype: int64\n",
      "tech_10x\n",
      "3_prime_gex    1179\n",
      "5_prime_gex     873\n",
      "multiome        122\n",
      "Name: count, dtype: int64\n",
      "cell_prep\n",
      "single_cell    2223\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## count values in each character column\n",
    "for col in [\"lib_prep\", \"tech_10x\", \"cell_prep\"]:\n",
    "    print(chris_cur_meta[col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "## count values in each character column\n",
    "for col in [\"organism\", \"tissue\", \"disease\", \"purturbation\", \"cell_line\"]:\n",
    "    print(chris_cur_meta[col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to file\n",
    "outfile = os.path.join(\"../data/\", \"20241227_ChrisC_metadata_v2.tsv\")\n",
    "chris_cur_meta.to_csv(outfile, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upserted ../data/20241227_ChrisC_metadata_v2.tsv into srx_metadata\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "../scripts/db-tools.py --upsert-csv ../data/20241227_ChrisC_metadata_v2.tsv --upsert-target srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t29110064\tERX11148781\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t73f82ac8-15cc-4fcd-87f8-5683723fce7f\tDevelopmental cell programs are co-opted in inflammatory skin disease\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t19007785\tSRX13670569\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t91c8e321-566f-4f9d-b89e-3a164be654d5\tNeuron type-specific effects of human aging and sex on DNA methylation and transcription\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t12488012\tSRX9556597\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\ted9185e3-5b82-40c7-9824-b2141590c7f0\tTime-resolved Systems Immunology Reveals a Late Juncture Linked to Fatal COVID-19\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t21270435\tERX8791959\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t62ef75e4-cbea-454e-a0ce-998ec40223d3\tCross-tissue immune cell analysis reveals tissue-specific features in humans\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t34493113\tERX12558784\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\tec329aed-22bc-4d6e-8935-8282dcb1acac\tEarly human lung immune cell development and its role in epithelial cell fate\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "../scripts/db-tools.py --glimpse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>29115023</td>\n",
       "      <td>ERX11157721</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>36178506</td>\n",
       "      <td>ERX11887200</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>5920965</td>\n",
       "      <td>ERX2188110</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>macaque</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>34439895</td>\n",
       "      <td>ERX9692805</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>chicken</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eval1</td>\n",
       "      <td>gds</td>\n",
       "      <td>200271457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>eval1</td>\n",
       "      <td>gds</td>\n",
       "      <td>200281884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>18060880</td>\n",
       "      <td>SRX13201194</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>30749595</td>\n",
       "      <td>SRX22716300</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>31054998</td>\n",
       "      <td>SRX22985325</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>31679394</td>\n",
       "      <td>SRX23448232</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>32664033</td>\n",
       "      <td>SRX24349377</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>macaque</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>33176418</td>\n",
       "      <td>SRX24842084</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>35087715</td>\n",
       "      <td>SRX25994842</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>35200088</td>\n",
       "      <td>SRX26085693</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>chicken</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>eval1</td>\n",
       "      <td>gds</td>\n",
       "      <td>200277303</td>\n",
       "      <td>SRX26099173</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>35536066</td>\n",
       "      <td>SRX26313156</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>35537624</td>\n",
       "      <td>SRX26314714</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>35979902</td>\n",
       "      <td>SRX26636208</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>36106630</td>\n",
       "      <td>SRX26727599</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>36209780</td>\n",
       "      <td>SRX26798077</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>eval1</td>\n",
       "      <td>sra</td>\n",
       "      <td>36280358</td>\n",
       "      <td>SRX26854925</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>2024-12-18 17:38:54.774086</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>27454880</td>\n",
       "      <td>ERX10668381</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>27454942</td>\n",
       "      <td>ERX10668443</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>34046074</td>\n",
       "      <td>ERX10987225</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>29108810</td>\n",
       "      <td>ERX11146221</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>36096868</td>\n",
       "      <td>ERX11740068</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>27238143</td>\n",
       "      <td>SRX19867681</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_nucleus</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>29740817</td>\n",
       "      <td>SRX21870188</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>29914843</td>\n",
       "      <td>SRX22001630</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>30464257</td>\n",
       "      <td>SRX22478468</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>27694423</td>\n",
       "      <td>SRX20273490</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>plexWell</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>27694586</td>\n",
       "      <td>SRX20273653</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>plexWell</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>30774926</td>\n",
       "      <td>SRX22738634</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>30775044</td>\n",
       "      <td>SRX22738752</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>30864717</td>\n",
       "      <td>SRX22822772</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>30865012</td>\n",
       "      <td>SRX22823067</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>32927456</td>\n",
       "      <td>SRX24608694</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>32927626</td>\n",
       "      <td>SRX24608864</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>mouse</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>34602376</td>\n",
       "      <td>SRX25601590</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>eval2</td>\n",
       "      <td>sra</td>\n",
       "      <td>35200088</td>\n",
       "      <td>SRX26085693</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>chicken</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "      <td>2025-01-02 23:08:54.386892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dataset_id database  entrez_id srx_accession is_illumina is_single_cell  \\\n",
       "0       eval1      sra   29115023   ERX11157721         yes            yes   \n",
       "1       eval1      sra   36178506   ERX11887200         yes            yes   \n",
       "2       eval1      sra    5920965    ERX2188110         yes            yes   \n",
       "3       eval1      sra   34439895    ERX9692805         yes            yes   \n",
       "4       eval1      gds  200271457           NaN          no             no   \n",
       "5       eval1      gds  200281884           NaN          no             no   \n",
       "6       eval1      sra   18060880   SRX13201194         yes             no   \n",
       "7       eval1      sra   30749595   SRX22716300         yes            yes   \n",
       "8       eval1      sra   31054998   SRX22985325         yes             no   \n",
       "9       eval1      sra   31679394   SRX23448232         yes             no   \n",
       "10      eval1      sra   32664033   SRX24349377         yes            yes   \n",
       "11      eval1      sra   33176418   SRX24842084         yes            yes   \n",
       "12      eval1      sra   35087715   SRX25994842         yes            yes   \n",
       "13      eval1      sra   35200088   SRX26085693         yes            yes   \n",
       "14      eval1      gds  200277303   SRX26099173         yes             no   \n",
       "15      eval1      sra   35536066   SRX26313156         yes             no   \n",
       "16      eval1      sra   35537624   SRX26314714          no            yes   \n",
       "17      eval1      sra   35979902   SRX26636208         yes             no   \n",
       "18      eval1      sra   36106630   SRX26727599         yes             no   \n",
       "19      eval1      sra   36209780   SRX26798077         yes             no   \n",
       "20      eval1      sra   36280358   SRX26854925         yes             no   \n",
       "21      eval2      sra   27454880   ERX10668381         yes            yes   \n",
       "22      eval2      sra   27454942   ERX10668443         yes            yes   \n",
       "23      eval2      sra   34046074   ERX10987225         yes            yes   \n",
       "24      eval2      sra   29108810   ERX11146221         yes            yes   \n",
       "25      eval2      sra   36096868   ERX11740068         yes            yes   \n",
       "26      eval2      sra   27238143   SRX19867681         yes            yes   \n",
       "27      eval2      sra   29740817   SRX21870188         yes            yes   \n",
       "28      eval2      sra   29914843   SRX22001630         yes            yes   \n",
       "29      eval2      sra   30464257   SRX22478468         yes            yes   \n",
       "30      eval2      sra   27694423   SRX20273490         yes            yes   \n",
       "31      eval2      sra   27694586   SRX20273653         yes            yes   \n",
       "32      eval2      sra   30774926   SRX22738634         yes            yes   \n",
       "33      eval2      sra   30775044   SRX22738752         yes            yes   \n",
       "34      eval2      sra   30864717   SRX22822772         yes            yes   \n",
       "35      eval2      sra   30865012   SRX22823067         yes            yes   \n",
       "36      eval2      sra   32927456   SRX24608694         yes            yes   \n",
       "37      eval2      sra   32927626   SRX24608864         yes            yes   \n",
       "38      eval2      sra   34602376  SRX25601590          yes            yes   \n",
       "39      eval2      sra   35200088   SRX26085693         yes            yes   \n",
       "\n",
       "   is_paired_end        lib_prep        tech_10x    organism       cell_prep  \\\n",
       "0            yes    10x_Genomics     3_prime_gex   zebrafish     single_cell   \n",
       "1            yes    10x_Genomics     3_prime_gex       mouse     single_cell   \n",
       "2            yes           other  not_applicable     macaque     single_cell   \n",
       "3            yes    10x_Genomics     3_prime_gex     chicken     single_cell   \n",
       "4             no  not_applicable  not_applicable       human  not_applicable   \n",
       "5             no  not_applicable  not_applicable       human  not_applicable   \n",
       "6             no  not_applicable  not_applicable       mouse  not_applicable   \n",
       "7            yes    10x_Genomics     3_prime_gex       human     single_cell   \n",
       "8            yes  not_applicable  not_applicable       human  not_applicable   \n",
       "9            yes  not_applicable  not_applicable  metagenome  not_applicable   \n",
       "10           yes    10x_Genomics     5_prime_gex     macaque     single_cell   \n",
       "11           yes           other  not_applicable   zebrafish     single_cell   \n",
       "12           yes    10x_Genomics     3_prime_gex       human     single_cell   \n",
       "13           yes           other  not_applicable     chicken     single_cell   \n",
       "14            no  not_applicable  not_applicable       mouse  not_applicable   \n",
       "15           yes  not_applicable  not_applicable       human  not_applicable   \n",
       "16            no           other  not_applicable       human     single_cell   \n",
       "17           yes  not_applicable  not_applicable       human  not_applicable   \n",
       "18           yes  not_applicable  not_applicable       mouse  not_applicable   \n",
       "19           yes  not_applicable  not_applicable       mouse  not_applicable   \n",
       "20           yes  not_applicable  not_applicable  metagenome  not_applicable   \n",
       "21           yes      Smart-seq2  not_applicable       mouse     single_cell   \n",
       "22           yes      Smart-seq2  not_applicable       mouse     single_cell   \n",
       "23           yes    10x_Genomics     3_prime_gex       human     single_cell   \n",
       "24           yes    10x_Genomics     3_prime_gex       human     single_cell   \n",
       "25           yes    10x_Genomics     3_prime_gex       mouse     single_cell   \n",
       "26           yes    10x_Genomics     3_prime_gex       human  single_nucleus   \n",
       "27           yes    10x_Genomics     5_prime_gex       human     single_cell   \n",
       "28           yes    10x_Genomics     5_prime_gex       human     single_cell   \n",
       "29           yes    10x_Genomics     5_prime_gex       mouse     single_cell   \n",
       "30           yes        plexWell  not_applicable       human     single_cell   \n",
       "31           yes        plexWell  not_applicable       human     single_cell   \n",
       "32           yes        MARS-seq  not_applicable       mouse     single_cell   \n",
       "33           yes        MARS-seq  not_applicable       mouse     single_cell   \n",
       "34           yes        MARS-seq  not_applicable       mouse     single_cell   \n",
       "35           yes        MARS-seq  not_applicable       mouse     single_cell   \n",
       "36           yes    10x_Genomics     3_prime_gex       mouse     single_cell   \n",
       "37           yes    10x_Genomics     3_prime_gex       mouse     single_cell   \n",
       "38           yes           other  not_applicable       human     single_cell   \n",
       "39           yes           other  not_applicable     chicken     single_cell   \n",
       "\n",
       "                   created_at                 updated_at  \n",
       "0  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "1  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "2  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "3  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "4  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "5  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "6  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "7  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "8  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "9  2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "10 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "11 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "12 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "13 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "14 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "15 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "16 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "17 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "18 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "19 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "20 2024-12-18 17:38:54.774086 2025-01-02 23:08:54.386892  \n",
       "21 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "22 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "23 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "24 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "25 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "26 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "27 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "28 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "29 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "30 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "31 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "32 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "33 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "34 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "35 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "36 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "37 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "38 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  \n",
       "39 2025-01-02 23:08:54.386892 2025-01-02 23:08:54.386892  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eval_test = db_get_table( \"eval\", conn_test)\n",
    "eval_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop created_at and updated_at\n",
    "eval_test.drop([\"created_at\", \"updated_at\"], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# upload to prod\n",
    "db_upsert(eval_test, \"eval\", conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check production database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX25926423\tSRR30502613\t2025-01-03 18:43:38.627686\t2025-01-03 18:43:38.627686\n",
      "SRX25926423\tSRR30502600\t2025-01-03 18:43:38.627686\t2025-01-03 18:43:38.627686\n",
      "SRX24914484\tSRR29400843\t2025-01-03 18:43:49.275138\t2025-01-03 18:43:49.275138\n",
      "SRX22006379\tSRR26297925\t2025-01-03 18:43:57.864762\t2025-01-03 18:43:57.864762\n",
      "SRX21883929\tSRR26171850\t2025-01-03 18:44:08.340937\t2025-01-03 18:44:08.340937\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t29110064\tERX11148781\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t73f82ac8-15cc-4fcd-87f8-5683723fce7f\tDevelopmental cell programs are co-opted in inflammatory skin disease\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t19007785\tSRX13670569\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t91c8e321-566f-4f9d-b89e-3a164be654d5\tNeuron type-specific effects of human aging and sex on DNA methylation and transcription\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t12488012\tSRX9556597\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\ted9185e3-5b82-40c7-9824-b2141590c7f0\tTime-resolved Systems Immunology Reveals a Late Juncture Linked to Fatal COVID-19\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t21270435\tERX8791959\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t62ef75e4-cbea-454e-a0ce-998ec40223d3\tCross-tissue immune cell analysis reveals tissue-specific features in humans\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "sra\t34493113\tERX12558784\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\tec329aed-22bc-4d6e-8935-8282dcb1acac\tEarly human lung immune cell development and its role in epithelial cell fate\tProcessed by Chris Carpenter\t2025-01-03 18:39:40.221883\t2025-01-03 18:39:40.221883\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110064</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73f82ac8-15cc-4fcd-87f8-5683723fce7f</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>19007785</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488012</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ed9185e3-5b82-40c7-9824-b2141590c7f0</td>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270435</td>\n",
       "      <td>ERX8791959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>34493113</td>\n",
       "      <td>ERX12558784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ec329aed-22bc-4d6e-8935-8282dcb1acac</td>\n",
       "      <td>Early human lung immune cell development and i...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6055</th>\n",
       "      <td>sra</td>\n",
       "      <td>19008576</td>\n",
       "      <td>SRX13671360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6056</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270500</td>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6057</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110088</td>\n",
       "      <td>ERX11148805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73f82ac8-15cc-4fcd-87f8-5683723fce7f</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6058</th>\n",
       "      <td>sra</td>\n",
       "      <td>10884261</td>\n",
       "      <td>ERX4126865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4d74781b-8186-4c9a-b659-ff4dc4601d91</td>\n",
       "      <td>scRNA-seq assessment of the human lung, spleen...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6059</th>\n",
       "      <td>sra</td>\n",
       "      <td>12773490</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71f4bccf-53d4-4c12-9e80-e73bfb89e398</td>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6060 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     database  entrez_id srx_accession is_illumina is_single_cell  \\\n",
       "0         sra   29110064   ERX11148781         NaN            NaN   \n",
       "1         sra   19007785   SRX13670569         NaN            NaN   \n",
       "2         sra   12488012    SRX9556597         NaN            NaN   \n",
       "3         sra   21270435    ERX8791959         NaN            NaN   \n",
       "4         sra   34493113   ERX12558784         NaN            NaN   \n",
       "...       ...        ...           ...         ...            ...   \n",
       "6055      sra   19008576   SRX13671360         NaN            NaN   \n",
       "6056      sra   21270500    ERX8792024         NaN            NaN   \n",
       "6057      sra   29110088   ERX11148805         NaN            NaN   \n",
       "6058      sra   10884261    ERX4126865         NaN            NaN   \n",
       "6059      sra   12773490    SRX9770752         NaN            NaN   \n",
       "\n",
       "     is_paired_end      lib_prep     tech_10x    cell_prep organism tissue  \\\n",
       "0              NaN  10x_Genomics  5_prime_gex  single_cell      NaN    NaN   \n",
       "1              NaN           NaN          NaN          NaN      NaN    NaN   \n",
       "2              NaN  10x_Genomics  5_prime_gex  single_cell      NaN    NaN   \n",
       "3              NaN  10x_Genomics  5_prime_gex  single_cell      NaN    NaN   \n",
       "4              NaN  10x_Genomics  5_prime_gex  single_cell      NaN    NaN   \n",
       "...            ...           ...          ...          ...      ...    ...   \n",
       "6055           NaN           NaN          NaN          NaN      NaN    NaN   \n",
       "6056           NaN  10x_Genomics  5_prime_gex  single_cell      NaN    NaN   \n",
       "6057           NaN           NaN          NaN          NaN      NaN    NaN   \n",
       "6058           NaN           NaN          NaN          NaN      NaN    NaN   \n",
       "6059           NaN  10x_Genomics  3_prime_gex  single_cell      NaN    NaN   \n",
       "\n",
       "     disease purturbation cell_line                     czi_collection_id  \\\n",
       "0        NaN          NaN       NaN  73f82ac8-15cc-4fcd-87f8-5683723fce7f   \n",
       "1        NaN          NaN       NaN  91c8e321-566f-4f9d-b89e-3a164be654d5   \n",
       "2        NaN          NaN       NaN  ed9185e3-5b82-40c7-9824-b2141590c7f0   \n",
       "3        NaN          NaN       NaN  62ef75e4-cbea-454e-a0ce-998ec40223d3   \n",
       "4        NaN          NaN       NaN  ec329aed-22bc-4d6e-8935-8282dcb1acac   \n",
       "...      ...          ...       ...                                   ...   \n",
       "6055     NaN          NaN       NaN  91c8e321-566f-4f9d-b89e-3a164be654d5   \n",
       "6056     NaN          NaN       NaN  62ef75e4-cbea-454e-a0ce-998ec40223d3   \n",
       "6057     NaN          NaN       NaN  73f82ac8-15cc-4fcd-87f8-5683723fce7f   \n",
       "6058     NaN          NaN       NaN  4d74781b-8186-4c9a-b659-ff4dc4601d91   \n",
       "6059     NaN          NaN       NaN  71f4bccf-53d4-4c12-9e80-e73bfb89e398   \n",
       "\n",
       "                                    czi_collection_name  \\\n",
       "0     Developmental cell programs are co-opted in in...   \n",
       "1     Neuron type-specific effects of human aging an...   \n",
       "2     Time-resolved Systems Immunology Reveals a Lat...   \n",
       "3     Cross-tissue immune cell analysis reveals tiss...   \n",
       "4     Early human lung immune cell development and i...   \n",
       "...                                                 ...   \n",
       "6055  Neuron type-specific effects of human aging an...   \n",
       "6056  Cross-tissue immune cell analysis reveals tiss...   \n",
       "6057  Developmental cell programs are co-opted in in...   \n",
       "6058  scRNA-seq assessment of the human lung, spleen...   \n",
       "6059  An Integrated Single Cell Atlas of Human Perio...   \n",
       "\n",
       "                             notes                 created_at  \\\n",
       "0     Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "1     Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "2     Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "3     Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "4     Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "...                            ...                        ...   \n",
       "6055  Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "6056  Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "6057  Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "6058  Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "6059  Processed by Chris Carpenter 2025-01-03 18:39:40.221883   \n",
       "\n",
       "                     updated_at  \n",
       "0    2025-01-03 18:39:40.221883  \n",
       "1    2025-01-03 18:39:40.221883  \n",
       "2    2025-01-03 18:39:40.221883  \n",
       "3    2025-01-03 18:39:40.221883  \n",
       "4    2025-01-03 18:39:40.221883  \n",
       "...                         ...  \n",
       "6055 2025-01-03 18:39:40.221883  \n",
       "6056 2025-01-03 18:39:40.221883  \n",
       "6057 2025-01-03 18:39:40.221883  \n",
       "6058 2025-01-03 18:39:40.221883  \n",
       "6059 2025-01-03 18:39:40.221883  \n",
       "\n",
       "[6060 rows x 19 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srx_metadata_prod = db_get_table(\"srx_metadata\", conn_prod)\n",
    "srx_metadata_prod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [database, entrez_id, srx_accession, is_illumina, is_single_cell, is_paired_end, lib_prep, tech_10x, cell_prep, organism, tissue, disease, purturbation, cell_line, czi_collection_id, czi_collection_name, notes, created_at, updated_at]\n",
       "Index: []"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter out Chris' records\n",
    "srx_metadata_prod_noChris = srx_metadata_prod[srx_metadata_prod[\"notes\"] != \"Processed by Chris Carpenter\"]\n",
    "srx_metadata_prod_noChris"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sessionInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /home/nickyoungblut/miniforge3/envs/SRAgent:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       2_gnu    conda-forge\n",
      "aiohappyeyeballs          2.4.4                    pypi_0    pypi\n",
      "aiohttp                   3.11.10                  pypi_0    pypi\n",
      "aiosignal                 1.3.2                    pypi_0    pypi\n",
      "annotated-types           0.7.0                    pypi_0    pypi\n",
      "anyio                     4.7.0                    pypi_0    pypi\n",
      "asttokens                 3.0.0              pyhd8ed1ab_1    conda-forge\n",
      "attrs                     24.2.0                   pypi_0    pypi\n",
      "beautifulsoup4            4.12.3                   pypi_0    pypi\n",
      "biopython                 1.84                     pypi_0    pypi\n",
      "build                     1.2.2.post1              pypi_0    pypi\n",
      "bzip2                     1.0.8                h4bc722e_7    conda-forge\n",
      "c-ares                    1.34.4               hb9d3cd8_0    conda-forge\n",
      "ca-certificates           2024.12.14           hbcca054_0    conda-forge\n",
      "cachetools                5.5.0                    pypi_0    pypi\n",
      "certifi                   2024.12.14               pypi_0    pypi\n",
      "charset-normalizer        3.4.0                    pypi_0    pypi\n",
      "comm                      0.2.2              pyhd8ed1ab_1    conda-forge\n",
      "curl                      8.11.1               h332b0f4_0    conda-forge\n",
      "dataclasses-json          0.6.7                    pypi_0    pypi\n",
      "db-dtypes                 1.3.1                    pypi_0    pypi\n",
      "debugpy                   1.8.11          py312h2ec8cdc_0    conda-forge\n",
      "decorator                 5.1.1              pyhd8ed1ab_1    conda-forge\n",
      "distro                    1.9.0                    pypi_0    pypi\n",
      "dynaconf                  3.2.6                    pypi_0    pypi\n",
      "exceptiongroup            1.2.2              pyhd8ed1ab_1    conda-forge\n",
      "executing                 2.1.0              pyhd8ed1ab_1    conda-forge\n",
      "frozenlist                1.5.0                    pypi_0    pypi\n",
      "google-api-core           2.24.0                   pypi_0    pypi\n",
      "google-auth               2.37.0                   pypi_0    pypi\n",
      "google-cloud-bigquery     3.27.0                   pypi_0    pypi\n",
      "google-cloud-bigquery-storage 2.27.0                   pypi_0    pypi\n",
      "google-cloud-core         2.4.1                    pypi_0    pypi\n",
      "google-cloud-secret-manager 2.22.0                   pypi_0    pypi\n",
      "google-crc32c             1.6.0                    pypi_0    pypi\n",
      "google-resumable-media    2.7.2                    pypi_0    pypi\n",
      "googleapis-common-protos  1.66.0                   pypi_0    pypi\n",
      "greenlet                  3.1.1                    pypi_0    pypi\n",
      "grpc-google-iam-v1        0.13.1                   pypi_0    pypi\n",
      "grpcio                    1.68.1                   pypi_0    pypi\n",
      "grpcio-status             1.68.1                   pypi_0    pypi\n",
      "h11                       0.14.0                   pypi_0    pypi\n",
      "httpcore                  1.0.7                    pypi_0    pypi\n",
      "httpx                     0.28.1                   pypi_0    pypi\n",
      "httpx-sse                 0.4.0                    pypi_0    pypi\n",
      "icu                       70.1                 h27087fc_0    conda-forge\n",
      "idna                      3.10                     pypi_0    pypi\n",
      "importlib-metadata        8.5.0              pyha770c72_1    conda-forge\n",
      "ipykernel                 6.29.5             pyh3099207_0    conda-forge\n",
      "ipython                   8.30.0             pyh707e725_0    conda-forge\n",
      "jedi                      0.19.2             pyhd8ed1ab_1    conda-forge\n",
      "jiter                     0.8.2                    pypi_0    pypi\n",
      "jsonpatch                 1.33                     pypi_0    pypi\n",
      "jsonpointer               3.0.0                    pypi_0    pypi\n",
      "jupyter_client            8.6.3              pyhd8ed1ab_1    conda-forge\n",
      "jupyter_core              5.7.2              pyh31011fe_1    conda-forge\n",
      "keyutils                  1.6.1                h166bdaf_0    conda-forge\n",
      "krb5                      1.21.3               h659f571_0    conda-forge\n",
      "langchain                 0.3.7                    pypi_0    pypi\n",
      "langchain-community       0.3.5                    pypi_0    pypi\n",
      "langchain-core            0.3.25                   pypi_0    pypi\n",
      "langchain-openai          0.2.6                    pypi_0    pypi\n",
      "langchain-text-splitters  0.3.3                    pypi_0    pypi\n",
      "langchainhub              0.1.21                   pypi_0    pypi\n",
      "langgraph                 0.2.45                   pypi_0    pypi\n",
      "langgraph-checkpoint      2.0.9                    pypi_0    pypi\n",
      "langgraph-sdk             0.1.45                   pypi_0    pypi\n",
      "langsmith                 0.1.147                  pypi_0    pypi\n",
      "ld_impl_linux-64          2.43                 h712a8e2_2    conda-forge\n",
      "libcurl                   8.11.1               h332b0f4_0    conda-forge\n",
      "libedit                   3.1.20191231         he28a2e2_2    conda-forge\n",
      "libev                     4.33                 hd590300_2    conda-forge\n",
      "libexpat                  2.6.4                h5888daf_0    conda-forge\n",
      "libffi                    3.4.2                h7f98852_5    conda-forge\n",
      "libgcc                    14.2.0               h77fa898_1    conda-forge\n",
      "libgcc-ng                 14.2.0               h69a702a_1    conda-forge\n",
      "libgomp                   14.2.0               h77fa898_1    conda-forge\n",
      "libiconv                  1.17                 hd590300_2    conda-forge\n",
      "liblzma                   5.6.3                hb9d3cd8_1    conda-forge\n",
      "liblzma-devel             5.6.3                hb9d3cd8_1    conda-forge\n",
      "libnghttp2                1.64.0               h161d5f1_0    conda-forge\n",
      "libnsl                    2.0.1                hd590300_0    conda-forge\n",
      "libsodium                 1.0.20               h4ab18f5_0    conda-forge\n",
      "libsqlite                 3.47.2               hee588c1_0    conda-forge\n",
      "libssh2                   1.11.1               hf672d98_0    conda-forge\n",
      "libstdcxx                 14.2.0               hc0a3c3a_1    conda-forge\n",
      "libstdcxx-ng              14.2.0               h4852527_1    conda-forge\n",
      "libuuid                   2.38.1               h0b41bf4_0    conda-forge\n",
      "libxcrypt                 4.4.36               hd590300_1    conda-forge\n",
      "libxml2                   2.9.14               h22db469_4    conda-forge\n",
      "libzlib                   1.3.1                hb9d3cd8_2    conda-forge\n",
      "marshmallow               3.23.1                   pypi_0    pypi\n",
      "matplotlib-inline         0.1.7              pyhd8ed1ab_1    conda-forge\n",
      "msgpack                   1.1.0                    pypi_0    pypi\n",
      "multidict                 6.1.0                    pypi_0    pypi\n",
      "mypy-extensions           1.0.0                    pypi_0    pypi\n",
      "ncbi-vdb                  3.1.1                h9948957_3    bioconda\n",
      "ncurses                   6.5                  he02047a_1    conda-forge\n",
      "nest-asyncio              1.6.0              pyhd8ed1ab_1    conda-forge\n",
      "numpy                     1.26.4                   pypi_0    pypi\n",
      "openai                    1.57.2                   pypi_0    pypi\n",
      "openssl                   3.4.0                hb9d3cd8_0    conda-forge\n",
      "orjson                    3.10.12                  pypi_0    pypi\n",
      "ossuuid                   1.6.2             hf484d3e_1000    conda-forge\n",
      "packaging                 24.2               pyhd8ed1ab_2    conda-forge\n",
      "pandas                    2.2.3                    pypi_0    pypi\n",
      "parso                     0.8.4              pyhd8ed1ab_1    conda-forge\n",
      "perl                      5.26.2            h36c2ea0_1008    conda-forge\n",
      "perl-app-cpanminus        1.7044                  pl526_1    bioconda\n",
      "perl-business-isbn        3.004                   pl526_0    bioconda\n",
      "perl-business-isbn-data   20140910.003            pl526_0    bioconda\n",
      "perl-carp                 1.38                    pl526_3    bioconda\n",
      "perl-constant             1.33                    pl526_1    bioconda\n",
      "perl-data-dumper          2.173                   pl526_0    bioconda\n",
      "perl-encode               2.88                    pl526_1    bioconda\n",
      "perl-exporter             5.72                    pl526_1    bioconda\n",
      "perl-extutils-makemaker   7.36                    pl526_1    bioconda\n",
      "perl-file-path            2.16                    pl526_0    bioconda\n",
      "perl-file-temp            0.2304                  pl526_2    bioconda\n",
      "perl-mime-base64          3.15                    pl526_1    bioconda\n",
      "perl-parent               0.236                   pl526_1    bioconda\n",
      "perl-uri                  1.76                    pl526_0    bioconda\n",
      "perl-xml-libxml           2.0132          pl526h7ec2d77_1    bioconda\n",
      "perl-xml-namespacesupport 1.12                    pl526_0    bioconda\n",
      "perl-xml-sax              1.02                    pl526_0    bioconda\n",
      "perl-xml-sax-base         1.09                    pl526_0    bioconda\n",
      "perl-xsloader             0.24                    pl526_0    bioconda\n",
      "pexpect                   4.9.0              pyhd8ed1ab_1    conda-forge\n",
      "pickleshare               0.7.5           pyhd8ed1ab_1004    conda-forge\n",
      "pip                       24.3.1             pyh8b19718_0    conda-forge\n",
      "platformdirs              4.3.6              pyhd8ed1ab_1    conda-forge\n",
      "prompt-toolkit            3.0.48             pyha770c72_1    conda-forge\n",
      "propcache                 0.2.1                    pypi_0    pypi\n",
      "proto-plus                1.25.0                   pypi_0    pypi\n",
      "protobuf                  5.29.1                   pypi_0    pypi\n",
      "psutil                    6.1.0           py312h66e93f0_0    conda-forge\n",
      "psycopg2-binary           2.9.10                   pypi_0    pypi\n",
      "ptyprocess                0.7.0              pyhd8ed1ab_1    conda-forge\n",
      "pure_eval                 0.2.3              pyhd8ed1ab_1    conda-forge\n",
      "pyarrow                   18.1.0                   pypi_0    pypi\n",
      "pyasn1                    0.6.1                    pypi_0    pypi\n",
      "pyasn1-modules            0.4.1                    pypi_0    pypi\n",
      "pydantic                  2.10.3                   pypi_0    pypi\n",
      "pydantic-core             2.27.1                   pypi_0    pypi\n",
      "pydantic-settings         2.7.0                    pypi_0    pypi\n",
      "pygments                  2.18.0             pyhd8ed1ab_1    conda-forge\n",
      "pypika                    0.48.9                   pypi_0    pypi\n",
      "pyproject-hooks           1.2.0                    pypi_0    pypi\n",
      "python                    3.12.8          h9e4cc4f_1_cpython    conda-forge\n",
      "python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge\n",
      "python-dotenv             1.0.1                    pypi_0    pypi\n",
      "python_abi                3.12                    5_cp312    conda-forge\n",
      "pytz                      2024.2                   pypi_0    pypi\n",
      "pyyaml                    6.0.2                    pypi_0    pypi\n",
      "pyzmq                     26.2.0          py312hbf22597_3    conda-forge\n",
      "readline                  8.2                  h8228510_1    conda-forge\n",
      "regex                     2024.11.6                pypi_0    pypi\n",
      "requests                  2.32.3                   pypi_0    pypi\n",
      "requests-toolbelt         1.0.0                    pypi_0    pypi\n",
      "rsa                       4.9                      pypi_0    pypi\n",
      "setuptools                75.6.0             pyhff2d567_1    conda-forge\n",
      "six                       1.17.0             pyhd8ed1ab_0    conda-forge\n",
      "sniffio                   1.3.1                    pypi_0    pypi\n",
      "soupsieve                 2.6                      pypi_0    pypi\n",
      "sqlalchemy                2.0.35                   pypi_0    pypi\n",
      "sra-tools                 3.1.1                h4304569_2    bioconda\n",
      "sragent                   0.1.0                    pypi_0    pypi\n",
      "stack_data                0.6.3              pyhd8ed1ab_1    conda-forge\n",
      "tabulate                  0.9.0                    pypi_0    pypi\n",
      "tenacity                  9.0.0                    pypi_0    pypi\n",
      "tiktoken                  0.8.0                    pypi_0    pypi\n",
      "tk                        8.6.13          noxft_h4845f30_101    conda-forge\n",
      "tornado                   6.4.2           py312h66e93f0_0    conda-forge\n",
      "tqdm                      4.67.1                   pypi_0    pypi\n",
      "traitlets                 5.14.3             pyhd8ed1ab_1    conda-forge\n",
      "types-requests            2.32.0.20241016          pypi_0    pypi\n",
      "typing-inspect            0.9.0                    pypi_0    pypi\n",
      "typing_extensions         4.12.2             pyha770c72_1    conda-forge\n",
      "tzdata                    2024.2                   pypi_0    pypi\n",
      "urllib3                   2.2.3                    pypi_0    pypi\n",
      "wcwidth                   0.2.13             pyhd8ed1ab_1    conda-forge\n",
      "wheel                     0.45.1             pyhd8ed1ab_1    conda-forge\n",
      "xmltodict                 0.14.2                   pypi_0    pypi\n",
      "xz                        5.6.3                hbcc6ac9_1    conda-forge\n",
      "xz-gpl-tools              5.6.3                hbcc6ac9_1    conda-forge\n",
      "xz-tools                  5.6.3                hb9d3cd8_1    conda-forge\n",
      "yarl                      1.18.3                   pypi_0    pypi\n",
      "zeromq                    4.3.5                h3b0a872_7    conda-forge\n",
      "zipp                      3.21.0             pyhd8ed1ab_1    conda-forge\n",
      "zstd                      1.5.6                ha6fb4c9_0    conda-forge\n"
     ]
    }
   ],
   "source": [
    "!conda list"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
