{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Migrate data from test to production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.upsert import db_upsert\n",
    "from SRAgent.db.utils import db_list_tables, db_glimpse_tables, execute_query\n",
    "from SRAgent.db.get import db_find_srx\n",
    "from SRAgent.db.create import create_table, create_table_router"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "srx_srr\n",
      "screcounter_star\n",
      "eval\n",
      "screcounter_log\n",
      "srx_metadata\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in test\n",
    "os.environ['DYNACONF'] = 'test'\n",
    "conn_test = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_test)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX26727599\tSRR31350667\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23538581\tSRR27876733\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23261451\tSRR27592690\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23261451\tSRR27592688\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "SRX23261451\tSRR27592689\t2024-12-18 17:37:23.074909\t2024-12-18 17:37:23.074909\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "SRX19729045\tSRR23917610\tgene\t139\t1.0\t\t4.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.171386\t556\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tgene_ex50\t197\t1.0\t\t3.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.149675\t784\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tgene_ex_int\t198\t1.0\t\t3.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.1507\t789\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tgene_full\t188\t1.0\t\t4.0\t3.0\t\t1.0\t1.0\t100000\t0\t0.147486\t763\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "SRX19729045\tSRR23917610\tvelocyto\t0\t\t\t\t\t\t\t\t100000\t0\t-inf\t0\t2024-12-23 21:49:11.388231\t2024-12-23 21:49:11.388231\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2024-12-18 17:38:54.774086\t2025-01-02 23:08:54.386892\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "ERX10987225\tERR11583756\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583807\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583914\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583981\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "ERX10987225\tERR11583986\tget database accessions\tFinal\tSuccess\tObtained database accession for processing\t2024-12-23 03:56:14.065204\t2024-12-23 03:56:14.065204\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tnotes\tcreated_at\tupdated_at\n",
      "sra\t33249542\tSRX24914804\tyes\tyes\tyes\tSmart-seq2\tnot_applicable\tsingle_cell\thuman\tblood\tHepatitis C Virus (HCV) reinfection outcomes\tAkebia Saponin D (ASD) and TNF-induced optic nerve damage\tMemory B cells\tMetadata obtained by SRAgent\t2024-12-18 18:23:36.181001\t2024-12-18 18:25:14.322667\n",
      "sra\t34603139\tSRX25602353\tyes\tyes\tyes\t10x_Genomics\tother\tsingle_cell\thuman\tPBMCs\tunsure\teffects of salt intake in healthy humans\tunsure\tMetadata obtained by SRAgent\t2024-12-18 18:23:36.181001\t2024-12-18 18:25:33.937056\n",
      "sra\t31225478\tSRX23114342\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tsingle_cell\tmouse\tsciatic nerve\tspontaneous autoimmune neuropathy (CIDP)\tunsure\timmune cells (CD45+)\tMetadata obtained by SRAgent\t2024-12-18 18:40:50.574099\t2024-12-18 18:42:11.558669\n",
      "sra\t34603098\tSRX25602312\tyes\tyes\tyes\tother\tnot_applicable\tsingle_cell\thuman\tPBMCs\tnone\teffects of salt intake\tnone\tMetadata obtained by SRAgent\t2024-12-18 18:40:50.574099\t2024-12-18 18:42:33.593136\n",
      "sra\t31787850\tSRX23530678\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tsingle_cell\tmouse\tbone marrow\tneuroblastoma\tMxra7 knockout\tbone marrow cells from Mxra7 knockout mice\tMetadata obtained by SRAgent\t2024-12-18 18:43:35.370648\t2024-12-18 18:45:18.280506\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "srx_srr\n",
      "eval\n",
      "screcounter_log\n",
      "screcounter_star\n",
      "srx_metadata\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in prod\n",
    "os.environ['DYNACONF'] = 'prod'\n",
    "conn_prod = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_prod)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"../db_bkup/2025-01-02\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eval.csv\n",
      "srx_metadata.csv\n",
      "screcounter_log.csv\n",
      "screcounter_star.csv\n",
      "srx_srr.csv\n"
     ]
    }
   ],
   "source": [
    "# list csv files in the data directory\n",
    "import os\n",
    "csv_files = [f for f in os.listdir(data_dir) if f.endswith('.csv')]\n",
    "print(\"\\n\".join(csv_files))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>33249542</td>\n",
       "      <td>SRX24914804</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>Hepatitis C Virus (HCV) reinfection outcomes</td>\n",
       "      <td>Akebia Saponin D (ASD) and TNF-induced optic n...</td>\n",
       "      <td>Memory B cells</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2024-12-18 18:23:36.181001</td>\n",
       "      <td>2024-12-18 18:25:14.322667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>34603139</td>\n",
       "      <td>SRX25602353</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>unsure</td>\n",
       "      <td>effects of salt intake in healthy humans</td>\n",
       "      <td>unsure</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2024-12-18 18:23:36.181001</td>\n",
       "      <td>2024-12-18 18:25:33.937056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>31225478</td>\n",
       "      <td>SRX23114342</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>sciatic nerve</td>\n",
       "      <td>spontaneous autoimmune neuropathy (CIDP)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>immune cells (CD45+)</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2024-12-18 18:40:50.574099</td>\n",
       "      <td>2024-12-18 18:42:11.558669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>34603098</td>\n",
       "      <td>SRX25602312</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>none</td>\n",
       "      <td>effects of salt intake</td>\n",
       "      <td>none</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2024-12-18 18:40:50.574099</td>\n",
       "      <td>2024-12-18 18:42:33.593136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>31787850</td>\n",
       "      <td>SRX23530678</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>Mxra7 knockout</td>\n",
       "      <td>bone marrow cells from Mxra7 knockout mice</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2024-12-18 18:43:35.370648</td>\n",
       "      <td>2024-12-18 18:45:18.280506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  database  entrez_id srx_accession is_illumina is_single_cell is_paired_end  \\\n",
       "0      sra   33249542   SRX24914804         yes            yes           yes   \n",
       "1      sra   34603139   SRX25602353         yes            yes           yes   \n",
       "2      sra   31225478   SRX23114342         yes            yes           yes   \n",
       "3      sra   34603098   SRX25602312         yes            yes           yes   \n",
       "4      sra   31787850   SRX23530678         yes            yes           yes   \n",
       "\n",
       "       lib_prep        tech_10x    cell_prep organism         tissue  \\\n",
       "0    Smart-seq2  not_applicable  single_cell    human          blood   \n",
       "1  10x_Genomics           other  single_cell    human          PBMCs   \n",
       "2  10x_Genomics     3_prime_gex  single_cell    mouse  sciatic nerve   \n",
       "3         other  not_applicable  single_cell    human          PBMCs   \n",
       "4  10x_Genomics     3_prime_gex  single_cell    mouse    bone marrow   \n",
       "\n",
       "                                        disease  \\\n",
       "0  Hepatitis C Virus (HCV) reinfection outcomes   \n",
       "1                                        unsure   \n",
       "2      spontaneous autoimmune neuropathy (CIDP)   \n",
       "3                                          none   \n",
       "4                                 neuroblastoma   \n",
       "\n",
       "                                        purturbation  \\\n",
       "0  Akebia Saponin D (ASD) and TNF-induced optic n...   \n",
       "1           effects of salt intake in healthy humans   \n",
       "2                                             unsure   \n",
       "3                             effects of salt intake   \n",
       "4                                     Mxra7 knockout   \n",
       "\n",
       "                                    cell_line                         notes  \\\n",
       "0                              Memory B cells  Metadata obtained by SRAgent   \n",
       "1                                      unsure  Metadata obtained by SRAgent   \n",
       "2                        immune cells (CD45+)  Metadata obtained by SRAgent   \n",
       "3                                        none  Metadata obtained by SRAgent   \n",
       "4  bone marrow cells from Mxra7 knockout mice  Metadata obtained by SRAgent   \n",
       "\n",
       "                   created_at                  updated_at  \n",
       "0  2024-12-18 18:23:36.181001  2024-12-18 18:25:14.322667  \n",
       "1  2024-12-18 18:23:36.181001  2024-12-18 18:25:33.937056  \n",
       "2  2024-12-18 18:40:50.574099  2024-12-18 18:42:11.558669  \n",
       "3  2024-12-18 18:40:50.574099  2024-12-18 18:42:33.593136  \n",
       "4  2024-12-18 18:43:35.370648  2024-12-18 18:45:18.280506  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "infile = os.path.join(data_dir, \"srx_metadata.csv\")\n",
    "srx_metadata = pd.read_csv(infile)\n",
    "srx_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Chris' processed records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Metadata obtained by SRAgent',\n",
       "       'New dataset found by Find-Datasets agent',\n",
       "       'Processed by Chris Carpenter', nan], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Chris' records\n",
    "srx_metadata[\"notes\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110064</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>sra</td>\n",
       "      <td>19007785</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>sra</td>\n",
       "      <td>27175908</td>\n",
       "      <td>SRX19828711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>sra</td>\n",
       "      <td>27176348</td>\n",
       "      <td>SRX19829151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488012</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16774</th>\n",
       "      <td>sra</td>\n",
       "      <td>12773490</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16775</th>\n",
       "      <td>sra</td>\n",
       "      <td>15435584</td>\n",
       "      <td>SRX11523521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16776</th>\n",
       "      <td>sra</td>\n",
       "      <td>27177388</td>\n",
       "      <td>SRX19830191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16777</th>\n",
       "      <td>sra</td>\n",
       "      <td>27172504</td>\n",
       "      <td>SRX19825307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16778</th>\n",
       "      <td>sra</td>\n",
       "      <td>1644947</td>\n",
       "      <td>SRX1128586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>2024-12-18 17:31:16.385013</td>\n",
       "      <td>2024-12-18 17:31:35.121583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16660 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession is_illumina is_single_cell  \\\n",
       "103        sra   29110064   ERX11148781         NaN            NaN   \n",
       "108        sra   19007785   SRX13670569         NaN            NaN   \n",
       "109        sra   27175908   SRX19828711         NaN            NaN   \n",
       "110        sra   27176348   SRX19829151         NaN            NaN   \n",
       "123        sra   12488012    SRX9556597         NaN            NaN   \n",
       "...        ...        ...           ...         ...            ...   \n",
       "16774      sra   12773490    SRX9770752         NaN            NaN   \n",
       "16775      sra   15435584   SRX11523521         NaN            NaN   \n",
       "16776      sra   27177388   SRX19830191         NaN            NaN   \n",
       "16777      sra   27172504   SRX19825307         NaN            NaN   \n",
       "16778      sra    1644947    SRX1128586         NaN            NaN   \n",
       "\n",
       "      is_paired_end lib_prep tech_10x cell_prep organism tissue disease  \\\n",
       "103             NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "108             NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "109             NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "110             NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "123             NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "...             ...      ...      ...       ...      ...    ...     ...   \n",
       "16774           NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "16775           NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "16776           NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "16777           NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "16778           NaN      NaN      NaN       NaN      NaN    NaN     NaN   \n",
       "\n",
       "      purturbation cell_line                         notes  \\\n",
       "103            NaN       NaN  Processed by Chris Carpenter   \n",
       "108            NaN       NaN  Processed by Chris Carpenter   \n",
       "109            NaN       NaN  Processed by Chris Carpenter   \n",
       "110            NaN       NaN  Processed by Chris Carpenter   \n",
       "123            NaN       NaN  Processed by Chris Carpenter   \n",
       "...            ...       ...                           ...   \n",
       "16774          NaN       NaN  Processed by Chris Carpenter   \n",
       "16775          NaN       NaN  Processed by Chris Carpenter   \n",
       "16776          NaN       NaN  Processed by Chris Carpenter   \n",
       "16777          NaN       NaN  Processed by Chris Carpenter   \n",
       "16778          NaN       NaN  Processed by Chris Carpenter   \n",
       "\n",
       "                       created_at                  updated_at  \n",
       "103    2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "108    2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "109    2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "110    2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "123    2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "...                           ...                         ...  \n",
       "16774  2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "16775  2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "16776  2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "16777  2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "16778  2024-12-18 17:31:16.385013  2024-12-18 17:31:35.121583  \n",
       "\n",
       "[16660 rows x 17 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srx_metadata_chris = srx_metadata[srx_metadata[\"notes\"] == \"Processed by Chris Carpenter\"]\n",
    "srx_metadata_chris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "      <th>experiment</th>\n",
       "      <th>CZI_collection_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110064</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>19007785</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>27175908</td>\n",
       "      <td>SRX19828711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX19828711</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>27176348</td>\n",
       "      <td>SRX19829151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX19829151</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488012</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69490</th>\n",
       "      <td>sra</td>\n",
       "      <td>12773490</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>3'v3</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69492</th>\n",
       "      <td>sra</td>\n",
       "      <td>15435584</td>\n",
       "      <td>SRX11523521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX11523521</td>\n",
       "      <td>MSK SPECTRUM ‚Äì Ovarian cancer mutational pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69494</th>\n",
       "      <td>sra</td>\n",
       "      <td>27177388</td>\n",
       "      <td>SRX19830191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX19830191</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69495</th>\n",
       "      <td>sra</td>\n",
       "      <td>27172504</td>\n",
       "      <td>SRX19825307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX19825307</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69496</th>\n",
       "      <td>sra</td>\n",
       "      <td>1644947</td>\n",
       "      <td>SRX1128586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>31:16.4</td>\n",
       "      <td>31:35.1</td>\n",
       "      <td>SRX1128586</td>\n",
       "      <td>Adult mouse cortical cell taxonomy revealed by...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16660 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession  is_illumina  is_single_cell  \\\n",
       "0          sra   29110064   ERX11148781          NaN             NaN   \n",
       "1          sra   19007785   SRX13670569          NaN             NaN   \n",
       "2          sra   27175908   SRX19828711          NaN             NaN   \n",
       "3          sra   27176348   SRX19829151          NaN             NaN   \n",
       "4          sra   12488012    SRX9556597          NaN             NaN   \n",
       "...        ...        ...           ...          ...             ...   \n",
       "69490      sra   12773490    SRX9770752          NaN             NaN   \n",
       "69492      sra   15435584   SRX11523521          NaN             NaN   \n",
       "69494      sra   27177388   SRX19830191          NaN             NaN   \n",
       "69495      sra   27172504   SRX19825307          NaN             NaN   \n",
       "69496      sra    1644947    SRX1128586          NaN             NaN   \n",
       "\n",
       "       is_paired_end        lib_prep tech_10x    cell_prep  organism  tissue  \\\n",
       "0                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "1                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "2                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "3                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "4                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "...              ...             ...      ...          ...       ...     ...   \n",
       "69490            NaN  \\t10x_Genomics     3'v3  single_cell       NaN     NaN   \n",
       "69492            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "69494            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "69495            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "69496            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "\n",
       "       disease  purturbation  cell_line                         notes  \\\n",
       "0          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "1          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "2          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "3          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "4          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "...        ...           ...        ...                           ...   \n",
       "69490      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69492      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69494      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69495      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69496      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "\n",
       "      created_at updated_at   experiment  \\\n",
       "0        31:16.4    31:35.1  ERX11148781   \n",
       "1        31:16.4    31:35.1  SRX13670569   \n",
       "2        31:16.4    31:35.1  SRX19828711   \n",
       "3        31:16.4    31:35.1  SRX19829151   \n",
       "4        31:16.4    31:35.1   SRX9556597   \n",
       "...          ...        ...          ...   \n",
       "69490    31:16.4    31:35.1   SRX9770752   \n",
       "69492    31:16.4    31:35.1  SRX11523521   \n",
       "69494    31:16.4    31:35.1  SRX19830191   \n",
       "69495    31:16.4    31:35.1  SRX19825307   \n",
       "69496    31:16.4    31:35.1   SRX1128586   \n",
       "\n",
       "                                     CZI_collection_name  \n",
       "0      Developmental cell programs are co-opted in in...  \n",
       "1      Neuron type-specific effects of human aging an...  \n",
       "2      A single-cell transcriptional timelapse of mou...  \n",
       "3      A single-cell transcriptional timelapse of mou...  \n",
       "4      Time-resolved Systems Immunology Reveals a Lat...  \n",
       "...                                                  ...  \n",
       "69490  An Integrated Single Cell Atlas of Human Perio...  \n",
       "69492  MSK SPECTRUM ‚Äì Ovarian cancer mutational pro...  \n",
       "69494  A single-cell transcriptional timelapse of mou...  \n",
       "69495  A single-cell transcriptional timelapse of mou...  \n",
       "69496  Adult mouse cortical cell taxonomy revealed by...  \n",
       "\n",
       "[16660 rows x 19 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "infile = \"../data/20241227_ChrisC_metadata.tsv\"\n",
    "chris_cur_meta = pd.read_csv(infile, sep=\"\\t\").drop_duplicates()\n",
    "chris_cur_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the columns that are not in the database\n",
    "chris_cur_meta.drop([\"experiment\", \"created_at\", \"updated_at\"], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>notes</th>\n",
       "      <th>czi_collection_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>29110064</td>\n",
       "      <td>ERX11148781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>19007785</td>\n",
       "      <td>SRX13670569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>27175908</td>\n",
       "      <td>SRX19828711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>27176348</td>\n",
       "      <td>SRX19829151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488012</td>\n",
       "      <td>SRX9556597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>5'</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69490</th>\n",
       "      <td>sra</td>\n",
       "      <td>12773490</td>\n",
       "      <td>SRX9770752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\\t10x_Genomics</td>\n",
       "      <td>3'v3</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69492</th>\n",
       "      <td>sra</td>\n",
       "      <td>15435584</td>\n",
       "      <td>SRX11523521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>MSK SPECTRUM ‚Äì Ovarian cancer mutational pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69494</th>\n",
       "      <td>sra</td>\n",
       "      <td>27177388</td>\n",
       "      <td>SRX19830191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69495</th>\n",
       "      <td>sra</td>\n",
       "      <td>27172504</td>\n",
       "      <td>SRX19825307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69496</th>\n",
       "      <td>sra</td>\n",
       "      <td>1644947</td>\n",
       "      <td>SRX1128586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Processed by Chris Carpenter</td>\n",
       "      <td>Adult mouse cortical cell taxonomy revealed by...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16660 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      database  entrez_id srx_accession  is_illumina  is_single_cell  \\\n",
       "0          sra   29110064   ERX11148781          NaN             NaN   \n",
       "1          sra   19007785   SRX13670569          NaN             NaN   \n",
       "2          sra   27175908   SRX19828711          NaN             NaN   \n",
       "3          sra   27176348   SRX19829151          NaN             NaN   \n",
       "4          sra   12488012    SRX9556597          NaN             NaN   \n",
       "...        ...        ...           ...          ...             ...   \n",
       "69490      sra   12773490    SRX9770752          NaN             NaN   \n",
       "69492      sra   15435584   SRX11523521          NaN             NaN   \n",
       "69494      sra   27177388   SRX19830191          NaN             NaN   \n",
       "69495      sra   27172504   SRX19825307          NaN             NaN   \n",
       "69496      sra    1644947    SRX1128586          NaN             NaN   \n",
       "\n",
       "       is_paired_end        lib_prep tech_10x    cell_prep  organism  tissue  \\\n",
       "0                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "1                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "2                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "3                NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "4                NaN  \\t10x_Genomics       5'  single_cell       NaN     NaN   \n",
       "...              ...             ...      ...          ...       ...     ...   \n",
       "69490            NaN  \\t10x_Genomics     3'v3  single_cell       NaN     NaN   \n",
       "69492            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "69494            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "69495            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "69496            NaN             NaN      NaN          NaN       NaN     NaN   \n",
       "\n",
       "       disease  purturbation  cell_line                         notes  \\\n",
       "0          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "1          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "2          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "3          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "4          NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "...        ...           ...        ...                           ...   \n",
       "69490      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69492      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69494      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69495      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "69496      NaN           NaN        NaN  Processed by Chris Carpenter   \n",
       "\n",
       "                                     czi_collection_name  \n",
       "0      Developmental cell programs are co-opted in in...  \n",
       "1      Neuron type-specific effects of human aging an...  \n",
       "2      A single-cell transcriptional timelapse of mou...  \n",
       "3      A single-cell transcriptional timelapse of mou...  \n",
       "4      Time-resolved Systems Immunology Reveals a Lat...  \n",
       "...                                                  ...  \n",
       "69490  An Integrated Single Cell Atlas of Human Perio...  \n",
       "69492  MSK SPECTRUM ‚Äì Ovarian cancer mutational pro...  \n",
       "69494  A single-cell transcriptional timelapse of mou...  \n",
       "69495  A single-cell transcriptional timelapse of mou...  \n",
       "69496  Adult mouse cortical cell taxonomy revealed by...  \n",
       "\n",
       "[16660 rows x 16 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lower case column names\n",
    "chris_cur_meta.columns = chris_cur_meta.columns.str.lower()\n",
    "chris_cur_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Format metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "## count values in each character column\n",
    "for col in [\"is_illumina\", \"is_single_cell\", \"is_paired_end\"]:\n",
    "    print(chris_cur_meta[col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format tech_10x\n",
    "idx = {\n",
    "    \"5'\": \"5_prime_gex\",\n",
    "    \"3'v3\": \"3_prime_gex\",\n",
    "    \"3'v2\": \"3_prime_gex\",\n",
    "    \"multiome GEX\": \"multiome\",\n",
    "}\n",
    "for k,v in idx.items():\n",
    "    chris_cur_meta[\"tech_10x\"] = chris_cur_meta[\"tech_10x\"].replace(k, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format lib_prep\n",
    "idx = {\n",
    "    \"\\t10x_Genomics\": \"10x_Genomics\",\n",
    "}\n",
    "for k,v in idx.items():\n",
    "    chris_cur_meta[\"lib_prep\"] = chris_cur_meta[\"lib_prep\"].replace(k, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lib_prep\n",
      "10x_Genomics    2367\n",
      "Name: count, dtype: int64\n",
      "tech_10x\n",
      "3_prime_gex    1323\n",
      "5_prime_gex     873\n",
      "multiome        122\n",
      "Name: count, dtype: int64\n",
      "cell_prep\n",
      "single_cell    2367\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## count values in each character column\n",
    "for col in [\"lib_prep\", \"tech_10x\", \"cell_prep\"]:\n",
    "    print(chris_cur_meta[col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n",
      "Series([], Name: count, dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "## count values in each character column\n",
    "for col in [\"organism\", \"tissue\", \"disease\", \"purturbation\", \"cell_line\"]:\n",
    "    print(chris_cur_meta[col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to file\n",
    "outfile = os.path.join(\"../data/\", \"20241227_ChrisC_metadata_v2.tsv\")\n",
    "chris_cur_meta.to_csv(outfile, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upserted ../data/20241227_ChrisC_metadata_v2.tsv into srx_metadata\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "../scripts/db-tools.py --upsert-csv ../data/20241227_ChrisC_metadata_v2.tsv --upsert-target srx_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: screcounter_star --#\n",
      "sample\taccession\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\treads_with_valid_barcodes\tsequencing_saturation\tumis_in_cells\tcreated_at\tupdated_at\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t29110064\tERX11148781\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t\tDevelopmental cell programs are co-opted in inflammatory skin disease\tProcessed by Chris Carpenter\t2025-01-03 17:09:51.186682\t2025-01-03 17:09:51.186682\n",
      "sra\t19007785\tSRX13670569\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t\tNeuron type-specific effects of human aging and sex on DNA methylation and transcription\tProcessed by Chris Carpenter\t2025-01-03 17:09:51.186682\t2025-01-03 17:09:51.186682\n",
      "sra\t27175908\tSRX19828711\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t\tA single-cell transcriptional timelapse of mouse embryonic development, from gastrula to pup\tProcessed by Chris Carpenter\t2025-01-03 17:09:51.186682\t2025-01-03 17:09:51.186682\n",
      "sra\t27176348\tSRX19829151\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\tNaN\t\tA single-cell transcriptional timelapse of mouse embryonic development, from gastrula to pup\tProcessed by Chris Carpenter\t2025-01-03 17:09:51.186682\t2025-01-03 17:09:51.186682\n",
      "sra\t12488012\tSRX9556597\tNaN\tNaN\tNaN\t10x_Genomics\t5_prime_gex\tsingle_cell\tNaN\tNaN\tNaN\tNaN\tNaN\t\tTime-resolved Systems Immunology Reveals a Late Juncture Linked to Fatal COVID-19\tProcessed by Chris Carpenter\t2025-01-03 17:09:51.186682\t2025-01-03 17:09:51.186682\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "../scripts/db-tools.py --glimpse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sessionInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /home/nickyoungblut/miniforge3/envs/SRAgent:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       2_gnu    conda-forge\n",
      "aiohappyeyeballs          2.4.4                    pypi_0    pypi\n",
      "aiohttp                   3.11.10                  pypi_0    pypi\n",
      "aiosignal                 1.3.2                    pypi_0    pypi\n",
      "annotated-types           0.7.0                    pypi_0    pypi\n",
      "anyio                     4.7.0                    pypi_0    pypi\n",
      "asttokens                 3.0.0              pyhd8ed1ab_1    conda-forge\n",
      "attrs                     24.2.0                   pypi_0    pypi\n",
      "beautifulsoup4            4.12.3                   pypi_0    pypi\n",
      "biopython                 1.84                     pypi_0    pypi\n",
      "build                     1.2.2.post1              pypi_0    pypi\n",
      "bzip2                     1.0.8                h4bc722e_7    conda-forge\n",
      "c-ares                    1.34.4               hb9d3cd8_0    conda-forge\n",
      "ca-certificates           2024.12.14           hbcca054_0    conda-forge\n",
      "cachetools                5.5.0                    pypi_0    pypi\n",
      "certifi                   2024.12.14               pypi_0    pypi\n",
      "charset-normalizer        3.4.0                    pypi_0    pypi\n",
      "comm                      0.2.2              pyhd8ed1ab_1    conda-forge\n",
      "curl                      8.11.1               h332b0f4_0    conda-forge\n",
      "dataclasses-json          0.6.7                    pypi_0    pypi\n",
      "db-dtypes                 1.3.1                    pypi_0    pypi\n",
      "debugpy                   1.8.11          py312h2ec8cdc_0    conda-forge\n",
      "decorator                 5.1.1              pyhd8ed1ab_1    conda-forge\n",
      "distro                    1.9.0                    pypi_0    pypi\n",
      "dynaconf                  3.2.6                    pypi_0    pypi\n",
      "exceptiongroup            1.2.2              pyhd8ed1ab_1    conda-forge\n",
      "executing                 2.1.0              pyhd8ed1ab_1    conda-forge\n",
      "frozenlist                1.5.0                    pypi_0    pypi\n",
      "google-api-core           2.24.0                   pypi_0    pypi\n",
      "google-auth               2.37.0                   pypi_0    pypi\n",
      "google-cloud-bigquery     3.27.0                   pypi_0    pypi\n",
      "google-cloud-bigquery-storage 2.27.0                   pypi_0    pypi\n",
      "google-cloud-core         2.4.1                    pypi_0    pypi\n",
      "google-cloud-secret-manager 2.22.0                   pypi_0    pypi\n",
      "google-crc32c             1.6.0                    pypi_0    pypi\n",
      "google-resumable-media    2.7.2                    pypi_0    pypi\n",
      "googleapis-common-protos  1.66.0                   pypi_0    pypi\n",
      "greenlet                  3.1.1                    pypi_0    pypi\n",
      "grpc-google-iam-v1        0.13.1                   pypi_0    pypi\n",
      "grpcio                    1.68.1                   pypi_0    pypi\n",
      "grpcio-status             1.68.1                   pypi_0    pypi\n",
      "h11                       0.14.0                   pypi_0    pypi\n",
      "httpcore                  1.0.7                    pypi_0    pypi\n",
      "httpx                     0.28.1                   pypi_0    pypi\n",
      "httpx-sse                 0.4.0                    pypi_0    pypi\n",
      "icu                       70.1                 h27087fc_0    conda-forge\n",
      "idna                      3.10                     pypi_0    pypi\n",
      "importlib-metadata        8.5.0              pyha770c72_1    conda-forge\n",
      "ipykernel                 6.29.5             pyh3099207_0    conda-forge\n",
      "ipython                   8.30.0             pyh707e725_0    conda-forge\n",
      "jedi                      0.19.2             pyhd8ed1ab_1    conda-forge\n",
      "jiter                     0.8.2                    pypi_0    pypi\n",
      "jsonpatch                 1.33                     pypi_0    pypi\n",
      "jsonpointer               3.0.0                    pypi_0    pypi\n",
      "jupyter_client            8.6.3              pyhd8ed1ab_1    conda-forge\n",
      "jupyter_core              5.7.2              pyh31011fe_1    conda-forge\n",
      "keyutils                  1.6.1                h166bdaf_0    conda-forge\n",
      "krb5                      1.21.3               h659f571_0    conda-forge\n",
      "langchain                 0.3.7                    pypi_0    pypi\n",
      "langchain-community       0.3.5                    pypi_0    pypi\n",
      "langchain-core            0.3.25                   pypi_0    pypi\n",
      "langchain-openai          0.2.6                    pypi_0    pypi\n",
      "langchain-text-splitters  0.3.3                    pypi_0    pypi\n",
      "langchainhub              0.1.21                   pypi_0    pypi\n",
      "langgraph                 0.2.45                   pypi_0    pypi\n",
      "langgraph-checkpoint      2.0.9                    pypi_0    pypi\n",
      "langgraph-sdk             0.1.45                   pypi_0    pypi\n",
      "langsmith                 0.1.147                  pypi_0    pypi\n",
      "ld_impl_linux-64          2.43                 h712a8e2_2    conda-forge\n",
      "libcurl                   8.11.1               h332b0f4_0    conda-forge\n",
      "libedit                   3.1.20191231         he28a2e2_2    conda-forge\n",
      "libev                     4.33                 hd590300_2    conda-forge\n",
      "libexpat                  2.6.4                h5888daf_0    conda-forge\n",
      "libffi                    3.4.2                h7f98852_5    conda-forge\n",
      "libgcc                    14.2.0               h77fa898_1    conda-forge\n",
      "libgcc-ng                 14.2.0               h69a702a_1    conda-forge\n",
      "libgomp                   14.2.0               h77fa898_1    conda-forge\n",
      "libiconv                  1.17                 hd590300_2    conda-forge\n",
      "liblzma                   5.6.3                hb9d3cd8_1    conda-forge\n",
      "liblzma-devel             5.6.3                hb9d3cd8_1    conda-forge\n",
      "libnghttp2                1.64.0               h161d5f1_0    conda-forge\n",
      "libnsl                    2.0.1                hd590300_0    conda-forge\n",
      "libsodium                 1.0.20               h4ab18f5_0    conda-forge\n",
      "libsqlite                 3.47.2               hee588c1_0    conda-forge\n",
      "libssh2                   1.11.1               hf672d98_0    conda-forge\n",
      "libstdcxx                 14.2.0               hc0a3c3a_1    conda-forge\n",
      "libstdcxx-ng              14.2.0               h4852527_1    conda-forge\n",
      "libuuid                   2.38.1               h0b41bf4_0    conda-forge\n",
      "libxcrypt                 4.4.36               hd590300_1    conda-forge\n",
      "libxml2                   2.9.14               h22db469_4    conda-forge\n",
      "libzlib                   1.3.1                hb9d3cd8_2    conda-forge\n",
      "marshmallow               3.23.1                   pypi_0    pypi\n",
      "matplotlib-inline         0.1.7              pyhd8ed1ab_1    conda-forge\n",
      "msgpack                   1.1.0                    pypi_0    pypi\n",
      "multidict                 6.1.0                    pypi_0    pypi\n",
      "mypy-extensions           1.0.0                    pypi_0    pypi\n",
      "ncbi-vdb                  3.1.1                h9948957_3    bioconda\n",
      "ncurses                   6.5                  he02047a_1    conda-forge\n",
      "nest-asyncio              1.6.0              pyhd8ed1ab_1    conda-forge\n",
      "numpy                     1.26.4                   pypi_0    pypi\n",
      "openai                    1.57.2                   pypi_0    pypi\n",
      "openssl                   3.4.0                hb9d3cd8_0    conda-forge\n",
      "orjson                    3.10.12                  pypi_0    pypi\n",
      "ossuuid                   1.6.2             hf484d3e_1000    conda-forge\n",
      "packaging                 24.2               pyhd8ed1ab_2    conda-forge\n",
      "pandas                    2.2.3                    pypi_0    pypi\n",
      "parso                     0.8.4              pyhd8ed1ab_1    conda-forge\n",
      "perl                      5.26.2            h36c2ea0_1008    conda-forge\n",
      "perl-app-cpanminus        1.7044                  pl526_1    bioconda\n",
      "perl-business-isbn        3.004                   pl526_0    bioconda\n",
      "perl-business-isbn-data   20140910.003            pl526_0    bioconda\n",
      "perl-carp                 1.38                    pl526_3    bioconda\n",
      "perl-constant             1.33                    pl526_1    bioconda\n",
      "perl-data-dumper          2.173                   pl526_0    bioconda\n",
      "perl-encode               2.88                    pl526_1    bioconda\n",
      "perl-exporter             5.72                    pl526_1    bioconda\n",
      "perl-extutils-makemaker   7.36                    pl526_1    bioconda\n",
      "perl-file-path            2.16                    pl526_0    bioconda\n",
      "perl-file-temp            0.2304                  pl526_2    bioconda\n",
      "perl-mime-base64          3.15                    pl526_1    bioconda\n",
      "perl-parent               0.236                   pl526_1    bioconda\n",
      "perl-uri                  1.76                    pl526_0    bioconda\n",
      "perl-xml-libxml           2.0132          pl526h7ec2d77_1    bioconda\n",
      "perl-xml-namespacesupport 1.12                    pl526_0    bioconda\n",
      "perl-xml-sax              1.02                    pl526_0    bioconda\n",
      "perl-xml-sax-base         1.09                    pl526_0    bioconda\n",
      "perl-xsloader             0.24                    pl526_0    bioconda\n",
      "pexpect                   4.9.0              pyhd8ed1ab_1    conda-forge\n",
      "pickleshare               0.7.5           pyhd8ed1ab_1004    conda-forge\n",
      "pip                       24.3.1             pyh8b19718_0    conda-forge\n",
      "platformdirs              4.3.6              pyhd8ed1ab_1    conda-forge\n",
      "prompt-toolkit            3.0.48             pyha770c72_1    conda-forge\n",
      "propcache                 0.2.1                    pypi_0    pypi\n",
      "proto-plus                1.25.0                   pypi_0    pypi\n",
      "protobuf                  5.29.1                   pypi_0    pypi\n",
      "psutil                    6.1.0           py312h66e93f0_0    conda-forge\n",
      "psycopg2-binary           2.9.10                   pypi_0    pypi\n",
      "ptyprocess                0.7.0              pyhd8ed1ab_1    conda-forge\n",
      "pure_eval                 0.2.3              pyhd8ed1ab_1    conda-forge\n",
      "pyarrow                   18.1.0                   pypi_0    pypi\n",
      "pyasn1                    0.6.1                    pypi_0    pypi\n",
      "pyasn1-modules            0.4.1                    pypi_0    pypi\n",
      "pydantic                  2.10.3                   pypi_0    pypi\n",
      "pydantic-core             2.27.1                   pypi_0    pypi\n",
      "pydantic-settings         2.7.0                    pypi_0    pypi\n",
      "pygments                  2.18.0             pyhd8ed1ab_1    conda-forge\n",
      "pypika                    0.48.9                   pypi_0    pypi\n",
      "pyproject-hooks           1.2.0                    pypi_0    pypi\n",
      "python                    3.12.8          h9e4cc4f_1_cpython    conda-forge\n",
      "python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge\n",
      "python-dotenv             1.0.1                    pypi_0    pypi\n",
      "python_abi                3.12                    5_cp312    conda-forge\n",
      "pytz                      2024.2                   pypi_0    pypi\n",
      "pyyaml                    6.0.2                    pypi_0    pypi\n",
      "pyzmq                     26.2.0          py312hbf22597_3    conda-forge\n",
      "readline                  8.2                  h8228510_1    conda-forge\n",
      "regex                     2024.11.6                pypi_0    pypi\n",
      "requests                  2.32.3                   pypi_0    pypi\n",
      "requests-toolbelt         1.0.0                    pypi_0    pypi\n",
      "rsa                       4.9                      pypi_0    pypi\n",
      "setuptools                75.6.0             pyhff2d567_1    conda-forge\n",
      "six                       1.17.0             pyhd8ed1ab_0    conda-forge\n",
      "sniffio                   1.3.1                    pypi_0    pypi\n",
      "soupsieve                 2.6                      pypi_0    pypi\n",
      "sqlalchemy                2.0.35                   pypi_0    pypi\n",
      "sra-tools                 3.1.1                h4304569_2    bioconda\n",
      "sragent                   0.1.0                    pypi_0    pypi\n",
      "stack_data                0.6.3              pyhd8ed1ab_1    conda-forge\n",
      "tabulate                  0.9.0                    pypi_0    pypi\n",
      "tenacity                  9.0.0                    pypi_0    pypi\n",
      "tiktoken                  0.8.0                    pypi_0    pypi\n",
      "tk                        8.6.13          noxft_h4845f30_101    conda-forge\n",
      "tornado                   6.4.2           py312h66e93f0_0    conda-forge\n",
      "tqdm                      4.67.1                   pypi_0    pypi\n",
      "traitlets                 5.14.3             pyhd8ed1ab_1    conda-forge\n",
      "types-requests            2.32.0.20241016          pypi_0    pypi\n",
      "typing-inspect            0.9.0                    pypi_0    pypi\n",
      "typing_extensions         4.12.2             pyha770c72_1    conda-forge\n",
      "tzdata                    2024.2                   pypi_0    pypi\n",
      "urllib3                   2.2.3                    pypi_0    pypi\n",
      "wcwidth                   0.2.13             pyhd8ed1ab_1    conda-forge\n",
      "wheel                     0.45.1             pyhd8ed1ab_1    conda-forge\n",
      "xmltodict                 0.14.2                   pypi_0    pypi\n",
      "xz                        5.6.3                hbcc6ac9_1    conda-forge\n",
      "xz-gpl-tools              5.6.3                hbcc6ac9_1    conda-forge\n",
      "xz-tools                  5.6.3                hb9d3cd8_1    conda-forge\n",
      "yarl                      1.18.3                   pypi_0    pypi\n",
      "zeromq                    4.3.5                h3b0a872_7    conda-forge\n",
      "zipp                      3.21.0             pyhd8ed1ab_1    conda-forge\n",
      "zstd                      1.5.6                ha6fb4c9_0    conda-forge\n"
     ]
    }
   ],
   "source": [
    "!conda list"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
